{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from multiprocessing import Pool\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from rdkit.Chem import rdMolDescriptors\n",
    "from rdkit.Chem import PandasTools\n",
    "from rdkit import Chem\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "import numpy as np\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "from utils.RMT import RMTClassifier\n",
    "\n",
    "import importlib\n",
    "importlib.reload(RMT)\n",
    "\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_df(df, drop_non_binders = True):\n",
    "\n",
    "    df_simple = df[['Molecule ChEMBL ID', 'Molecule Name',\n",
    "                          'Uo Units', 'Ligand Efficiency BEI', 'Smiles',\n",
    "                          'Standard Value', 'Standard Units']]\n",
    "\n",
    "    df_simple['Standard Value'] = pd.to_numeric(df_simple['Standard Value'],\n",
    "                                         errors = 'coerce')\n",
    "\n",
    "    #df_simple = df_simple.dropna()\n",
    "    df_simple = df_simple.reset_index(drop = True)\n",
    "\n",
    "    #print(df_simple['Standard Units'].unique())\n",
    "    \n",
    "    df_simple = df_simple[df_simple['Standard Units'] == 'nM']\n",
    "\n",
    "    \n",
    "    df_simple = df_simple.dropna(subset = ['Smiles'])\n",
    "\n",
    "    if drop_non_binders:\n",
    "        df_simple = df_simple[df_simple['Standard Value'] < 1000]\n",
    "\n",
    "    df_simple = df_simple.drop_duplicates(subset = 'Smiles',\n",
    "                                          keep = 'first')\n",
    "    \n",
    "    return df_simple\n",
    "\n",
    "\n",
    "def get_unfolded_fingerprint(smile):   \n",
    "    \n",
    "    compound = Chem.MolFromSmiles(smile)\n",
    "    #fp       = Chem.\\\n",
    "    #            .FingerprintMols.FingerprintMol(compound, 2)\n",
    "    \n",
    "    #fp       = Chem.AllChem.\\\n",
    "    #           GetMorganFingerprintAsBitVect(compound,\n",
    "    #                                        2, nBits = 2048)\n",
    "    \n",
    "    bi = {}\n",
    "    #fp = Chem.AllChem.GetMorganFingerprint(compound, radius=1, bitInfo=bi)\n",
    "            \n",
    "    #fp       = Chem.AllChem.GetMorganFingerprint(compound, 2)\n",
    "    \n",
    "    fp        = Chem.UnfoldedRDKFingerprintCountBased(compound, minPath = 1)\n",
    "\n",
    "    #fp       = Chem.RDKFingerprint(compound, maxPath = 10)\n",
    "\n",
    "    return fp\n",
    "\n",
    "\n",
    "def forrest_fingerprint(smile):\n",
    "    \n",
    "    \n",
    "    # Obtain dictionary of fingerprints and order\n",
    "    \n",
    "    bi = {}\n",
    "    Chem.AllChem.GetMorganFingerprint(Chem.MolFromSmiles(smile), radius=1, bitInfo=bi)\n",
    "\n",
    "    # Iterate over dictionary to store as pd.Series\n",
    "    mol = pd.DataFrame()\n",
    "    for key, value in bi.items():\n",
    "        for i in range(len(value)):\n",
    "            sub = pd.Series()\n",
    "            sub['val'] = str(key)\n",
    "\n",
    "            sub['order'] = value[i][0]\n",
    "            sub['radius'] = value[i][1]\n",
    "\n",
    "            mol = mol.append(sub, ignore_index=True)\n",
    "\n",
    "    # Correctly order fingerprint\n",
    "    mol = mol.sort_values(by=['order', 'radius']).reset_index(drop=True)\n",
    "\n",
    "    # Convert ordered fingerprint into single string\n",
    "    mol_string = \" \".join(mol.val.tolist())\n",
    "        \n",
    "    return mol_string.strip()\n",
    "\n",
    "\n",
    "def get_morgan_matrix(smile):\n",
    "    \n",
    "    compound = Chem.MolFromSmiles(smile)\n",
    "    \n",
    "    fp       = Chem.AllChem.\\\n",
    "               GetMorganFingerprintAsBitVect(compound,\n",
    "                                            2, nBits = 2048)\n",
    "    \n",
    "    \n",
    "    \n",
    "    return [int(i) for i in fp]\n",
    "\n",
    "\n",
    "def conv_rdkit_to_csr(rdkit_sparse):\n",
    "\n",
    "    data = np.array(list(rdkit_sparse.GetNonzeroElements().values()))\n",
    "    rows = np.array([0 for i in range(len(rdkit_sparse.GetNonzeroElements()))])\n",
    "    columns = np.array(list(rdkit_sparse.GetNonzeroElements().keys()))\n",
    "    \n",
    "    return csr_matrix((data, (rows,columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module rdkit.Chem.AllChem in rdkit.Chem:\n",
      "\n",
      "NAME\n",
      "    rdkit.Chem.AllChem - Import all RDKit chemistry modules\n",
      "\n",
      "FUNCTIONS\n",
      "    AddHs(...)\n",
      "        AddHs( (Mol)mol [, (bool)explicitOnly=False [, (bool)addCoords=False [, (AtomPairsParameters)onlyOnAtoms=None]]]) -> Mol :\n",
      "            Adds hydrogens to the graph of a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - explicitOnly: (optional) if this toggle is set, only explicit Hs will\n",
      "                  be added to the molecule.  Default value is 0 (add implicit and explicit Hs).\n",
      "            \n",
      "                - addCoords: (optional) if this toggle is set, The Hs will have 3D coordinates\n",
      "                  set.  Default value is 0 (no 3D coords).\n",
      "            \n",
      "                - onlyOnHs: (optional) if this sequence is provided, only these atoms will be\n",
      "                  considered to have Hs added to them\n",
      "            \n",
      "              RETURNS: a new molecule with added Hs\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "                - Much of the code assumes that Hs are not included in the molecular\n",
      "                  topology, so be *very* careful with the molecule that comes back from\n",
      "                  this function.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* AddHs(RDKit::ROMol [,bool=False [,bool=False [,boost::python::api::object=None]]])\n",
      "    \n",
      "    AddRecursiveQuery(...)\n",
      "        AddRecursiveQuery( (Mol)mol, (Mol)query, (int)atomIdx [, (bool)preserveExistingQuery=True]) -> None :\n",
      "            Adds a recursive query to an atom\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as the recursive query (this will be copied)\n",
      "            \n",
      "                - atomIdx: the atom to modify\n",
      "            \n",
      "                - preserveExistingQuery: (optional) if this is set, existing query information on the atom will be preserved\n",
      "            \n",
      "              RETURNS: None\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AddRecursiveQuery(RDKit::ROMol {lvalue},RDKit::ROMol,unsigned int [,bool=True])\n",
      "    \n",
      "    AdjustQueryProperties(...)\n",
      "        AdjustQueryProperties( (Mol)mol [, (AtomPairsParameters)params=None]) -> Mol :\n",
      "            Returns a new molecule where the query properties of atoms have been modified.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* AdjustQueryProperties(RDKit::ROMol [,boost::python::api::object=None])\n",
      "    \n",
      "    AlignMol(...)\n",
      "        AlignMol( (Mol)prbMol, (Mol)refMol [, (int)prbCid=-1 [, (int)refCid=-1 [, (AtomPairsParameters)atomMap=[] [, (AtomPairsParameters)weights=[] [, (bool)reflect=False [, (int)maxIters=50]]]]]]) -> float :\n",
      "            Optimally (minimum RMSD) align a molecule to another molecule\n",
      "                 \n",
      "                  The 3D transformation required to align the specied conformation in the probe molecule\n",
      "                  to a specified conformation in the reference molecule is computed so that the root mean\n",
      "                  squared distance between a specified set of atoms is minimized. \n",
      "                  This transform is then applied to the specified conformation in the probe molecule\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol    molecule that is to be aligned\n",
      "                  - refMol    molecule used as the reference for the alignment\n",
      "                  - prbCid    ID of the conformation in the probe to be used \n",
      "                                   for the alignment (defaults to first conformation)\n",
      "                  - refCid    ID of the conformation in the ref molecule to which \n",
      "                                   the alignment is computed (defaults to first conformation)\n",
      "                  - atomMap   a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                   used to compute the alignments. If this mapping is \n",
      "                                   not specified an attempt is made to generate on by\n",
      "                                   substructure matching\n",
      "                  - weights   Optionally specify weights for each of the atom pairs\n",
      "                  - reflect   if true reflect the conformation of the probe molecule\n",
      "                  - maxIters  maximum number of iterations used in mimizing the RMSD\n",
      "                   \n",
      "                  RETURNS\n",
      "                  RMSD value\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double AlignMol(RDKit::ROMol {lvalue},RDKit::ROMol [,int=-1 [,int=-1 [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,unsigned int=50]]]]]])\n",
      "    \n",
      "    AlignMolConformers(...)\n",
      "        AlignMolConformers( (Mol)mol [, (AtomPairsParameters)atomIds=[] [, (AtomPairsParameters)confIds=[] [, (AtomPairsParameters)weights=[] [, (bool)reflect=False [, (int)maxIters=50 [, (AtomPairsParameters)RMSlist=None]]]]]]) -> None :\n",
      "            Alignment conformations in a molecule to each other\n",
      "                 \n",
      "                  The first conformation in the molecule is used as the reference\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - mol       molecule of interest\n",
      "                  - atomIds   List of atom ids to use a points for alingment - defaults to all atoms\n",
      "                  - confIds   Ids of conformations to align - defaults to all conformers \n",
      "                  - weights   Optionally specify weights for each of the atom pairs\n",
      "                  - reflect   if true reflect the conformation of the probe molecule\n",
      "                  - maxIters  maximum number of iterations used in mimizing the RMSD\n",
      "                  - RMSlist   if provided, fills in the RMS values between the reference\n",
      "                          conformation and the other aligned conformations\n",
      "                   \n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AlignMolConformers(RDKit::ROMol {lvalue} [,boost::python::api::object=[] [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,unsigned int=50 [,boost::python::api::object=None]]]]]])\n",
      "    \n",
      "    AssignAtomChiralTagsFromStructure(...)\n",
      "        AssignAtomChiralTagsFromStructure( (Mol)mol [, (int)confId=-1 [, (bool)replaceExistingTags=True]]) -> None :\n",
      "            Sets the chiral tags on a molecule's atoms based on \n",
      "              a 3D conformation.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - confId: the conformer id to use, -1 for the default \n",
      "                - replaceExistingTags: if True, existing stereochemistry information will be cleared \n",
      "                                       before running the calculation. \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignAtomChiralTagsFromStructure(RDKit::ROMol {lvalue} [,int=-1 [,bool=True]])\n",
      "    \n",
      "    AssignBondOrdersFromTemplate(refmol, mol)\n",
      "        assigns bond orders to a molecule based on the\n",
      "            bond orders in a template molecule\n",
      "        \n",
      "        Arguments\n",
      "          - refmol: the template molecule\n",
      "          - mol: the molecule to assign bond orders to\n",
      "        \n",
      "          An example, start by generating a template from a SMILES\n",
      "          and read in the PDB structure of the molecule\n",
      "          >>> import os\n",
      "          >>> from rdkit.Chem import AllChem\n",
      "          >>> template = AllChem.MolFromSmiles(\"CN1C(=NC(C1=O)(c2ccccc2)c3ccccc3)N\")\n",
      "          >>> mol = AllChem.MolFromPDBFile(os.path.join(RDConfig.RDCodeDir, 'Chem', 'test_data', '4DJU_lig.pdb'))\n",
      "          >>> len([1 for b in template.GetBonds() if b.GetBondTypeAsDouble() == 1.0])\n",
      "          8\n",
      "          >>> len([1 for b in mol.GetBonds() if b.GetBondTypeAsDouble() == 1.0])\n",
      "          22\n",
      "        \n",
      "          Now assign the bond orders based on the template molecule\n",
      "          >>> newMol = AllChem.AssignBondOrdersFromTemplate(template, mol)\n",
      "          >>> len([1 for b in newMol.GetBonds() if b.GetBondTypeAsDouble() == 1.0])\n",
      "          8\n",
      "        \n",
      "          Note that the template molecule should have no explicit hydrogens\n",
      "          else the algorithm will fail.\n",
      "        \n",
      "          It also works if there are different formal charges (this was github issue 235):\n",
      "          >>> template=AllChem.MolFromSmiles('CN(C)C(=O)Cc1ccc2c(c1)NC(=O)c3ccc(cc3N2)c4ccc(c(c4)OC)[N+](=O)[O-]')\n",
      "          >>> mol = AllChem.MolFromMolFile(os.path.join(RDConfig.RDCodeDir, 'Chem', 'test_data', '4FTR_lig.mol'))\n",
      "          >>> AllChem.MolToSmiles(mol)\n",
      "          'COC1CC(C2CCC3C(O)NC4CC(CC(O)N(C)C)CCC4NC3C2)CCC1N(O)O'\n",
      "          >>> newMol = AllChem.AssignBondOrdersFromTemplate(template, mol)\n",
      "          >>> AllChem.MolToSmiles(newMol)\n",
      "          'COc1cc(-c2ccc3c(c2)Nc2ccc(CC(=O)N(C)C)cc2NC3=O)ccc1[N+](=O)[O-]'\n",
      "    \n",
      "    AssignRadicals(...)\n",
      "        AssignRadicals( (Mol)mol) -> None :\n",
      "            Assigns radical counts to atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignRadicals(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    AssignStereochemistry(...)\n",
      "        AssignStereochemistry( (Mol)mol [, (bool)cleanIt=False [, (bool)force=False [, (bool)flagPossibleStereoCenters=False]]]) -> None :\n",
      "            Does the CIP stereochemistry assignment \n",
      "              for the molecule's atoms (R/S) and double bond (Z/E).\n",
      "              Chiral atoms will have a property '_CIPCode' indicating\n",
      "              their chiral code.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - cleanIt: (optional) if provided, atoms with a chiral specifier that aren't\n",
      "                  actually chiral (e.g. atoms with duplicate substituents or only 2 substituents,\n",
      "                  etc.) will have their chiral code set to CHI_UNSPECIFIED\n",
      "                - force: (optional) causes the calculation to be repeated, even if it has already\n",
      "                  been done\n",
      "                - flagPossibleStereoCenters (optional)   set the _ChiralityPossible property on\n",
      "                  atoms that are possible stereocenters\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignStereochemistry(RDKit::ROMol {lvalue} [,bool=False [,bool=False [,bool=False]]])\n",
      "    \n",
      "    AtomNumEqualsQueryAtom(...)\n",
      "        AtomNumEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where AtomNum is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* AtomNumEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    AtomNumGreaterQueryAtom(...)\n",
      "        AtomNumGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where AtomNum is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* AtomNumGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    AtomNumLessQueryAtom(...)\n",
      "        AtomNumLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where AtomNum is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* AtomNumLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    BuildFeatureFactory(...)\n",
      "        BuildFeatureFactory( (str)arg1) -> MolChemicalFeatureFactory :\n",
      "            Construct a feature factory given a feature definition in a file\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolChemicalFeatureFactory* BuildFeatureFactory(std::string)\n",
      "    \n",
      "    BuildFeatureFactoryFromString(...)\n",
      "        BuildFeatureFactoryFromString( (str)arg1) -> MolChemicalFeatureFactory :\n",
      "            Construct a feature factory given a feature definition block\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolChemicalFeatureFactory* BuildFeatureFactoryFromString(std::string)\n",
      "    \n",
      "    CalcAUTOCORR2D(...)\n",
      "        CalcAUTOCORR2D( (Mol)mol) -> list :\n",
      "            Returns 2D Autocorrelation descriptors vector\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcAUTOCORR2D(RDKit::ROMol)\n",
      "    \n",
      "    CalcAUTOCORR3D(...)\n",
      "        CalcAUTOCORR3D( (Mol)mol [, (int)confId=-1]) -> list :\n",
      "            Returns 3D Autocorrelation descriptors vector\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcAUTOCORR3D(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    CalcAsphericity(...)\n",
      "        CalcAsphericity( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcAsphericity(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcChi0n(...)\n",
      "        CalcChi0n( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi0n(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi0v(...)\n",
      "        CalcChi0v( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi0v(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi1n(...)\n",
      "        CalcChi1n( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi1n(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi1v(...)\n",
      "        CalcChi1v( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi1v(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi2n(...)\n",
      "        CalcChi2n( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi2n(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi2v(...)\n",
      "        CalcChi2v( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi2v(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi3n(...)\n",
      "        CalcChi3n( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi3n(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi3v(...)\n",
      "        CalcChi3v( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi3v(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi4n(...)\n",
      "        CalcChi4n( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi4n(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChi4v(...)\n",
      "        CalcChi4v( (Mol)mol [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChi4v(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcChiNn(...)\n",
      "        CalcChiNn( (Mol)mol, (int)n [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChiNn(RDKit::ROMol,unsigned int [,bool=False])\n",
      "    \n",
      "    CalcChiNv(...)\n",
      "        CalcChiNv( (Mol)mol, (int)n [, (bool)force=False]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcChiNv(RDKit::ROMol,unsigned int [,bool=False])\n",
      "    \n",
      "    CalcCrippenDescriptors(...)\n",
      "        CalcCrippenDescriptors( (Mol)mol [, (bool)includeHs=True [, (bool)force=False]]) -> tuple :\n",
      "            returns a 2-tuple with the Wildman-Crippen logp,mr values\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple CalcCrippenDescriptors(RDKit::ROMol [,bool=True [,bool=False]])\n",
      "    \n",
      "    CalcEccentricity(...)\n",
      "        CalcEccentricity( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcEccentricity(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcExactMolWt(...)\n",
      "        CalcExactMolWt( (Mol)mol [, (bool)onlyHeavy=False]) -> float :\n",
      "            returns the molecule's exact molecular weight\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcExactMolWt(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcFractionCSP3(...)\n",
      "        CalcFractionCSP3( (Mol)mol) -> float :\n",
      "            returns the fraction of C atoms that are SP3 hybridized\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcFractionCSP3(RDKit::ROMol)\n",
      "    \n",
      "    CalcGETAWAY(...)\n",
      "        CalcGETAWAY( (Mol)mol [, (int)confId=-1 [, (float)precision=2]]) -> list :\n",
      "            Returns the GETAWAY descriptors vector\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcGETAWAY(RDKit::ROMol [,int=-1 [,double=2]])\n",
      "    \n",
      "    CalcHallKierAlpha(...)\n",
      "        CalcHallKierAlpha( (Mol)mol [, (AtomPairsParameters)atomContribs=None]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcHallKierAlpha(RDKit::ROMol [,boost::python::api::object=None])\n",
      "    \n",
      "    CalcInertialShapeFactor(...)\n",
      "        CalcInertialShapeFactor( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcInertialShapeFactor(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcKappa1(...)\n",
      "        CalcKappa1( (Mol)mol) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcKappa1(RDKit::ROMol)\n",
      "    \n",
      "    CalcKappa2(...)\n",
      "        CalcKappa2( (Mol)mol) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcKappa2(RDKit::ROMol)\n",
      "    \n",
      "    CalcKappa3(...)\n",
      "        CalcKappa3( (Mol)mol) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcKappa3(RDKit::ROMol)\n",
      "    \n",
      "    CalcLabuteASA(...)\n",
      "        CalcLabuteASA( (Mol)mol [, (bool)includeHs=True [, (bool)force=False]]) -> float :\n",
      "            returns the Labute ASA value for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcLabuteASA(RDKit::ROMol [,bool=True [,bool=False]])\n",
      "    \n",
      "    CalcMORSE(...)\n",
      "        CalcMORSE( (Mol)mol [, (int)confId=-1]) -> list :\n",
      "            Returns Molecule Representation of Structures based on Electron diffraction descriptors\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcMORSE(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    CalcMolFormula(...)\n",
      "        CalcMolFormula( (Mol)mol [, (bool)separateIsotopes=False [, (bool)abbreviateHIsotopes=True]]) -> str :\n",
      "            returns the molecule's formula\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string CalcMolFormula(RDKit::ROMol [,bool=False [,bool=True]])\n",
      "    \n",
      "    CalcNPR1(...)\n",
      "        CalcNPR1( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcNPR1(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcNPR2(...)\n",
      "        CalcNPR2( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcNPR2(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcNumAliphaticCarbocycles(...)\n",
      "        CalcNumAliphaticCarbocycles( (Mol)mol) -> int :\n",
      "            returns the number of aliphatic (containing at least one non-aromatic bond) carbocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAliphaticCarbocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAliphaticHeterocycles(...)\n",
      "        CalcNumAliphaticHeterocycles( (Mol)mol) -> int :\n",
      "            returns the number of aliphatic (containing at least one non-aromatic bond) heterocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAliphaticHeterocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAliphaticRings(...)\n",
      "        CalcNumAliphaticRings( (Mol)mol) -> int :\n",
      "            returns the number of aliphatic (containing at least one non-aromatic bond) rings for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAliphaticRings(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAmideBonds(...)\n",
      "        CalcNumAmideBonds( (Mol)mol) -> int :\n",
      "            returns the number of amide bonds in a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAmideBonds(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAromaticCarbocycles(...)\n",
      "        CalcNumAromaticCarbocycles( (Mol)mol) -> int :\n",
      "            returns the number of aromatic carbocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAromaticCarbocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAromaticHeterocycles(...)\n",
      "        CalcNumAromaticHeterocycles( (Mol)mol) -> int :\n",
      "            returns the number of aromatic heterocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAromaticHeterocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAromaticRings(...)\n",
      "        CalcNumAromaticRings( (Mol)mol) -> int :\n",
      "            returns the number of aromatic rings for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAromaticRings(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumAtomStereoCenters(...)\n",
      "        CalcNumAtomStereoCenters( (Mol)mol) -> int :\n",
      "            Returns the total number of atomic stereocenters (specified and unspecified)\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumAtomStereoCenters(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumBridgeheadAtoms(...)\n",
      "        CalcNumBridgeheadAtoms( (Mol)mol [, (AtomPairsParameters)atoms=None]) -> int :\n",
      "            Returns the number of bridgehead atoms (atoms shared between rings that share at least two bonds)\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumBridgeheadAtoms(RDKit::ROMol [,boost::python::api::object=None])\n",
      "    \n",
      "    CalcNumHBA(...)\n",
      "        CalcNumHBA( (Mol)mol) -> int :\n",
      "            returns the number of H-bond acceptors for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumHBA(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumHBD(...)\n",
      "        CalcNumHBD( (Mol)mol) -> int :\n",
      "            returns the number of H-bond donors for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumHBD(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumHeteroatoms(...)\n",
      "        CalcNumHeteroatoms( (Mol)mol) -> int :\n",
      "            returns the number of heteroatoms for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumHeteroatoms(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumHeterocycles(...)\n",
      "        CalcNumHeterocycles( (Mol)mol) -> int :\n",
      "            returns the number of heterocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumHeterocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumLipinskiHBA(...)\n",
      "        CalcNumLipinskiHBA( (Mol)mol) -> int :\n",
      "            returns the number of Lipinski H-bond acceptors for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumLipinskiHBA(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumLipinskiHBD(...)\n",
      "        CalcNumLipinskiHBD( (Mol)mol) -> int :\n",
      "            returns the number of Lipinski H-bond donors for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumLipinskiHBD(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumRings(...)\n",
      "        CalcNumRings( (Mol)mol) -> int :\n",
      "            returns the number of rings for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumRings(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumRotatableBonds(...)\n",
      "        CalcNumRotatableBonds( (Mol)mol, (bool)strict) -> int :\n",
      "            returns the number of rotatable bonds for a molecule.\n",
      "               strict = NumRotatableBondsOptions.NonStrict - Simple rotatable bond definition.\n",
      "               strict = NumRotatableBondsOptions.Strict - (default) does not count things like\n",
      "                        amide or ester bonds\n",
      "               strict = NumRotatableBondsOptions.StrictLinkages - handles linkages between ring\n",
      "                  systems.\n",
      "                  - Single bonds between aliphatic ring Cs are always rotatable. This\n",
      "                    means that the central bond in CC1CCCC(C)C1-C1C(C)CCCC1C is now \n",
      "                    considered rotatable; it was not before\n",
      "                  - Heteroatoms in the linked rings no longer affect whether or not\n",
      "                    the linking bond is rotatable\n",
      "                  - the linking bond in systems like Cc1cccc(C)c1-c1c(C)cccc1 is now\n",
      "                     considered non-rotatable\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumRotatableBonds(RDKit::ROMol,bool)\n",
      "        \n",
      "        CalcNumRotatableBonds( (Mol)mol [, (NumRotatableBondsOptions)strict=rdkit.Chem.rdMolDescriptors.NumRotatableBondsOptions.Default]) -> int :\n",
      "            returns the number of rotatable bonds for a molecule.\n",
      "               strict = NumRotatableBondsOptions.NonStrict - Simple rotatable bond definition.\n",
      "               strict = NumRotatableBondsOptions.Strict - (default) does not count things like\n",
      "                        amide or ester bonds\n",
      "               strict = NumRotatableBondsOptions.StrictLinkages - handles linkages between ring\n",
      "                  systems.\n",
      "                  - Single bonds between aliphatic ring Cs are always rotatable. This\n",
      "                    means that the central bond in CC1CCCC(C)C1-C1C(C)CCCC1C is now \n",
      "                    considered rotatable; it was not before\n",
      "                  - Heteroatoms in the linked rings no longer affect whether or not\n",
      "                    the linking bond is rotatable\n",
      "                  - the linking bond in systems like Cc1cccc(C)c1-c1c(C)cccc1 is now\n",
      "                     considered non-rotatable\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumRotatableBonds(RDKit::ROMol [,RDKit::Descriptors::NumRotatableBondsOptions=rdkit.Chem.rdMolDescriptors.NumRotatableBondsOptions.Default])\n",
      "    \n",
      "    CalcNumSaturatedCarbocycles(...)\n",
      "        CalcNumSaturatedCarbocycles( (Mol)mol) -> int :\n",
      "            returns the number of saturated carbocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumSaturatedCarbocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumSaturatedHeterocycles(...)\n",
      "        CalcNumSaturatedHeterocycles( (Mol)mol) -> int :\n",
      "            returns the number of saturated heterocycles for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumSaturatedHeterocycles(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumSaturatedRings(...)\n",
      "        CalcNumSaturatedRings( (Mol)mol) -> int :\n",
      "            returns the number of saturated rings for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumSaturatedRings(RDKit::ROMol)\n",
      "    \n",
      "    CalcNumSpiroAtoms(...)\n",
      "        CalcNumSpiroAtoms( (Mol)mol [, (AtomPairsParameters)atoms=None]) -> int :\n",
      "            Returns the number of spiro atoms (atoms shared between rings that share exactly one atom)\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumSpiroAtoms(RDKit::ROMol [,boost::python::api::object=None])\n",
      "    \n",
      "    CalcNumUnspecifiedAtomStereoCenters(...)\n",
      "        CalcNumUnspecifiedAtomStereoCenters( (Mol)mol) -> int :\n",
      "            Returns the number of unspecified atomic stereocenters\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int CalcNumUnspecifiedAtomStereoCenters(RDKit::ROMol)\n",
      "    \n",
      "    CalcPBF(...)\n",
      "        CalcPBF( (Mol)mol [, (int)confId=-1]) -> float :\n",
      "            Returns the PBF (plane of best fit) descriptor (http://dx.doi.org/10.1021/ci300293f)\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcPBF(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    CalcPMI1(...)\n",
      "        CalcPMI1( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcPMI1(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcPMI2(...)\n",
      "        CalcPMI2( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcPMI2(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcPMI3(...)\n",
      "        CalcPMI3( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcPMI3(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcRDF(...)\n",
      "        CalcRDF( (Mol)mol [, (int)confId=-1]) -> list :\n",
      "            Returns radial distribution fonction descriptors (RDF)\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcRDF(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    CalcRadiusOfGyration(...)\n",
      "        CalcRadiusOfGyration( (Mol)mol [, (int)confId=-1 [, (bool)useAtomicMasses=True]]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcRadiusOfGyration(RDKit::ROMol [,int=-1 [,bool=True]])\n",
      "    \n",
      "    CalcSpherocityIndex(...)\n",
      "        CalcSpherocityIndex( (Mol)mol [, (int)confId=-1]) -> float :\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcSpherocityIndex(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    CalcTPSA(...)\n",
      "        CalcTPSA( (Mol)mol [, (bool)force=False]) -> float :\n",
      "            returns the TPSA value for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                double CalcTPSA(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    CalcWHIM(...)\n",
      "        CalcWHIM( (Mol)mol [, (int)confId=-1 [, (float)thresh=0.001]]) -> list :\n",
      "            Returns the WHIM descriptors vector\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list CalcWHIM(RDKit::ROMol [,int=-1 [,double=0.001]])\n",
      "    \n",
      "    CanonicalRankAtoms(...)\n",
      "        CanonicalRankAtoms( (Mol)mol [, (bool)breakTies=True [, (bool)includeChirality=True [, (bool)includeIsotopes=True]]]) -> _vectj :\n",
      "            Returns the canonical atom ranking for each atom of a molecule fragment.\n",
      "              If breakTies is False, this returns the symmetry class for each atom.  The symmetry\n",
      "              class is used by the canonicalization routines to type each atom based on the whole\n",
      "              chemistry of the molecular graph.  Any atom with the same rank (symmetry class) is\n",
      "              indistinguishable.  For example:\n",
      "            \n",
      "                > mol = MolFromSmiles('C1NCN1')\n",
      "                > list(CanonicalRankAtoms(mol, breakTies=False))\n",
      "                [0,1,0,1]\n",
      "            \n",
      "              In this case the carbons have the same symmetry class and the nitrogens have the same\n",
      "              symmetry class.  From the perspective of the Molecular Graph, they are indentical.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - breakTies: (optional) force breaking of ranked ties [default=True]\n",
      "                - includeChirality: (optional) use chiral information when computing rank [default=True]\n",
      "                - includeIsotopes: (optional) use isotope information when computing rank [default=True]\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<unsigned int, std::allocator<unsigned int> > CanonicalRankAtoms(RDKit::ROMol [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    CanonicalRankAtomsInFragment(...)\n",
      "        CanonicalRankAtomsInFragment( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (AtomPairsParameters)atomSymbols=0 [, (AtomPairsParameters)bondSymbols=0 [, (bool)breakTies=True]]]]) -> _vecti :\n",
      "            Returns the canonical atom ranking for each atom of a molecule fragment\n",
      "              See help(CanonicalRankAtoms) for more information.\n",
      "            \n",
      "               > mol = MolFromSmiles('C1NCN1.C1NCN1')\n",
      "               > list(CanonicalRankAtomsInFragment(mol, atomsToUse=range(0,4), breakTies=False))\n",
      "               [0,1,0,1,-1,-1,-1,-1]\n",
      "               > list(CanonicalRankAtomsInFragment(mol, atomsToUse=range(4,8), breakTies=False))\n",
      "               [-1,-1,-1,-1,0,1,0,1]\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse : a list of atoms to include in the fragment\n",
      "                - bondsToUse : (optional) a list of bonds to include in the fragment\n",
      "                               if not provided, all bonds between the atoms provided\n",
      "                               will be included.\n",
      "                - atomSymbols : (optional) a list with the symbols to use for the atoms\n",
      "                                in the SMILES. This should have be mol.GetNumAtoms() long.\n",
      "                - bondSymbols : (optional) a list with the symbols to use for the bonds\n",
      "                                in the SMILES. This should have be mol.GetNumBonds() long.\n",
      "                - breakTies: (optional) force breaking of ranked ties\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<int, std::allocator<int> > CanonicalRankAtomsInFragment(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=True]]]])\n",
      "    \n",
      "    CanonicalizeConformer(...)\n",
      "        CanonicalizeConformer( (Conformer)conf [, (Point3D)center=None [, (bool)normalizeCovar=False [, (bool)ignoreHs=True]]]) -> None :\n",
      "            Canonicalize the orientation of a conformer so that its principal axes\n",
      "                           around the specified center point coincide with the x, y, z axes\n",
      "              \n",
      "              ARGUMENTS:\n",
      "                - conf : conformer of interest \n",
      "                - center : optionally center point about which the principal axes are computed \n",
      "                                      if not specified the centroid of the conformer will be used\n",
      "                - normalizeCovar : Optionally normalize the covariance matrix by the number of atoms\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void CanonicalizeConformer(RDKit::Conformer {lvalue} [,RDGeom::Point3D const*=None [,bool=False [,bool=True]]])\n",
      "    \n",
      "    CanonicalizeMol(...)\n",
      "        CanonicalizeMol( (Mol)mol [, (bool)normalizeCovar=False [, (bool)ignoreHs=True]]) -> None :\n",
      "            Loop over the conformers in a molecule and canonicalize their orientation\n",
      "        \n",
      "            C++ signature :\n",
      "                void CanonicalizeMol(RDKit::ROMol {lvalue} [,bool=False [,bool=True]])\n",
      "    \n",
      "    Cleanup(...)\n",
      "        Cleanup( (Mol)mol) -> None :\n",
      "            cleans up certain common bad functionalities in the molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void Cleanup(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    CombineMols(...)\n",
      "        CombineMols( (Mol)mol1, (Mol)mol2 [, (Point3D)offset=<rdkit.Geometry.rdGeometry.Point3D object at 0x7fdd89af8030>]) -> Mol :\n",
      "            Combine the atoms from two molecules to produce a third\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* CombineMols(RDKit::ROMol,RDKit::ROMol [,RDGeom::Point3D=<rdkit.Geometry.rdGeometry.Point3D object at 0x7fdd89af8030>])\n",
      "    \n",
      "    Compute2DCoords(...)\n",
      "        Compute2DCoords( (Mol)mol [, (bool)canonOrient=True [, (bool)clearConfs=True [, (dict)coordMap={} [, (int)nFlipsPerSample=0 [, (int)nSample=0 [, (int)sampleSeed=0 [, (bool)permuteDeg4Nodes=False [, (float)bondLength=-1.0]]]]]]]]) -> int :\n",
      "            Compute 2D coordinates for a molecule. \n",
      "              The resulting coordinates are stored on each atom of the molecule \n",
      "            \n",
      "              ARGUMENTS: \n",
      "            \n",
      "                 mol - the molecule of interest\n",
      "                 canonOrient - orient the molecule in a canonical way\n",
      "                 clearConfs - if true, all existing conformations on the molecule\n",
      "                         will be cleared\n",
      "                 coordMap - a dictionary mapping atom Ids -> Point2D objects \n",
      "                            with starting coordinates for atoms that should\n",
      "                            have their positions locked.\n",
      "                 nFlipsPerSample - number of rotatable bonds that are\n",
      "                            flipped at random at a time.\n",
      "                 nSample - Number of random samplings of rotatable bonds.\n",
      "                 sampleSeed - seed for the random sampling process.\n",
      "                 permuteDeg4Nodes - allow permutation of bonds at a degree 4\n",
      "                             node during the sampling process \n",
      "                 bondLength - change the default bond length for depiction \n",
      "            \n",
      "              RETURNS: \n",
      "            \n",
      "                 ID of the conformation added to the molecule\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int Compute2DCoords(RDKit::ROMol {lvalue} [,bool=True [,bool=True [,boost::python::dict {lvalue}={} [,unsigned int=0 [,unsigned int=0 [,int=0 [,bool=False [,double=-1.0]]]]]]]])\n",
      "    \n",
      "    Compute2DCoordsForReaction(...)\n",
      "        Compute2DCoordsForReaction( (ChemicalReaction)reaction [, (float)spacing=2.0 [, (bool)updateProps=True [, (bool)canonOrient=True [, (int)nFlipsPerSample=0 [, (int)nSample=0 [, (int)sampleSeed=0 [, (bool)permuteDeg4Nodes=False [, (float)bondLength=-1.0]]]]]]]]) -> None :\n",
      "            Compute 2D coordinates for a reaction. \n",
      "              ARGUMENTS: \n",
      "            \n",
      "                 reaction - the reaction of interest\n",
      "                 spacing - the amount of space left between components of the reaction\n",
      "                 canonOrient - orient the reactants and products in a canonical way\n",
      "                 updateProps - if set, properties such as conjugation and\n",
      "                    hybridization will be calculated for the reactant and product\n",
      "                    templates before generating coordinates. This should result in\n",
      "                    better depictions, but can lead to errors in some cases.\n",
      "                 nFlipsPerSample - number of rotatable bonds that are\n",
      "                            flipped at random at a time.\n",
      "                 nSample - Number of random samplings of rotatable bonds.\n",
      "                 sampleSeed - seed for the random sampling process.\n",
      "                 permuteDeg4Nodes - allow permutation of bonds at a degree 4\n",
      "                             node during the sampling process \n",
      "                 bondLength - change the default bond length for depiction \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void Compute2DCoordsForReaction(RDKit::ChemicalReaction {lvalue} [,double=2.0 [,bool=True [,bool=True [,unsigned int=0 [,unsigned int=0 [,int=0 [,bool=False [,double=-1.0]]]]]]]])\n",
      "    \n",
      "    Compute2DCoordsMimicDistmat(...)\n",
      "        Compute2DCoordsMimicDistmat( (Mol)mol, (AtomPairsParameters)distMat [, (bool)canonOrient=False [, (bool)clearConfs=True [, (float)weightDistMat=0.5 [, (int)nFlipsPerSample=3 [, (int)nSample=100 [, (int)sampleSeed=100 [, (bool)permuteDeg4Nodes=True [, (float)bondLength=-1.0]]]]]]]]) -> int :\n",
      "            Compute 2D coordinates for a molecule such \n",
      "              that the inter-atom distances mimic those in a user-provided\n",
      "              distance matrix. \n",
      "              The resulting coordinates are stored on each atom of the molecule \n",
      "            \n",
      "              ARGUMENTS: \n",
      "            \n",
      "                 mol - the molecule of interest\n",
      "                 distMat - distance matrix that we want the 2D structure to mimic\n",
      "                 canonOrient - orient the molecule in a canonical way\n",
      "                 clearConfs - if true, all existing conformations on the molecule\n",
      "                         will be cleared\n",
      "                 weightDistMat - weight assigned in the cost function to mimicing\n",
      "                                 the distance matrix.\n",
      "                                 This must be between (0.0,1.0). (1.0-weightDistMat)\n",
      "                                 is then the weight assigned to improving \n",
      "                                 the density of the 2D structure i.e. try to\n",
      "                                 make it spread out\n",
      "                 nFlipsPerSample - number of rotatable bonds that are\n",
      "                            flipped at random at a time.\n",
      "                 nSample - Number of random samplings of rotatable bonds.\n",
      "                 sampleSeed - seed for the random sampling process.\n",
      "                 permuteDeg4Nodes - allow permutation of bonds at a degree 4\n",
      "                             node during the sampling process \n",
      "                 bondLength - change the default bond length for depiction \n",
      "            \n",
      "              RETURNS: \n",
      "            \n",
      "                 ID of the conformation added to the molecule\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int Compute2DCoordsMimicDistmat(RDKit::ROMol {lvalue},boost::python::api::object [,bool=False [,bool=True [,double=0.5 [,unsigned int=3 [,unsigned int=100 [,int=100 [,bool=True [,double=-1.0]]]]]]]])\n",
      "    \n",
      "    ComputeCanonicalTransform(...)\n",
      "        ComputeCanonicalTransform( (Conformer)conf [, (Point3D)center=None [, (bool)normalizeCovar=False [, (bool)ignoreHs=True]]]) -> object :\n",
      "            Compute the transformation required aligna conformer so that\n",
      "                           the the principal axes align up with the x,y, z axes\n",
      "                           The conformer itself is left unchanged\n",
      "              ARGUMENTS:\n",
      "                - conf : the conformer of interest\n",
      "                - center : optional center point to compute the principal axes around (defaults to the centroid)\n",
      "                - normalizeCovar : optionally normalize the covariance matrix by the number of atoms\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* ComputeCanonicalTransform(RDKit::Conformer [,RDGeom::Point3D const*=None [,bool=False [,bool=True]]])\n",
      "    \n",
      "    ComputeCentroid(...)\n",
      "        ComputeCentroid( (Conformer)conf [, (bool)ignoreHs=True]) -> Point3D :\n",
      "            Compute the centroid of the conformation - hydrogens are ignored and no attention\n",
      "                                       if paid to the difference in sizes of the heavy atoms\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDGeom::Point3D ComputeCentroid(RDKit::Conformer [,bool=True])\n",
      "    \n",
      "    ComputeConfBox(...)\n",
      "        ComputeConfBox( (Conformer)conf [, (AtomPairsParameters)trans=None [, (float)padding=2.0]]) -> tuple :\n",
      "            Compute the lower and upper corners of a cuboid that will fit the conformer\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple ComputeConfBox(RDKit::Conformer [,boost::python::api::object=None [,double=2.0]])\n",
      "    \n",
      "    ComputeConfDimsAndOffset(...)\n",
      "        ComputeConfDimsAndOffset( (Conformer)conf [, (AtomPairsParameters)trans=None [, (float)padding=2.0]]) -> tuple :\n",
      "            Compute the size of the box that can fit the conformations, and offset \n",
      "               of the box from the origin\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple ComputeConfDimsAndOffset(RDKit::Conformer [,boost::python::api::object=None [,double=2.0]])\n",
      "    \n",
      "    ComputeGasteigerCharges(...)\n",
      "        ComputeGasteigerCharges( (Mol)mol [, (int)nIter=12 [, (bool)throwOnParamFailure=False]]) -> None :\n",
      "            Compute Gasteiger partial charges for molecule\n",
      "            \n",
      "             The charges are computed using an iterative procedure presented in \n",
      "             \n",
      "             Ref : J.Gasteiger, M. Marseli, Iterative Equalization of Oribital Electronegatiity \n",
      "             A Rapid Access to Atomic Charges, Tetrahedron Vol 36 p3219 1980\n",
      "             \n",
      "             The computed charges are stored on each atom are stored a computed property ( under the name \n",
      "             _GasteigerCharge). In addition, each atom also stored the total charge for the implicit hydrogens \n",
      "             on the atom (under the property name _GasteigerHCharge)\n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interrest\n",
      "                - nIter : number of iteration (defaults to 12)\n",
      "                - throwOnParamFailure : toggles whether or not an exception should be raised if parameters\n",
      "                  for an atom cannot be found.  If this is false (the default), all parameters for unknown\n",
      "                  atoms will be set to zero.  This has the effect of removing that atom from the iteration.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void ComputeGasteigerCharges(RDKit::ROMol const* [,int=12 [,bool=False]])\n",
      "    \n",
      "    ComputeMolShape(mol, confId=-1, boxDim=(20, 20, 20), spacing=0.5, **kwargs)\n",
      "        returns a grid representation of the molecule's shape\n",
      "    \n",
      "    ComputeMolVolume(mol, confId=-1, gridSpacing=0.2, boxMargin=2.0)\n",
      "        Calculates the volume of a particular conformer of a molecule\n",
      "        based on a grid-encoding of the molecular shape.\n",
      "    \n",
      "    ComputeUnionBox(...)\n",
      "        ComputeUnionBox( (tuple)arg1, (tuple)arg2) -> tuple :\n",
      "            Compute the union of two boxes, so that all the points in both boxes are \n",
      "                contained in the new box\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple ComputeUnionBox(boost::python::tuple,boost::python::tuple)\n",
      "    \n",
      "    ConstrainedEmbed(mol, core, useTethers=True, coreConfId=-1, randomseed=2342, getForceField=<Boost.Python.function object at 0x55a87be3a010>, **kwargs)\n",
      "        generates an embedding of a molecule where part of the molecule\n",
      "        is constrained to have particular coordinates\n",
      "        \n",
      "        Arguments\n",
      "          - mol: the molecule to embed\n",
      "          - core: the molecule to use as a source of constraints\n",
      "          - useTethers: (optional) if True, the final conformation will be\n",
      "              optimized subject to a series of extra forces that pull the\n",
      "              matching atoms to the positions of the core atoms. Otherwise\n",
      "              simple distance constraints based on the core atoms will be\n",
      "              used in the optimization.\n",
      "          - coreConfId: (optional) id of the core conformation to use\n",
      "          - randomSeed: (optional) seed for the random number generator\n",
      "        \n",
      "        \n",
      "          An example, start by generating a template with a 3D structure:\n",
      "          >>> from rdkit.Chem import AllChem\n",
      "          >>> template = AllChem.MolFromSmiles(\"c1nn(Cc2ccccc2)cc1\")\n",
      "          >>> AllChem.EmbedMolecule(template)\n",
      "          0\n",
      "          >>> AllChem.UFFOptimizeMolecule(template)\n",
      "          0\n",
      "        \n",
      "          Here's a molecule:\n",
      "          >>> mol = AllChem.MolFromSmiles(\"c1nn(Cc2ccccc2)cc1-c3ccccc3\")\n",
      "        \n",
      "          Now do the constrained embedding\n",
      "          >>> newmol=AllChem.ConstrainedEmbed(mol, template)\n",
      "        \n",
      "          Demonstrate that the positions are the same:\n",
      "          >>> newp=newmol.GetConformer().GetAtomPosition(0)\n",
      "          >>> molp=mol.GetConformer().GetAtomPosition(0)\n",
      "          >>> list(newp-molp)==[0.0,0.0,0.0]\n",
      "          True\n",
      "          >>> newp=newmol.GetConformer().GetAtomPosition(1)\n",
      "          >>> molp=mol.GetConformer().GetAtomPosition(1)\n",
      "          >>> list(newp-molp)==[0.0,0.0,0.0]\n",
      "          True\n",
      "    \n",
      "    CreateDifferenceFingerprintForReaction(...)\n",
      "        CreateDifferenceFingerprintForReaction( (ChemicalReaction)reaction [, (ReactionFingerprintParams)ReactionFingerPrintParams=<rdkit.Chem.rdChemReactions.ReactionFingerprintParams object at 0x7fdd834ca030>]) -> UIntSparseIntVect :\n",
      "            construct a difference fingerprint for a ChemicalReaction by subtracting the reactant fingerprint from the product fingerprint\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<unsigned int>* CreateDifferenceFingerprintForReaction(RDKit::ChemicalReaction [,RDKit::ReactionFingerprintParams=<rdkit.Chem.rdChemReactions.ReactionFingerprintParams object at 0x7fdd834ca030>])\n",
      "    \n",
      "    CreateStructuralFingerprintForReaction(...)\n",
      "        CreateStructuralFingerprintForReaction( (ChemicalReaction)reaction [, (ReactionFingerprintParams)ReactionFingerPrintParams=<rdkit.Chem.rdChemReactions.ReactionFingerprintParams object at 0x7fdd834ca098>]) -> ExplicitBitVect :\n",
      "            construct a structural fingerprint for a ChemicalReaction by concatenating the reactant fingerprint and the product fingerprint\n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* CreateStructuralFingerprintForReaction(RDKit::ChemicalReaction [,RDKit::ReactionFingerprintParams=<rdkit.Chem.rdChemReactions.ReactionFingerprintParams object at 0x7fdd834ca098>])\n",
      "    \n",
      "    DeleteSubstructs(...)\n",
      "        DeleteSubstructs( (Mol)mol, (Mol)query [, (bool)onlyFrags=False [, (bool)useChirality=False]]) -> Mol :\n",
      "            Removes atoms matching a substructure query from a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as a substructure query\n",
      "            \n",
      "                - onlyFrags: (optional) if this toggle is set, atoms will only be removed if\n",
      "                  the entire fragment in which they are found is matched by the query.\n",
      "                  See below for examples.\n",
      "                  Default value is 0 (remove the atoms whether or not the entire fragment matches)\n",
      "            \n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a new molecule with the substructure removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - DeleteSubstructs('CCOC','OC') -> 'CC'\n",
      "            \n",
      "                - DeleteSubstructs('CCOC','OC',1) -> 'CCOC'\n",
      "            \n",
      "                - DeleteSubstructs('CCOCCl.Cl','Cl',1) -> 'CCOCCl'\n",
      "            \n",
      "                - DeleteSubstructs('CCOCCl.Cl','Cl') -> 'CCOC'\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* DeleteSubstructs(RDKit::ROMol,RDKit::ROMol [,bool=False [,bool=False]])\n",
      "    \n",
      "    DetectBondStereoChemistry(...)\n",
      "        DetectBondStereoChemistry( (Mol)mol, (Conformer)conformer) -> None :\n",
      "            Assign stereochemistry to bonds based on coordinates and a conformer.\n",
      "                    DEPRECATED: Please use the version that takes a conformer ID instead.\n",
      "                    \n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "                - conformer: Conformer providing the coordinates\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void DetectBondStereoChemistry(RDKit::ROMol {lvalue},RDKit::Conformer const*)\n",
      "    \n",
      "    DetectBondStereochemistry(...)\n",
      "        DetectBondStereochemistry( (Mol)mol [, (int)confId=-1]) -> None :\n",
      "            Assign stereochemistry to bonds based on coordinates.\n",
      "                    \n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "                - confId: Conformer to use for the coordinates\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void DetectBondStereochemistry(RDKit::ROMol {lvalue} [,int=-1])\n",
      "    \n",
      "    ETDG(...)\n",
      "        ETDG() -> EmbedParameters :\n",
      "            Returns an EmbedParameters object for the ETDG method.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::DGeomHelpers::EmbedParameters* ETDG()\n",
      "    \n",
      "    ETKDG(...)\n",
      "        ETKDG() -> EmbedParameters :\n",
      "            Returns an EmbedParameters object for the ETKDG method.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::DGeomHelpers::EmbedParameters* ETKDG()\n",
      "    \n",
      "    EmbedMolecule(...)\n",
      "        EmbedMolecule( (Mol)mol [, (int)maxAttempts=0 [, (int)randomSeed=-1 [, (bool)clearConfs=True [, (bool)useRandomCoords=False [, (float)boxSizeMult=2.0 [, (bool)randNegEig=True [, (int)numZeroFail=1 [, (dict)coordMap={} [, (float)forceTol=0.001 [, (bool)ignoreSmoothingFailures=False [, (bool)enforceChirality=True [, (bool)useExpTorsionAnglePrefs=False [, (bool)useBasicKnowledge=False [, (bool)printExpTorsionAngles=False]]]]]]]]]]]]]]) -> int :\n",
      "            Use distance geometry to obtain intial \n",
      "             coordinates for a molecule\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - maxAttempts : the maximum number of attempts to try embedding \n",
      "                - randomSeed : provide a seed for the random number generator \n",
      "                               so that the same coordinates can be obtained \n",
      "                               for a molecule on multiple runs. The default \n",
      "                               (-1) uses a random seed \n",
      "                - clearConfs : clear all existing conformations on the molecule\n",
      "                - useRandomCoords : Start the embedding from random coordinates instead of\n",
      "                                    using eigenvalues of the distance matrix.\n",
      "                - boxSizeMult    Determines the size of the box that is used for\n",
      "                                 random coordinates. If this is a positive number, the \n",
      "                                 side length will equal the largest element of the distance\n",
      "                                 matrix times boxSizeMult. If this is a negative number,\n",
      "                                 the side length will equal -boxSizeMult (i.e. independent\n",
      "                                 of the elements of the distance matrix).\n",
      "                - randNegEig : If the embedding yields a negative eigenvalue, \n",
      "                               pick coordinates that correspond \n",
      "                               to this component at random \n",
      "                - numZeroFail : fail embedding if we have at least this many zero eigenvalues \n",
      "                - coordMap : a dictionary mapping atom IDs->coordinates. Use this to \n",
      "                             require some atoms to have fixed coordinates in the resulting \n",
      "                             conformation.\n",
      "                - forceTol : tolerance to be used during the force-field minimization with \n",
      "                             the distance geometry force field.\n",
      "                - ignoreSmoothingFailures : try to embed the molecule even if triangle smoothing\n",
      "                             of the bounds matrix fails.\n",
      "                - enforceChirality : enforce the correct chirality if chiral centers are present.\n",
      "                - useExpTorsionAnglePrefs : impose experimental torsion angle preferences\n",
      "                - useBasicKnowledge : impose basic knowledge such as flat rings\n",
      "                - printExpTorsionAngles : print the output from the experimental torsion angles\n",
      "            \n",
      "             RETURNS:\n",
      "            \n",
      "                ID of the new conformation added to the molecule \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int EmbedMolecule(RDKit::ROMol {lvalue} [,unsigned int=0 [,int=-1 [,bool=True [,bool=False [,double=2.0 [,bool=True [,unsigned int=1 [,boost::python::dict {lvalue}={} [,double=0.001 [,bool=False [,bool=True [,bool=False [,bool=False [,bool=False]]]]]]]]]]]]]])\n",
      "        \n",
      "        EmbedMolecule( (Mol)mol, (EmbedParameters)params) -> int :\n",
      "            Use distance geometry to obtain intial \n",
      "             coordinates for a molecule\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - params : an EmbedParameters object \n",
      "            \n",
      "             RETURNS:\n",
      "            \n",
      "                ID of the new conformation added to the molecule \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int EmbedMolecule(RDKit::ROMol {lvalue},RDKit::DGeomHelpers::EmbedParameters {lvalue})\n",
      "    \n",
      "    EmbedMultipleConfs(...)\n",
      "        EmbedMultipleConfs( (Mol)mol [, (int)numConfs=10 [, (int)maxAttempts=0 [, (int)randomSeed=-1 [, (bool)clearConfs=True [, (bool)useRandomCoords=False [, (float)boxSizeMult=2.0 [, (bool)randNegEig=True [, (int)numZeroFail=1 [, (float)pruneRmsThresh=-1.0 [, (dict)coordMap={} [, (float)forceTol=0.001 [, (bool)ignoreSmoothingFailures=False [, (bool)enforceChirality=True [, (int)numThreads=1 [, (bool)useExpTorsionAnglePrefs=False [, (bool)useBasicKnowledge=False [, (bool)printExpTorsionAngles=False]]]]]]]]]]]]]]]]]) -> _vecti :\n",
      "            Use distance geometry to obtain multiple sets of \n",
      "             coordinates for a molecule\n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "              - mol : the molecule of interest\n",
      "              - numConfs : the number of conformers to generate \n",
      "              - maxAttempts : the maximum number of attempts to try embedding \n",
      "              - randomSeed : provide a seed for the random number generator \n",
      "                             so that the same coordinates can be obtained \n",
      "                             for a molecule on multiple runs. The default \n",
      "                             (-1) uses a random seed \n",
      "              - clearConfs : clear all existing conformations on the molecule\n",
      "              - useRandomCoords : Start the embedding from random coordinates instead of\n",
      "                                  using eigenvalues of the distance matrix.\n",
      "              - boxSizeMult    Determines the size of the box that is used for\n",
      "                               random coordinates. If this is a positive number, the \n",
      "                               side length will equal the largest element of the distance\n",
      "                               matrix times boxSizeMult. If this is a negative number,\n",
      "                               the side length will equal -boxSizeMult (i.e. independent\n",
      "                               of the elements of the distance matrix).\n",
      "              - randNegEig : If the embedding yields a negative eigenvalue, \n",
      "                             pick coordinates that correspond \n",
      "                             to this component at random \n",
      "              - numZeroFail : fail embedding if we have at least this many zero eigenvalues \n",
      "              - pruneRmsThresh : Retain only the conformations out of 'numConfs' \n",
      "                                after embedding that are at least \n",
      "                                this far apart from each other. \n",
      "                      RMSD is computed on the heavy atoms. \n",
      "                      Pruning is greedy; i.e. the first embedded conformation\n",
      "                      is retained and from then on only those that are at\n",
      "                      least pruneRmsThresh away from all retained conformations\n",
      "                      are kept. The pruning is done after embedding and \n",
      "                      bounds violation minimization. No pruning by default.\n",
      "                - coordMap : a dictionary mapping atom IDs->coordinates. Use this to \n",
      "                             require some atoms to have fixed coordinates in the resulting \n",
      "                             conformation.\n",
      "                - forceTol : tolerance to be used during the force-field minimization with \n",
      "                             the distance geometry force field.\n",
      "                - ignoreSmoothingFailures : try to embed the molecule even if triangle smoothing\n",
      "                             of the bounds matrix fails.\n",
      "                - enforceChirality : enforce the correct chirality if chiral centers are present.\n",
      "                - numThreads : number of threads to use while embedding. This only has an effect if the RDKit\n",
      "                             was built with multi-thread support.\n",
      "                            If set to zero, the max supported by the system will be used.\n",
      "                - useExpTorsionAnglePrefs : impose experimental torsion angle preferences\n",
      "                - useBasicKnowledge : impose basic knowledge such as flat rings\n",
      "                - printExpTorsionAngles : print the output from the experimental torsion angles\n",
      "             RETURNS:\n",
      "            \n",
      "                List of new conformation IDs \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<int, std::allocator<int> > EmbedMultipleConfs(RDKit::ROMol {lvalue} [,unsigned int=10 [,unsigned int=0 [,int=-1 [,bool=True [,bool=False [,double=2.0 [,bool=True [,unsigned int=1 [,double=-1.0 [,boost::python::dict {lvalue}={} [,double=0.001 [,bool=False [,bool=True [,int=1 [,bool=False [,bool=False [,bool=False]]]]]]]]]]]]]]]]])\n",
      "        \n",
      "        EmbedMultipleConfs( (Mol)mol, (int)numConfs, (EmbedParameters)params) -> _vecti :\n",
      "            Use distance geometry to obtain multiple sets of \n",
      "             coordinates for a molecule\n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "              - mol : the molecule of interest\n",
      "              - numConfs : the number of conformers to generate \n",
      "              - params : an EmbedParameters object \n",
      "             RETURNS:\n",
      "            \n",
      "                List of new conformation IDs \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<int, std::allocator<int> > EmbedMultipleConfs(RDKit::ROMol {lvalue},unsigned int,RDKit::DGeomHelpers::EmbedParameters {lvalue})\n",
      "    \n",
      "    EncodeShape(...)\n",
      "        EncodeShape( (Mol)mol, (UniformGrid3D_)grid [, (int)confId=-1 [, (AtomPairsParameters)trans=None [, (float)vdwScale=0.8 [, (float)stepSize=0.25 [, (int)maxLayers=-1 [, (bool)ignoreHs=True]]]]]]) -> None :\n",
      "            Encode the shape of a molecule (one of its conformer) onto a grid\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - grid : grid onto which the encoding is written \n",
      "                - confId : id of the conformation of interest on mol (defaults to the first one) \n",
      "                - trans : any transformation that needs to be used to encode onto the grid (note the molecule remains unchanged) \n",
      "                - vdwScale : Scaling factor for the radius of the atoms to determine the base radius \n",
      "                             used in the encoding - grid points inside this sphere carry the maximum occupancy \n",
      "                - setpSize : thickness of the layers outside the base radius, the occupancy value is decreased \n",
      "                             from layer to layer from the maximum value \n",
      "                - maxLayers : the maximum number of layers - defaults to the number of bits \n",
      "                              used per grid point - e.g. two bits per grid point will allow 3 layers\n",
      "                - ignoreHs : when set, the contribution of Hs to the shape will be ignored\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void EncodeShape(RDKit::ROMol,RDGeom::UniformGrid3D {lvalue} [,int=-1 [,boost::python::api::object=None [,double=0.8 [,double=0.25 [,int=-1 [,bool=True]]]]]])\n",
      "    \n",
      "    EnumerateLibraryCanSerialize(...)\n",
      "        EnumerateLibraryCanSerialize() -> bool :\n",
      "            Returns True if the EnumerateLibrary is serializable (requires boost serialization\n",
      "        \n",
      "            C++ signature :\n",
      "                bool EnumerateLibraryCanSerialize()\n",
      "    \n",
      "    EnumerateLibraryFromReaction(reaction, sidechainSets, returnReactants=False)\n",
      "        Returns a generator for the virtual library defined by\n",
      "         a reaction and a sequence of sidechain sets\n",
      "        \n",
      "        >>> from rdkit import Chem\n",
      "        >>> from rdkit.Chem import AllChem\n",
      "        >>> s1=[Chem.MolFromSmiles(x) for x in ('NC','NCC')]\n",
      "        >>> s2=[Chem.MolFromSmiles(x) for x in ('OC=O','OC(=O)C')]\n",
      "        >>> rxn = AllChem.ReactionFromSmarts('[O:2]=[C:1][OH].[N:3]>>[O:2]=[C:1][N:3]')\n",
      "        >>> r = AllChem.EnumerateLibraryFromReaction(rxn,[s2,s1])\n",
      "        >>> [Chem.MolToSmiles(x[0]) for x in list(r)]\n",
      "        ['CNC=O', 'CCNC=O', 'CNC(C)=O', 'CCNC(C)=O']\n",
      "        \n",
      "        Note that this is all done in a lazy manner, so \"infinitely\" large libraries can\n",
      "        be done without worrying about running out of memory. Your patience will run out first:\n",
      "        \n",
      "        Define a set of 10000 amines:\n",
      "        >>> amines = (Chem.MolFromSmiles('N'+'C'*x) for x in range(10000))\n",
      "        \n",
      "        ... a set of 10000 acids\n",
      "        >>> acids = (Chem.MolFromSmiles('OC(=O)'+'C'*x) for x in range(10000))\n",
      "        \n",
      "        ... now the virtual library (1e8 compounds in principle):\n",
      "        >>> r = AllChem.EnumerateLibraryFromReaction(rxn,[acids,amines])\n",
      "        \n",
      "        ... look at the first 4 compounds:\n",
      "        >>> [Chem.MolToSmiles(next(r)[0]) for x in range(4)]\n",
      "        ['NC=O', 'CNC=O', 'CCNC=O', 'CCCNC=O']\n",
      "    \n",
      "    ExplicitDegreeEqualsQueryAtom(...)\n",
      "        ExplicitDegreeEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitDegree is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitDegreeEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    ExplicitDegreeGreaterQueryAtom(...)\n",
      "        ExplicitDegreeGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitDegree is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitDegreeGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    ExplicitDegreeLessQueryAtom(...)\n",
      "        ExplicitDegreeLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitDegree is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitDegreeLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    ExplicitValenceEqualsQueryAtom(...)\n",
      "        ExplicitValenceEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitValence is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitValenceEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    ExplicitValenceGreaterQueryAtom(...)\n",
      "        ExplicitValenceGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitValence is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitValenceGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    ExplicitValenceLessQueryAtom(...)\n",
      "        ExplicitValenceLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where ExplicitValence is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* ExplicitValenceLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    FastFindRings(...)\n",
      "        FastFindRings( (Mol)arg1) -> None :\n",
      "            Does a non-SSSR ring finding for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: Nothing\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void FastFindRings(RDKit::ROMol)\n",
      "    \n",
      "    FindAllPathsOfLengthN(...)\n",
      "        FindAllPathsOfLengthN( (Mol)mol, (int)length [, (bool)useBonds=True [, (bool)useHs=False [, (int)rootedAtAtom=-1]]]) -> _listSt6vectorIiSaIiEE :\n",
      "            Finds all paths of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target length for the paths.\n",
      "            \n",
      "                - useBonds: (optional) toggles the use of bond indices in the paths.\n",
      "                  Otherwise atom indices are used.  *Note* this behavior is different\n",
      "                  from that for subgraphs.\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only paths from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of tuples with IDs for the bonds.\n",
      "            \n",
      "              NOTES: \n",
      "            \n",
      "               - Difference between _subgraphs_ and _paths_ :: \n",
      "            \n",
      "                   Subgraphs are potentially branched, whereas paths (in our \n",
      "                   terminology at least) cannot be.  So, the following graph: \n",
      "            \n",
      "                        C--0--C--1--C--3--C\n",
      "                              |\n",
      "                              2\n",
      "                              |\n",
      "                              C\n",
      "            \n",
      "                   has 3 _subgraphs_ of length 3: (0,1,2),(0,1,3),(2,1,3)\n",
      "                   but only 2 _paths_ of length 3: (0,1,3),(2,1,3)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::list<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > FindAllPathsOfLengthN(RDKit::ROMol,unsigned int [,bool=True [,bool=False [,int=-1]]])\n",
      "    \n",
      "    FindAllSubgraphsOfLengthMToN(...)\n",
      "        FindAllSubgraphsOfLengthMToN( (Mol)mol, (int)min, (int)max [, (bool)useHs=False [, (int)rootedAtAtom=-1]]) -> object :\n",
      "            Finds all subgraphs of a particular length in a molecule\n",
      "              See documentation for FindAllSubgraphsOfLengthN for definitions\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object FindAllSubgraphsOfLengthMToN(RDKit::ROMol,unsigned int,unsigned int [,bool=False [,int=-1]])\n",
      "    \n",
      "    FindAllSubgraphsOfLengthN(...)\n",
      "        FindAllSubgraphsOfLengthN( (Mol)mol, (int)length [, (bool)useHs=False [, (int)rootedAtAtom=-1]]) -> _listSt6vectorIiSaIiEE :\n",
      "            Finds all subgraphs of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target number of bonds for the subgraphs.\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only subgraphs from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of 2-tuples with bond IDs\n",
      "            \n",
      "              NOTES: \n",
      "            \n",
      "               - Difference between _subgraphs_ and _paths_ :: \n",
      "            \n",
      "                   Subgraphs are potentially branched, whereas paths (in our \n",
      "                   terminology at least) cannot be.  So, the following graph: \n",
      "            \n",
      "                        C--0--C--1--C--3--C\n",
      "                              |\n",
      "                              2\n",
      "                              |\n",
      "                              C\n",
      "              has 3 _subgraphs_ of length 3: (0,1,2),(0,1,3),(2,1,3)\n",
      "              but only 2 _paths_ of length 3: (0,1,3),(2,1,3)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::list<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > FindAllSubgraphsOfLengthN(RDKit::ROMol,unsigned int [,bool=False [,int=-1]])\n",
      "    \n",
      "    FindAtomEnvironmentOfRadiusN(...)\n",
      "        FindAtomEnvironmentOfRadiusN( (Mol)mol, (int)radius, (int)rootedAtAtom [, (bool)useHs=False]) -> _vecti :\n",
      "            Finds the bonds within a certain radius of an atom in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - radius: an integer with the target radius for the environment.\n",
      "            \n",
      "                - rootedAtAtom: the atom to consider\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "              RETURNS: a vector of bond IDs\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<int, std::allocator<int> > FindAtomEnvironmentOfRadiusN(RDKit::ROMol,unsigned int,unsigned int [,bool=False])\n",
      "    \n",
      "    FindPotentialStereoBonds(...)\n",
      "        FindPotentialStereoBonds( (Mol)mol [, (bool)cleanIt=False]) -> None :\n",
      "            Find bonds than can be cis/trans in a molecule and mark them as 'any'.\n",
      "                     This function finds any double bonds that can potentially be part\n",
      "                     of a cis/trans system. No attempt is made here to mark them cis or trans\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - cleanIt: (optional) if this option is set to true, any previous marking of _CIPCode\n",
      "                           on the bond is cleared - otherwise it is left untouched\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void FindPotentialStereoBonds(RDKit::ROMol {lvalue} [,bool=False])\n",
      "    \n",
      "    FindUniqueSubgraphsOfLengthN(...)\n",
      "        FindUniqueSubgraphsOfLengthN( (Mol)mol, (int)length [, (bool)useHs=False [, (bool)useBO=True [, (int)rootedAtAtom=-1]]]) -> _listSt6vectorIiSaIiEE :\n",
      "            Finds unique subgraphs of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target number of bonds for the subgraphs.\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - useBO: (optional) Toggles use of bond orders in distinguishing one subgraph from\n",
      "                  another.\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only subgraphs from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of tuples with bond IDs\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::list<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > FindUniqueSubgraphsOfLengthN(RDKit::ROMol,unsigned int [,bool=False [,bool=True [,int=-1]]])\n",
      "    \n",
      "    FormalChargeEqualsQueryAtom(...)\n",
      "        FormalChargeEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where FormalCharge is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* FormalChargeEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    FormalChargeGreaterQueryAtom(...)\n",
      "        FormalChargeGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where FormalCharge is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* FormalChargeGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    FormalChargeLessQueryAtom(...)\n",
      "        FormalChargeLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where FormalCharge is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* FormalChargeLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    FragmentOnBRICSBonds(...)\n",
      "        FragmentOnBRICSBonds( (Mol)mol) -> Mol :\n",
      "            Return a new molecule with all BRICS bonds broken\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* FragmentOnBRICSBonds(RDKit::ROMol)\n",
      "    \n",
      "    FragmentOnBonds(...)\n",
      "        FragmentOnBonds( (Mol)mol, (AtomPairsParameters)bondIndices [, (bool)addDummies=True [, (AtomPairsParameters)dummyLabels=None [, (AtomPairsParameters)bondTypes=None [, (list)cutsPerAtom=[]]]]]) -> Mol :\n",
      "            Return a new molecule with all specified bonds broken\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                  - mol            - the molecule to be modified\n",
      "                  - bondIndices    - indices of the bonds to be broken\n",
      "                  - addDummies  - toggles addition of dummy atoms to indicate where \n",
      "                      bonds were broken\n",
      "                  - dummyLabels - used to provide the labels to be used for the dummies.\n",
      "                      the first element in each pair is the label for the dummy\n",
      "                      that replaces the bond's beginAtom, the second is for the \n",
      "                      dummy that replaces the bond's endAtom. If not provided, the\n",
      "                      dummies are labeled with atom indices.\n",
      "                  - bondTypes - used to provide the bond type to use between the\n",
      "                      fragments and the dummy atoms. If not provided, defaults to single. \n",
      "                  - cutsPerAtom - used to return the number of cuts made at each atom. \n",
      "            \n",
      "              RETURNS:\n",
      "                  a new Mol with the modifications\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* FragmentOnBonds(RDKit::ROMol,boost::python::api::object [,bool=True [,boost::python::api::object=None [,boost::python::api::object=None [,boost::python::list=[]]]]])\n",
      "    \n",
      "    FragmentOnSomeBonds(...)\n",
      "        FragmentOnSomeBonds( (Mol)mol, (AtomPairsParameters)bondIndices [, (int)numToBreak=1 [, (bool)addDummies=True [, (AtomPairsParameters)dummyLabels=None [, (AtomPairsParameters)bondTypes=None [, (bool)returnCutsPerAtom=False]]]]]) -> tuple :\n",
      "            fragment on some bonds\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple FragmentOnSomeBonds(RDKit::ROMol,boost::python::api::object [,unsigned int=1 [,bool=True [,boost::python::api::object=None [,boost::python::api::object=None [,bool=False]]]]])\n",
      "    \n",
      "    GenerateDepictionMatching2DStructure(...)\n",
      "        GenerateDepictionMatching2DStructure( (Mol)mol, (Mol)reference [, (int)confId=-1 [, (AtomPairsParameters)refPatt=None [, (bool)acceptFailure=False]]]) -> None :\n",
      "            Generate a depiction for a molecule where a piece of the \n",
      "              molecule is constrained to have the same coordinates as a reference. \n",
      "            \n",
      "              This is useful for, for example, generating depictions of SAR data \n",
      "              sets so that the cores of the molecules are all oriented the same way. \n",
      "              ARGUMENTS: \n",
      "            \n",
      "              mol -    the molecule to be aligned, this will come back \n",
      "                       with a single conformer. \n",
      "              reference -    a molecule with the reference atoms to align to; \n",
      "                             this should have a depiction. \n",
      "              confId -       (optional) the id of the reference conformation to use \n",
      "              referencePattern -  (optional) a query molecule to be used to \n",
      "                                  generate the atom mapping between the molecule \n",
      "                                  and the reference. \n",
      "              acceptFailure - (optional) if True, standard depictions will be generated \n",
      "                              for molecules that don't have a substructure match to the \n",
      "                              reference; if False, throws a DepictException.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void GenerateDepictionMatching2DStructure(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,int=-1 [,boost::python::api::object=None [,bool=False]]])\n",
      "    \n",
      "    GenerateDepictionMatching3DStructure(...)\n",
      "        GenerateDepictionMatching3DStructure( (Mol)mol, (Mol)reference [, (int)confId=-1 [, (AtomPairsParameters)refPatt=None [, (bool)acceptFailure=False]]]) -> None :\n",
      "            Generate a depiction for a molecule where a piece of the molecule \n",
      "              is constrained to have coordinates similar to those of a 3D reference \n",
      "              structure.\n",
      "              ARGUMENTS: \n",
      "            \n",
      "              mol -    the molecule to be aligned, this will come back \n",
      "                       with a single conformer containing the 2D coordinates. \n",
      "              reference -    a molecule with the reference atoms to align to. \n",
      "                             By default this should be the same as mol, but with \n",
      "                             3D coordinates \n",
      "              confId -       (optional) the id of the reference conformation to use \n",
      "              referencePattern -  (optional) a query molecule to map a subset of \n",
      "                                  the reference onto the mol, so that only some of the \n",
      "                                  atoms are aligned. \n",
      "              acceptFailure - (optional) if True, standard depictions will be generated \n",
      "                              for molecules that don't match the reference or the\n",
      "                              referencePattern; if False, throws a DepictException.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void GenerateDepictionMatching3DStructure(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,int=-1 [,boost::python::api::object=None [,bool=False]]])\n",
      "    \n",
      "    GenerateErGFingerprintForReducedGraph(...)\n",
      "        GenerateErGFingerprintForReducedGraph( (Mol)mol [, (AtomPairsParameters)atomTypes=0 [, (float)fuzzIncrement=0.3 [, (int)minPath=1 [, (int)maxPath=15]]]]) -> object :\n",
      "            Returns the ErG fingerprint vector for a reduced graph\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GenerateErGFingerprintForReducedGraph(RDKit::ROMol [,boost::python::api::object=0 [,double=0.3 [,int=1 [,int=15]]]])\n",
      "    \n",
      "    GenerateMolExtendedReducedGraph(...)\n",
      "        GenerateMolExtendedReducedGraph( (Mol)mol [, (AtomPairsParameters)atomTypes=0]) -> Mol :\n",
      "            Returns the reduced graph for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* GenerateMolExtendedReducedGraph(RDKit::ROMol [,boost::python::api::object=0])\n",
      "    \n",
      "    Get3DDistanceMatrix(...)\n",
      "        Get3DDistanceMatrix( (Mol)mol [, (int)confId=-1 [, (bool)useAtomWts=False [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's 3D distance matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - confId: (optional) chooses the conformer Id to use\n",
      "                  Default value is -1.\n",
      "            \n",
      "                - useAtomWts: (optional) toggles using atom weights for the diagonal elements of the\n",
      "                  matrix (to return a \"Balaban\" distance matrix).\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats with the distance matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* Get3DDistanceMatrix(RDKit::ROMol {lvalue} [,int=-1 [,bool=False [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetAdjacencyMatrix(...)\n",
      "        GetAdjacencyMatrix( (Mol)mol [, (bool)useBO=False [, (int)emptyVal=0 [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's adjacency matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - useBO: (optional) toggles use of bond orders in calculating the matrix.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - emptyVal: (optional) sets the elements of the matrix between non-adjacent atoms\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats containing the adjacency matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetAdjacencyMatrix(RDKit::ROMol {lvalue} [,bool=False [,int=0 [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetAlignmentTransform(...)\n",
      "        GetAlignmentTransform( (Mol)prbMol, (Mol)refMol [, (int)prbCid=-1 [, (int)refCid=-1 [, (AtomPairsParameters)atomMap=[] [, (AtomPairsParameters)weights=[] [, (bool)reflect=False [, (int)maxIters=50]]]]]]) -> object :\n",
      "            Compute the transformation required to align a molecule\n",
      "                 \n",
      "                  The 3D transformation required to align the specied conformation in the probe molecule\n",
      "                  to a specified conformation in the reference molecule is computed so that the root mean\n",
      "                  squared distance between a specified set of atoms is minimized\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol    molecule that is to be aligned\n",
      "                  - refMol    molecule used as the reference for the alignment\n",
      "                  - prbCid    ID of the conformation in the probe to be used \n",
      "                                   for the alignment (defaults to first conformation)\n",
      "                  - refCid    ID of the conformation in the ref molecule to which \n",
      "                                   the alignment is computed (defaults to first conformation)\n",
      "                  - atomMap   a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                   used to compute the alignments. If this mapping is \n",
      "                                   not specified an attempt is made to generate on by\n",
      "                                   substructure matching\n",
      "                  - weights   Optionally specify weights for each of the atom pairs\n",
      "                  - reflect   if true reflect the conformation of the probe molecule\n",
      "                  - maxIters  maximum number of iterations used in mimizing the RMSD\n",
      "                   \n",
      "                  RETURNS\n",
      "                  a tuple of (RMSD value, transform matrix) \n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetAlignmentTransform(RDKit::ROMol,RDKit::ROMol [,int=-1 [,int=-1 [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,unsigned int=50]]]]]])\n",
      "    \n",
      "    GetAngleDeg(...)\n",
      "        GetAngleDeg( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId) -> float :\n",
      "            Returns the angle in degrees between atoms i, j, k\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetAngleDeg(RDKit::Conformer,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetAngleRad(...)\n",
      "        GetAngleRad( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId) -> float :\n",
      "            Returns the angle in radians between atoms i, j, k\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetAngleRad(RDKit::Conformer,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetAtomAlias(...)\n",
      "        GetAtomAlias( (Atom)atom) -> str :\n",
      "            Returns the atom's MDL alias text\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string GetAtomAlias(RDKit::Atom const*)\n",
      "    \n",
      "    GetAtomMatch(...)\n",
      "        GetAtomMatch( (AtomPairsParameters)featMatch [, (int)maxAts=1024]) -> object :\n",
      "            Returns an empty list if any of the features passed in share an atom.\n",
      "             Otherwise a list of lists of atom indices is returned.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object GetAtomMatch(boost::python::api::object [,int=1024])\n",
      "    \n",
      "    GetAtomPairAtomCode(...)\n",
      "        GetAtomPairAtomCode( (Atom)atom [, (int)branchSubtract=0 [, (bool)includeChirality=False]]) -> int :\n",
      "            Returns the atom code (hash) for an atom\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int GetAtomPairAtomCode(RDKit::Atom const* [,unsigned int=0 [,bool=False]])\n",
      "    \n",
      "    GetAtomPairCode(...)\n",
      "        GetAtomPairCode( (int)atom1Code, (int)atom2Code, (int)distance [, (bool)includeChirality=False]) -> int :\n",
      "            Returns the atom-pair code (hash) for a pair of atoms separated by a certain number of bonds\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int GetAtomPairCode(unsigned int,unsigned int,unsigned int [,bool=False])\n",
      "    \n",
      "    GetAtomPairFingerprint(...)\n",
      "        GetAtomPairFingerprint( (Mol)mol [, (int)minLength=1 [, (int)maxLength=30 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (bool)includeChirality=False [, (bool)use2D=True [, (int)confId=-1]]]]]]]]) -> IntSparseIntVect :\n",
      "            Returns the atom-pair fingerprint for a molecule as an IntSparseIntVect\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<int>* GetAtomPairFingerprint(RDKit::ROMol [,unsigned int=1 [,unsigned int=30 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=False [,bool=True [,int=-1]]]]]]]])\n",
      "    \n",
      "    GetAtomRLabel(...)\n",
      "        GetAtomRLabel( (Atom)atom) -> int :\n",
      "            Returns the atom's MDL AtomRLabel (this is an integer from 0 to 99)\n",
      "        \n",
      "            C++ signature :\n",
      "                int GetAtomRLabel(RDKit::Atom const*)\n",
      "    \n",
      "    GetAtomValue(...)\n",
      "        GetAtomValue( (Atom)atom) -> str :\n",
      "            Returns the atom's MDL alias text\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string GetAtomValue(RDKit::Atom const*)\n",
      "    \n",
      "    GetBestRMS(...)\n",
      "        GetBestRMS( (Mol)prbMol, (Mol)refMol [, (int)prbId=-1 [, (int)refId=-1 [, (AtomPairsParameters)map=None [, (int)maxMatches=1000000]]]]) -> float :\n",
      "            Returns the optimal RMS for aligning two molecules, taking\n",
      "                   symmetry into account. As a side-effect, the probe molecule is\n",
      "                   left in the aligned state.\n",
      "                  \n",
      "                   Note:\n",
      "                   This function will attempt to align all permutations of matching atom\n",
      "                   orders in both molecules, for some molecules it will lead to\n",
      "                   'combinatorial explosion' especially if hydrogens are present.\n",
      "                   Use 'rdkit.Chem.AllChem.AlignMol' to align molecules without changing\n",
      "                   the atom order.\n",
      "                  \n",
      "                   ARGUMENTS\n",
      "                    - prbMol:      the molecule to be aligned to the reference\n",
      "                    - refMol:      the reference molecule\n",
      "                    - prbId:       (optional) probe conformation to use\n",
      "                    - refId:       (optional) reference conformation to use\n",
      "                    - map:         (optional) a list of lists of (probeAtomId,refAtomId)\n",
      "                                   tuples with the atom-atom mappings of the two\n",
      "                                   molecules. If not provided, these will be generated\n",
      "                                   using a substructure search.\n",
      "                    - maxMatches:  (optional) if map isn't specified, this will be\n",
      "                                   the max number of matches found in a SubstructMatch()\n",
      "                   \n",
      "                  RETURNS\n",
      "                  The best RMSD found\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetBestRMS(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,int=-1 [,int=-1 [,boost::python::api::object=None [,int=1000000]]]])\n",
      "    \n",
      "    GetBondLength(...)\n",
      "        GetBondLength( (Conformer)conf, (int)iAtomId, (int)jAtomId) -> float :\n",
      "            Returns the bond length in angstrom between atoms i, j\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetBondLength(RDKit::Conformer,unsigned int,unsigned int)\n",
      "    \n",
      "    GetChemDrawRxnAdjustParams(...)\n",
      "        GetChemDrawRxnAdjustParams() -> AdjustQueryParameters :\n",
      "            (deprecated, see MatchOnlyAtRgroupsAdjustParams)\n",
      "                Returns the chemdraw style adjustment parameters for reactant templates\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolOps::AdjustQueryParameters GetChemDrawRxnAdjustParams()\n",
      "    \n",
      "    GetConformerRMS(mol, confId1, confId2, atomIds=None, prealigned=False)\n",
      "        Returns the RMS between two conformations.\n",
      "        By default, the conformers will be aligned to the first conformer\n",
      "        before the RMS calculation and, as a side-effect, the second will be left\n",
      "        in the aligned state.\n",
      "        \n",
      "        Arguments:\n",
      "          - mol:        the molecule\n",
      "          - confId1:    the id of the first conformer\n",
      "          - confId2:    the id of the second conformer\n",
      "          - atomIds:    (optional) list of atom ids to use a points for\n",
      "                        alingment - defaults to all atoms\n",
      "          - prealigned: (optional) by default the conformers are assumed\n",
      "                        be unaligned and the second conformer be aligned\n",
      "                        to the first\n",
      "    \n",
      "    GetConformerRMSMatrix(mol, atomIds=None, prealigned=False)\n",
      "        Returns the RMS matrix of the conformers of a molecule.\n",
      "        As a side-effect, the conformers will be aligned to the first\n",
      "        conformer (i.e. the reference) and will left in the aligned state.\n",
      "        \n",
      "        Arguments:\n",
      "          - mol:     the molecule\n",
      "          - atomIds: (optional) list of atom ids to use a points for\n",
      "                     alingment - defaults to all atoms\n",
      "          - prealigned: (optional) by default the conformers are assumed\n",
      "                        be unaligned and will therefore be aligned to the\n",
      "                        first conformer\n",
      "        \n",
      "        Note that the returned RMS matrix is symmetrical, i.e. it is the\n",
      "        lower half of the matrix, e.g. for 5 conformers:\n",
      "        rmsmatrix = [ a,\n",
      "                      b, c,\n",
      "                      d, e, f,\n",
      "                      g, h, i, j]\n",
      "        where a is the RMS between conformers 0 and 1, b is the RMS between\n",
      "        conformers 0 and 2, etc.\n",
      "        This way it can be directly used as distance matrix in e.g. Butina\n",
      "        clustering.\n",
      "    \n",
      "    GetConnectivityInvariants(...)\n",
      "        GetConnectivityInvariants( (Mol)mol [, (bool)includeRingMembership=True]) -> list :\n",
      "            Returns connectivity invariants (ECFP-like) for a molecule.\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetConnectivityInvariants(RDKit::ROMol [,bool=True])\n",
      "    \n",
      "    GetCrippenO3A(...)\n",
      "        GetCrippenO3A( (Mol)prbMol, (Mol)refMol [, (list)prbCrippenContribs=[] [, (list)refCrippenContribs=[] [, (int)prbCid=-1 [, (int)refCid=-1 [, (bool)reflect=False [, (int)maxIters=50 [, (int)options=0 [, (list)constraintMap=[] [, (list)constraintWeights=[]]]]]]]]]]) -> O3A :\n",
      "            Get an O3A object with atomMap and weights vectors to overlay\n",
      "                  the probe molecule onto the reference molecule based on\n",
      "                  Crippen logP atom contributions\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol                   molecule that is to be aligned\n",
      "                  - refMol                   molecule used as the reference for the alignment\n",
      "                  - prbCrippenContribs       Crippen atom contributions for the probe molecule\n",
      "                                             as a list of (logp, mr) tuples, as returned\n",
      "                                             by _CalcCrippenContribs()\n",
      "                  - refCrippenContribs       Crippen atom contributions for the reference molecule\n",
      "                                             as a list of (logp, mr) tuples, as returned\n",
      "                                             by _CalcCrippenContribs()\n",
      "                  - prbCid                   ID of the conformation in the probe to be used \n",
      "                                             for the alignment (defaults to first conformation)\n",
      "                  - refCid                   ID of the conformation in the ref molecule to which \n",
      "                                             the alignment is computed (defaults to first conformation)\n",
      "                  - reflect                  if true reflect the conformation of the probe molecule\n",
      "                                             (defaults to false)\n",
      "                  - maxIters                 maximum number of iterations used in mimizing the RMSD\n",
      "                                             (defaults to 50)\n",
      "                  - options                  least 2 significant bits encode accuracy\n",
      "                                             (0: maximum, 3: minimum; defaults to 0)\n",
      "                                             bit 3 triggers local optimization of the alignment\n",
      "                                             (no computation of the cost matrix; defaults: off)\n",
      "                  - constraintMap            a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                             which shall be used for the alignment (defaults to [])\n",
      "                  - constraintWeights        optionally specify weights for each of the constraints\n",
      "                                             (weights default to 100.0)\n",
      "                   \n",
      "                  RETURNS\n",
      "                  The O3A object\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolAlign::PyO3A* GetCrippenO3A(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,boost::python::list=[] [,boost::python::list=[] [,int=-1 [,int=-1 [,bool=False [,unsigned int=50 [,unsigned int=0 [,boost::python::list=[] [,boost::python::list=[]]]]]]]]]])\n",
      "    \n",
      "    GetCrippenO3AForProbeConfs(...)\n",
      "        GetCrippenO3AForProbeConfs( (Mol)prbMol, (Mol)refMol [, (int)numThreads=1 [, (list)prbCrippenContribs=[] [, (list)refCrippenContribs=[] [, (int)refCid=-1 [, (bool)reflect=False [, (int)maxIters=50 [, (int)options=0 [, (list)constraintMap=[] [, (list)constraintWeights=[]]]]]]]]]]) -> tuple :\n",
      "            Get a vector of O3A objects for the overlay of all \n",
      "                  the probe molecule's conformations onto the reference molecule based on\n",
      "                  MMFF atom types and charges\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol                   molecule that is to be aligned\n",
      "                  - refMol                   molecule used as the reference for the alignment\n",
      "                  - numThreads :             the number of threads to use, only has an effect if\n",
      "                                             the RDKit was built with thread support (defaults to 1)\n",
      "                  - prbCrippenContribs       Crippen atom contributions for the probe molecule\n",
      "                                             as a list of (logp, mr) tuples, as returned\n",
      "                                             by _CalcCrippenContribs()\n",
      "                  - refCrippenContribs       Crippen atom contributions for the reference molecule\n",
      "                                             as a list of (logp, mr) tuples, as returned\n",
      "                                             by _CalcCrippenContribs()\n",
      "                  - refCid                   ID of the conformation in the ref molecule to which \n",
      "                                             the alignment is computed (defaults to first conformation)\n",
      "                  - reflect                  if true reflect the conformation of the probe molecule\n",
      "                                             (defaults to false)\n",
      "                  - maxIters                 maximum number of iterations used in mimizing the RMSD\n",
      "                                             (defaults to 50)\n",
      "                  - options                  least 2 significant bits encode accuracy\n",
      "                                             (0: maximum, 3: minimum; defaults to 0)\n",
      "                                             bit 3 triggers local optimization of the alignment\n",
      "                                             (no computation of the cost matrix; defaults: off)\n",
      "                  - constraintMap            a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                             which shall be used for the alignment (defaults to [])\n",
      "                  - constraintWeights        optionally specify weights for each of the constraints\n",
      "                                             (weights default to 100.0)\n",
      "                   \n",
      "                  RETURNS\n",
      "                  A vector of O3A objects\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetCrippenO3AForProbeConfs(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,int=1 [,boost::python::list=[] [,boost::python::list=[] [,int=-1 [,bool=False [,unsigned int=50 [,unsigned int=0 [,boost::python::list=[] [,boost::python::list=[]]]]]]]]]])\n",
      "    \n",
      "    GetDefaultAdjustParams(...)\n",
      "        GetDefaultAdjustParams() -> AdjustQueryParameters :\n",
      "            Returns the default adjustment parameters for reactant templates\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolOps::AdjustQueryParameters GetDefaultAdjustParams()\n",
      "    \n",
      "    GetDefaultPickleProperties(...)\n",
      "        GetDefaultPickleProperties() -> int :\n",
      "            Get the current global mol pickler options.\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int GetDefaultPickleProperties()\n",
      "    \n",
      "    GetDihedralDeg(...)\n",
      "        GetDihedralDeg( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId, (int)lAtomId) -> float :\n",
      "            Returns the dihedral angle in degrees between atoms i, j, k, l\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetDihedralDeg(RDKit::Conformer,unsigned int,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetDihedralRad(...)\n",
      "        GetDihedralRad( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId, (int)lAtomId) -> float :\n",
      "            Returns the dihedral angle in radians between atoms i, j, k, l\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double GetDihedralRad(RDKit::Conformer,unsigned int,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetDistanceMatrix(...)\n",
      "        GetDistanceMatrix( (Mol)mol [, (bool)useBO=False [, (bool)useAtomWts=False [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's topological distance matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - useBO: (optional) toggles use of bond orders in calculating the distance matrix.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - useAtomWts: (optional) toggles using atom weights for the diagonal elements of the\n",
      "                  matrix (to return a \"Balaban\" distance matrix).\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats with the distance matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetDistanceMatrix(RDKit::ROMol {lvalue} [,bool=False [,bool=False [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetErGFingerprint(...)\n",
      "        GetErGFingerprint( (Mol)mol [, (AtomPairsParameters)atomTypes=0 [, (float)fuzzIncrement=0.3 [, (int)minPath=1 [, (int)maxPath=15]]]]) -> object :\n",
      "            Returns the ErG fingerprint vector for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetErGFingerprint(RDKit::ROMol [,boost::python::api::object=0 [,double=0.3 [,int=1 [,int=15]]]])\n",
      "    \n",
      "    GetFeatureInvariants(...)\n",
      "        GetFeatureInvariants( (Mol)mol) -> list :\n",
      "            Returns feature invariants (FCFP-like) for a molecule.\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetFeatureInvariants(RDKit::ROMol)\n",
      "    \n",
      "    GetFormalCharge(...)\n",
      "        GetFormalCharge( (Mol)arg1) -> int :\n",
      "            Returns the formal charge for the molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int GetFormalCharge(RDKit::ROMol)\n",
      "    \n",
      "    GetHashedAtomPairFingerprint(...)\n",
      "        GetHashedAtomPairFingerprint( (Mol)mol [, (int)nBits=2048 [, (int)minLength=1 [, (int)maxLength=30 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (bool)includeChirality=False [, (bool)use2D=True [, (int)confId=-1]]]]]]]]]) -> IntSparseIntVect :\n",
      "            Returns the hashed atom-pair fingerprint for a molecule as an IntSparseIntVect\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<int>* GetHashedAtomPairFingerprint(RDKit::ROMol [,unsigned int=2048 [,unsigned int=1 [,unsigned int=30 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=False [,bool=True [,int=-1]]]]]]]]])\n",
      "    \n",
      "    GetHashedAtomPairFingerprintAsBitVect(...)\n",
      "        GetHashedAtomPairFingerprintAsBitVect( (Mol)mol [, (int)nBits=2048 [, (int)minLength=1 [, (int)maxLength=30 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (int)nBitsPerEntry=4 [, (bool)includeChirality=False [, (bool)use2D=True [, (int)confId=-1]]]]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns the atom-pair fingerprint for a molecule as an ExplicitBitVect\n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* GetHashedAtomPairFingerprintAsBitVect(RDKit::ROMol [,unsigned int=2048 [,unsigned int=1 [,unsigned int=30 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,unsigned int=4 [,bool=False [,bool=True [,int=-1]]]]]]]]]])\n",
      "    \n",
      "    GetHashedMorganFingerprint(...)\n",
      "        GetHashedMorganFingerprint( (Mol)mol, (int)radius [, (int)nBits=2048 [, (AtomPairsParameters)invariants=[] [, (AtomPairsParameters)fromAtoms=[] [, (bool)useChirality=False [, (bool)useBondTypes=True [, (bool)useFeatures=False [, (AtomPairsParameters)bitInfo=None]]]]]]]) -> UIntSparseIntVect :\n",
      "            Returns a hashed Morgan fingerprint for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<unsigned int>* GetHashedMorganFingerprint(RDKit::ROMol,int [,int=2048 [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,bool=True [,bool=False [,boost::python::api::object=None]]]]]]])\n",
      "    \n",
      "    GetHashedTopologicalTorsionFingerprint(...)\n",
      "        GetHashedTopologicalTorsionFingerprint( (Mol)mol [, (int)nBits=2048 [, (int)targetSize=4 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (bool)includeChirality=False]]]]]]) -> LongSparseIntVect :\n",
      "            Returns the hashed topological-torsion fingerprint for a molecule as a LongIntSparseIntVect\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<long>* GetHashedTopologicalTorsionFingerprint(RDKit::ROMol [,unsigned int=2048 [,unsigned int=4 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=False]]]]]])\n",
      "    \n",
      "    GetHashedTopologicalTorsionFingerprintAsBitVect(...)\n",
      "        GetHashedTopologicalTorsionFingerprintAsBitVect( (Mol)mol [, (int)nBits=2048 [, (int)targetSize=4 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (int)nBitsPerEntry=4 [, (bool)includeChirality=False]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns the topological-torsion fingerprint for a molecule as an ExplicitBitVect\n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* GetHashedTopologicalTorsionFingerprintAsBitVect(RDKit::ROMol [,unsigned int=2048 [,unsigned int=4 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,unsigned int=4 [,bool=False]]]]]]])\n",
      "    \n",
      "    GetMACCSKeysFingerprint(...)\n",
      "        GetMACCSKeysFingerprint( (Mol)mol) -> ExplicitBitVect :\n",
      "            Returns the MACCS keys for a molecule as an ExplicitBitVect\n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* GetMACCSKeysFingerprint(RDKit::ROMol)\n",
      "    \n",
      "    GetMolFrags(...)\n",
      "        GetMolFrags( (Mol)mol [, (bool)asMols=False [, (bool)sanitizeFrags=True]]) -> tuple :\n",
      "            Finds the disconnected fragments from a molecule.\n",
      "            \n",
      "              For example, for the molecule 'CC(=O)[O-].[NH3+]C' GetMolFrags() returns\n",
      "              ((0, 1, 2, 3), (4, 5))\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - asMols: (optional) if this is provided and true, the fragments\n",
      "                  will be returned as molecules instead of atom ids.\n",
      "                - sanitizeFrags: (optional) if this is provided and true, the fragments\n",
      "                  molecules will be sanitized before returning them.\n",
      "            \n",
      "              RETURNS: a tuple of tuples with IDs for the atoms in each fragment\n",
      "                       or a tuple of molecules.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetMolFrags(RDKit::ROMol [,bool=False [,bool=True]])\n",
      "    \n",
      "    GetMoleculeBoundsMatrix(...)\n",
      "        GetMoleculeBoundsMatrix( (Mol)mol [, (bool)set15bounds=True [, (bool)scaleVDW=False]]) -> object :\n",
      "            Returns the distance bounds matrix for a molecule\n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - set15bounds : set bounds for 1-5 atom distances based on \n",
      "                                topology (otherwise stop at 1-4s)\n",
      "                - scaleVDW : scale down the sum of VDW radii when setting the \n",
      "                             lower bounds for atoms less than 5 bonds apart \n",
      "             RETURNS:\n",
      "            \n",
      "                the bounds matrix as a Numeric array with lower bounds in \n",
      "                the lower triangle and upper bounds in the upper triangle\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetMoleculeBoundsMatrix(RDKit::ROMol {lvalue} [,bool=True [,bool=False]])\n",
      "    \n",
      "    GetMorganFingerprint(...)\n",
      "        GetMorganFingerprint( (Mol)mol, (int)radius [, (AtomPairsParameters)invariants=[] [, (AtomPairsParameters)fromAtoms=[] [, (bool)useChirality=False [, (bool)useBondTypes=True [, (bool)useFeatures=False [, (bool)useCounts=True [, (AtomPairsParameters)bitInfo=None]]]]]]]) -> UIntSparseIntVect :\n",
      "            Returns a Morgan fingerprint for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<unsigned int>* GetMorganFingerprint(RDKit::ROMol,int [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,bool=True [,bool=False [,bool=True [,boost::python::api::object=None]]]]]]])\n",
      "    \n",
      "    GetMorganFingerprintAsBitVect(...)\n",
      "        GetMorganFingerprintAsBitVect( (Mol)mol, (int)radius [, (int)nBits=2048 [, (AtomPairsParameters)invariants=[] [, (AtomPairsParameters)fromAtoms=[] [, (bool)useChirality=False [, (bool)useBondTypes=True [, (bool)useFeatures=False [, (AtomPairsParameters)bitInfo=None]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns a Morgan fingerprint for a molecule as a bit vector\n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* GetMorganFingerprintAsBitVect(RDKit::ROMol,int [,unsigned int=2048 [,boost::python::api::object=[] [,boost::python::api::object=[] [,bool=False [,bool=True [,bool=False [,boost::python::api::object=None]]]]]]])\n",
      "    \n",
      "    GetO3A(...)\n",
      "        GetO3A( (Mol)prbMol, (Mol)refMol [, (AtomPairsParameters)prbPyMMFFMolProperties=None [, (AtomPairsParameters)refPyMMFFMolProperties=None [, (int)prbCid=-1 [, (int)refCid=-1 [, (bool)reflect=False [, (int)maxIters=50 [, (int)options=0 [, (list)constraintMap=[] [, (list)constraintWeights=[]]]]]]]]]]) -> O3A :\n",
      "            Get an O3A object with atomMap and weights vectors to overlay\n",
      "                  the probe molecule onto the reference molecule based on\n",
      "                  MMFF atom types and charges\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol                   molecule that is to be aligned\n",
      "                  - refMol                   molecule used as the reference for the alignment\n",
      "                  - prbPyMMFFMolProperties   PyMMFFMolProperties object for the probe molecule as returned\n",
      "                                             by SetupMMFFForceField()\n",
      "                  - refPyMMFFMolProperties   PyMMFFMolProperties object for the reference molecule as returned\n",
      "                                             by SetupMMFFForceField()\n",
      "                  - prbCid                   ID of the conformation in the probe to be used \n",
      "                                             for the alignment (defaults to first conformation)\n",
      "                  - refCid                   ID of the conformation in the ref molecule to which \n",
      "                                             the alignment is computed (defaults to first conformation)\n",
      "                  - reflect                  if true reflect the conformation of the probe molecule\n",
      "                                             (defaults to false)\n",
      "                  - maxIters                 maximum number of iterations used in mimizing the RMSD\n",
      "                                             (defaults to 50)\n",
      "                  - options                  least 2 significant bits encode accuracy\n",
      "                                             (0: maximum, 3: minimum; defaults to 0)\n",
      "                                             bit 3 triggers local optimization of the alignment\n",
      "                                             (no computation of the cost matrix; defaults: off)\n",
      "                  - constraintMap            a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                             which shall be used for the alignment (defaults to [])\n",
      "                  - constraintWeights        optionally specify weights for each of the constraints\n",
      "                                             (weights default to 100.0)\n",
      "                   \n",
      "                  RETURNS\n",
      "                  The O3A object\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolAlign::PyO3A* GetO3A(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,boost::python::api::object=None [,boost::python::api::object=None [,int=-1 [,int=-1 [,bool=False [,unsigned int=50 [,unsigned int=0 [,boost::python::list=[] [,boost::python::list=[]]]]]]]]]])\n",
      "    \n",
      "    GetO3AForProbeConfs(...)\n",
      "        GetO3AForProbeConfs( (Mol)prbMol, (Mol)refMol [, (int)numThreads=1 [, (AtomPairsParameters)prbPyMMFFMolProperties=None [, (AtomPairsParameters)refPyMMFFMolProperties=None [, (int)refCid=-1 [, (bool)reflect=False [, (int)maxIters=50 [, (int)options=0 [, (list)constraintMap=[] [, (list)constraintWeights=[]]]]]]]]]]) -> tuple :\n",
      "            Get a vector of O3A objects for the overlay of all \n",
      "                  the probe molecule's conformations onto the reference molecule based on\n",
      "                  MMFF atom types and charges\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - prbMol                   molecule that is to be aligned\n",
      "                  - refMol                   molecule used as the reference for the alignment\n",
      "                  - numThreads :             the number of threads to use, only has an effect if\n",
      "                                             the RDKit was built with thread support (defaults to 1)\n",
      "                                             If set to zero, the max supported by the system will be used.\n",
      "                  - prbPyMMFFMolProperties   PyMMFFMolProperties object for the probe molecule as returned\n",
      "                                             by SetupMMFFForceField()\n",
      "                  - refPyMMFFMolProperties   PyMMFFMolProperties object for the reference molecule as returned\n",
      "                                             by SetupMMFFForceField()\n",
      "                  - refCid                   ID of the conformation in the ref molecule to which \n",
      "                                             the alignment is computed (defaults to first conformation)\n",
      "                  - reflect                  if true reflect the conformation of the probe molecule\n",
      "                                             (defaults to false)\n",
      "                  - maxIters                 maximum number of iterations used in mimizing the RMSD\n",
      "                                             (defaults to 50)\n",
      "                  - options                  least 2 significant bits encode accuracy\n",
      "                                             (0: maximum, 3: minimum; defaults to 0)\n",
      "                                             bit 3 triggers local optimization of the alignment\n",
      "                                             (no computation of the cost matrix; defaults: off)\n",
      "                  - constraintMap            a vector of pairs of atom IDs (probe AtomId, ref AtomId)\n",
      "                                             which shall be used for the alignment (defaults to [])\n",
      "                  - constraintWeights        optionally specify weights for each of the constraints\n",
      "                                             (weights default to 100.0)\n",
      "                   \n",
      "                  RETURNS\n",
      "                  A vector of O3A objects\n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetO3AForProbeConfs(RDKit::ROMol {lvalue},RDKit::ROMol {lvalue} [,int=1 [,boost::python::api::object=None [,boost::python::api::object=None [,int=-1 [,bool=False [,unsigned int=50 [,unsigned int=0 [,boost::python::list=[] [,boost::python::list=[]]]]]]]]]])\n",
      "    \n",
      "    GetPeriodicTable(...)\n",
      "        GetPeriodicTable() -> PeriodicTable :\n",
      "            Returns the application's PeriodicTable instance.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::PeriodicTable* GetPeriodicTable()\n",
      "    \n",
      "    GetSSSR(...)\n",
      "        GetSSSR( (Mol)arg1) -> int :\n",
      "            Get the smallest set of simple rings for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: a sequence of sequences containing the rings found as atom ids\n",
      "                     The length of this will be equal to NumBonds-NumAtoms+1 for single-fragment molecules.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int GetSSSR(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    GetShortestPath(...)\n",
      "        GetShortestPath( (Mol)arg1, (int)arg2, (int)arg3) -> tuple :\n",
      "            Find the shortest path between two atoms using the Bellman-Ford algorithm.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - idx1: index of the first atom\n",
      "                - idx2: index of the second atom\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetShortestPath(RDKit::ROMol,int,int)\n",
      "    \n",
      "    GetSupplementalSmilesLabel(...)\n",
      "        GetSupplementalSmilesLabel( (Atom)atom) -> str :\n",
      "            Gets the supplemental smiles label on an atom, returns an empty string if not present.\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string GetSupplementalSmilesLabel(RDKit::Atom const*)\n",
      "    \n",
      "    GetSymmSSSR(...)\n",
      "        GetSymmSSSR( (Mol)arg1) -> _vectSt6vectorIiSaIiEE :\n",
      "            Get a symmetrized SSSR for a molecule.\n",
      "            \n",
      "              The symmetrized SSSR is at least as large as the SSSR for a molecule.\n",
      "              In certain highly-symmetric cases (e.g. cubane), the symmetrized SSSR can be\n",
      "              a bit larger (i.e. the number of symmetrized rings is >= NumBonds-NumAtoms+1).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: a sequence of sequences containing the rings found as atom ids\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > GetSymmSSSR(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    GetTopologicalTorsionFingerprint(...)\n",
      "        GetTopologicalTorsionFingerprint( (Mol)mol [, (int)targetSize=4 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)ignoreAtoms=0 [, (AtomPairsParameters)atomInvariants=0 [, (bool)includeChirality=False]]]]]) -> LongSparseIntVect :\n",
      "            Returns the topological-torsion fingerprint for a molecule as a LongIntSparseIntVect\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<long>* GetTopologicalTorsionFingerprint(RDKit::ROMol [,unsigned int=4 [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=False]]]]])\n",
      "    \n",
      "    GetUFFAngleBendParams(...)\n",
      "        GetUFFAngleBendParams( (Mol)mol, (int)idx1, (int)idx2, (int)idx3) -> object :\n",
      "            Retrieves UFF angle bend parameters for atoms with indexes idx1, idx2, idx3 as a (ka, theta0) tuple, or None if no parameters could be found\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetUFFAngleBendParams(RDKit::ROMol,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetUFFBondStretchParams(...)\n",
      "        GetUFFBondStretchParams( (Mol)mol, (int)idx1, (int)idx2) -> object :\n",
      "            Retrieves UFF bond stretch parameters for atoms with indexes idx1, idx2 as a (kb, r0) tuple, or None if no parameters could be found\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetUFFBondStretchParams(RDKit::ROMol,unsigned int,unsigned int)\n",
      "    \n",
      "    GetUFFInversionParams(...)\n",
      "        GetUFFInversionParams( (Mol)mol, (int)idx1, (int)idx2, (int)idx3, (int)idx4) -> object :\n",
      "            Retrieves UFF inversion parameters for atoms with indexes idx1, idx2, idx3, idx4 as a K float value, or None if no parameters could be found\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetUFFInversionParams(RDKit::ROMol,unsigned int,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetUFFTorsionParams(...)\n",
      "        GetUFFTorsionParams( (Mol)mol, (int)idx1, (int)idx2, (int)idx3, (int)idx4) -> object :\n",
      "            Retrieves UFF torsion parameters for atoms with indexes idx1, idx2, idx3, idx4 as a V float value, or None if no parameters could be found\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetUFFTorsionParams(RDKit::ROMol,unsigned int,unsigned int,unsigned int,unsigned int)\n",
      "    \n",
      "    GetUFFVdWParams(...)\n",
      "        GetUFFVdWParams( (Mol)mol, (int)idx1, (int)idx2) -> object :\n",
      "            Retrieves UFF van der Waals parameters for atoms with indexes idx1, idx2 as a (x_ij, D_ij) tuple, or None if no parameters could be found\n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetUFFVdWParams(RDKit::ROMol,unsigned int,unsigned int)\n",
      "    \n",
      "    GetUSR(...)\n",
      "        GetUSR( (Mol)mol [, (int)confId=-1]) -> list :\n",
      "            Returns a USR descriptor for one conformer of a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetUSR(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    GetUSRCAT(...)\n",
      "        GetUSRCAT( (Mol)mol [, (AtomPairsParameters)atomSelections=None [, (int)confId=-1]]) -> list :\n",
      "            Returns a USRCAT descriptor for one conformer of a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetUSRCAT(RDKit::ROMol [,boost::python::api::object=None [,int=-1]])\n",
      "    \n",
      "    GetUSRDistributions(...)\n",
      "        GetUSRDistributions( (AtomPairsParameters)coords [, (AtomPairsParameters)points=None]) -> list :\n",
      "            Returns the four USR distance distributions for a set of coordinates\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetUSRDistributions(boost::python::api::object [,boost::python::api::object=None])\n",
      "    \n",
      "    GetUSRDistributionsFromPoints(...)\n",
      "        GetUSRDistributionsFromPoints( (AtomPairsParameters)coords, (AtomPairsParameters)points) -> list :\n",
      "            Returns the USR distance distributions for a set of coordinates and points\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetUSRDistributionsFromPoints(boost::python::api::object,boost::python::api::object)\n",
      "    \n",
      "    GetUSRFromDistributions(...)\n",
      "        GetUSRFromDistributions( (AtomPairsParameters)distances) -> list :\n",
      "            Returns the USR descriptor from a set of distance distributions\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list GetUSRFromDistributions(boost::python::api::object)\n",
      "    \n",
      "    GetUSRScore(...)\n",
      "        GetUSRScore( (AtomPairsParameters)descriptor1, (AtomPairsParameters)descriptor2 [, (AtomPairsParameters)weights=[]]) -> float :\n",
      "            Returns the USR score for two USR or USRCAT descriptors\n",
      "        \n",
      "            C++ signature :\n",
      "                double GetUSRScore(boost::python::api::object,boost::python::api::object [,boost::python::api::object=[]])\n",
      "    \n",
      "    HCountEqualsQueryAtom(...)\n",
      "        HCountEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where HCount is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HCountEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    HCountGreaterQueryAtom(...)\n",
      "        HCountGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where HCount is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HCountGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    HCountLessQueryAtom(...)\n",
      "        HCountLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where HCount is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HCountLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    HasAgentTemplateSubstructMatch(...)\n",
      "        HasAgentTemplateSubstructMatch( (ChemicalReaction)reaction, (ChemicalReaction)queryReaction) -> bool :\n",
      "            tests if the agents of a queryReaction are the same as those of a reaction\n",
      "        \n",
      "            C++ signature :\n",
      "                bool HasAgentTemplateSubstructMatch(RDKit::ChemicalReaction,RDKit::ChemicalReaction)\n",
      "    \n",
      "    HasBoolPropWithValueQueryAtom(...)\n",
      "        HasBoolPropWithValueQueryAtom( (str)propname, (bool)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches when the propery 'propname' has the specified boolean value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasBoolPropWithValueQueryAtom(std::string,bool [,bool=False])\n",
      "    \n",
      "    HasBoolPropWithValueQueryBond(...)\n",
      "        HasBoolPropWithValueQueryBond( (str)propname, (bool)val [, (bool)negate=False]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' has the specified boolean value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasBoolPropWithValueQueryBond(std::string,bool [,bool=False])\n",
      "    \n",
      "    HasChiralTagQueryAtom(...)\n",
      "        HasChiralTagQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when HasChiralTag is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasChiralTagQueryAtom([ bool=False])\n",
      "    \n",
      "    HasDoublePropWithValueQueryAtom(...)\n",
      "        HasDoublePropWithValueQueryAtom( (str)propname, (float)val [, (bool)negate=False [, (float)tolerance=0.0]]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches when the propery 'propname' has the specified value +- tolerance\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasDoublePropWithValueQueryAtom(std::string,double [,bool=False [,double=0.0]])\n",
      "    \n",
      "    HasDoublePropWithValueQueryBond(...)\n",
      "        HasDoublePropWithValueQueryBond( (str)propname, (float)val [, (bool)negate=False [, (float)tolerance=0.0]]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' has the specified value +- tolerance\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasDoublePropWithValueQueryBond(std::string,double [,bool=False [,double=0.0]])\n",
      "    \n",
      "    HasIntPropWithValueQueryAtom(...)\n",
      "        HasIntPropWithValueQueryAtom( (str)propname, (int)val [, (bool)negate=False [, (int)tolerance=0]]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches when the propery 'propname' has the specified int value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasIntPropWithValueQueryAtom(std::string,int [,bool=False [,int=0]])\n",
      "    \n",
      "    HasIntPropWithValueQueryBond(...)\n",
      "        HasIntPropWithValueQueryBond( (str)propname, (int)val [, (bool)negate=False [, (int)tolerance=0]]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' has the specified int value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasIntPropWithValueQueryBond(std::string,int [,bool=False [,int=0]])\n",
      "    \n",
      "    HasProductTemplateSubstructMatch(...)\n",
      "        HasProductTemplateSubstructMatch( (ChemicalReaction)reaction, (ChemicalReaction)queryReaction) -> bool :\n",
      "            tests if the products of a queryReaction are substructures of the products of a reaction\n",
      "        \n",
      "            C++ signature :\n",
      "                bool HasProductTemplateSubstructMatch(RDKit::ChemicalReaction,RDKit::ChemicalReaction)\n",
      "    \n",
      "    HasPropQueryAtom(...)\n",
      "        HasPropQueryAtom( (str)propname [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches when the propery 'propname' exists in the atom.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasPropQueryAtom(std::string [,bool=False])\n",
      "    \n",
      "    HasPropQueryBond(...)\n",
      "        HasPropQueryBond( (str)propname [, (bool)negate=False]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' exists in the bond.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasPropQueryBond(std::string [,bool=False])\n",
      "        \n",
      "        HasPropQueryBond( (str)propname [, (bool)negate=False]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' exists in the bond.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasPropQueryBond(std::string [,bool=False])\n",
      "        \n",
      "        HasPropQueryBond( (str)propname [, (bool)negate=False]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' exists in the bond.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasPropQueryBond(std::string [,bool=False])\n",
      "    \n",
      "    HasReactantTemplateSubstructMatch(...)\n",
      "        HasReactantTemplateSubstructMatch( (ChemicalReaction)reaction, (ChemicalReaction)queryReaction) -> bool :\n",
      "            tests if the reactants of a queryReaction are substructures of the reactants of a reaction\n",
      "        \n",
      "            C++ signature :\n",
      "                bool HasReactantTemplateSubstructMatch(RDKit::ChemicalReaction,RDKit::ChemicalReaction)\n",
      "    \n",
      "    HasReactionAtomMapping(...)\n",
      "        HasReactionAtomMapping( (ChemicalReaction)arg1) -> bool :\n",
      "            tests if a reaction obtains any atom mapping\n",
      "        \n",
      "            C++ signature :\n",
      "                bool HasReactionAtomMapping(RDKit::ChemicalReaction)\n",
      "    \n",
      "    HasReactionSubstructMatch(...)\n",
      "        HasReactionSubstructMatch( (ChemicalReaction)reaction, (ChemicalReaction)queryReaction [, (bool)includeAgents=False]) -> bool :\n",
      "            tests if the queryReaction is a substructure of a reaction\n",
      "        \n",
      "            C++ signature :\n",
      "                bool HasReactionSubstructMatch(RDKit::ChemicalReaction,RDKit::ChemicalReaction [,bool=False])\n",
      "    \n",
      "    HasStringPropWithValueQueryAtom(...)\n",
      "        HasStringPropWithValueQueryAtom( (str)propname, (str)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches when the propery 'propname' has the specified string value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HasStringPropWithValueQueryAtom(std::string,std::string [,bool=False])\n",
      "    \n",
      "    HasStringPropWithValueQueryBond(...)\n",
      "        HasStringPropWithValueQueryBond( (str)propname, (str)val [, (bool)negate=False]) -> QueryBond :\n",
      "            Returns a QueryBond that matches when the propery 'propname' has the specified string value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryBond* HasStringPropWithValueQueryBond(std::string,std::string [,bool=False])\n",
      "    \n",
      "    HybridizationEqualsQueryAtom(...)\n",
      "        HybridizationEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Hybridization is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HybridizationEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    HybridizationGreaterQueryAtom(...)\n",
      "        HybridizationGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Hybridization is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HybridizationGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    HybridizationLessQueryAtom(...)\n",
      "        HybridizationLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Hybridization is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* HybridizationLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    InNRingsEqualsQueryAtom(...)\n",
      "        InNRingsEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where InNRings is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* InNRingsEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    InNRingsGreaterQueryAtom(...)\n",
      "        InNRingsGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where InNRings is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* InNRingsGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    InNRingsLessQueryAtom(...)\n",
      "        InNRingsLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where InNRings is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* InNRingsLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    IsAliphaticQueryAtom(...)\n",
      "        IsAliphaticQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when IsAliphatic is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsAliphaticQueryAtom([ bool=False])\n",
      "    \n",
      "    IsAromaticQueryAtom(...)\n",
      "        IsAromaticQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when IsAromatic is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsAromaticQueryAtom([ bool=False])\n",
      "    \n",
      "    IsInRingQueryAtom(...)\n",
      "        IsInRingQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when IsInRing is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsInRingQueryAtom([ bool=False])\n",
      "    \n",
      "    IsReactionTemplateMoleculeAgent(...)\n",
      "        IsReactionTemplateMoleculeAgent( (Mol)molecule, (float)agentThreshold) -> bool :\n",
      "            tests if a molecule can be classified as an agent depending on the ratio of mapped atoms and a give threshold\n",
      "        \n",
      "            C++ signature :\n",
      "                bool IsReactionTemplateMoleculeAgent(RDKit::ROMol,double)\n",
      "    \n",
      "    IsUnsaturatedQueryAtom(...)\n",
      "        IsUnsaturatedQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when IsUnsaturated is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsUnsaturatedQueryAtom([ bool=False])\n",
      "    \n",
      "    IsotopeEqualsQueryAtom(...)\n",
      "        IsotopeEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Isotope is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsotopeEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    IsotopeGreaterQueryAtom(...)\n",
      "        IsotopeGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Isotope is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsotopeGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    IsotopeLessQueryAtom(...)\n",
      "        IsotopeLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Isotope is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* IsotopeLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    KDG(...)\n",
      "        KDG() -> EmbedParameters :\n",
      "            Returns an EmbedParameters object for the KDG method.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::DGeomHelpers::EmbedParameters* KDG()\n",
      "    \n",
      "    Kekulize(...)\n",
      "        Kekulize( (Mol)mol [, (bool)clearAromaticFlags=False]) -> None :\n",
      "            Kekulizes the molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - clearAromaticFlags: (optional) if this toggle is set, all atoms and bonds in the \n",
      "                  molecule will be marked non-aromatic following the kekulization.\n",
      "                  Default value is 0.\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void Kekulize(RDKit::ROMol {lvalue} [,bool=False])\n",
      "    \n",
      "    LayeredFingerprint(...)\n",
      "        LayeredFingerprint( (Mol)mol [, (int)layerFlags=4294967295 [, (int)minPath=1 [, (int)maxPath=7 [, (int)fpSize=2048 [, (list)atomCounts=[] [, (ExplicitBitVect)setOnlyBits=None [, (bool)branchedPaths=True [, (AtomPairsParameters)fromAtoms=0]]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns a layered fingerprint for a molecule\n",
      "            \n",
      "              NOTE: This function is experimental. The API or results may change from\n",
      "                release to release.\n",
      "            \n",
      "              Explanation of the algorithm below.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - layerFlags: (optional) which layers to include in the fingerprint\n",
      "                  See below for definitions. Defaults to all.\n",
      "            \n",
      "                - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                  Defaults to 7.\n",
      "            \n",
      "                - fpSize: (optional) number of bits in the fingerprint\n",
      "                  Defaults to 2048.\n",
      "            \n",
      "                - atomCounts: (optional) \n",
      "                    if provided, this should be a list at least as long as the number of atoms\n",
      "                    in the molecule. It will be used to provide the count of the number \n",
      "                    of paths that set bits each atom is involved in.\n",
      "                    NOTE: the list is not zeroed out here.\n",
      "            \n",
      "                - setOnlyBits: (optional) \n",
      "                    if provided, only bits that are set in this bit vector will be set\n",
      "                    in the result. This is essentially the same as doing:\n",
      "                       res &= setOnlyBits\n",
      "                    but also has an impact on the atomCounts (if being used)\n",
      "            \n",
      "                - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                  used in the fingerprint.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                  starting from these atoms will be used.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "              RETURNS: a DataStructs.ExplicitBitVect with _fpSize_ bits\n",
      "            \n",
      "              Layer definitions:\n",
      "                 - 0x01: pure topology\n",
      "                 - 0x02: bond order\n",
      "                 - 0x04: atom types\n",
      "                 - 0x08: presence of rings\n",
      "                 - 0x10: ring sizes\n",
      "                 - 0x20: aromaticity\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* LayeredFingerprint(RDKit::ROMol [,unsigned int=4294967295 [,unsigned int=1 [,unsigned int=7 [,unsigned int=2048 [,boost::python::list=[] [,ExplicitBitVect*=None [,bool=True [,boost::python::api::object=0]]]]]]]])\n",
      "    \n",
      "    LogErrorMsg(...)\n",
      "        LogErrorMsg( (str)arg1) -> None :\n",
      "            Log a warning message to the RDKit error logs\n",
      "        \n",
      "            C++ signature :\n",
      "                void LogErrorMsg(std::string)\n",
      "    \n",
      "    LogWarningMsg(...)\n",
      "        LogWarningMsg( (str)arg1) -> None :\n",
      "            Log a warning message to the RDKit warning logs\n",
      "        \n",
      "            C++ signature :\n",
      "                void LogWarningMsg(std::string)\n",
      "    \n",
      "    MMFFGetMoleculeForceField(...)\n",
      "        MMFFGetMoleculeForceField( (Mol)mol, (MMFFMolProperties)pyMMFFMolProperties [, (float)nonBondedThresh=100.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]) -> ForceField :\n",
      "            returns a MMFF force field for a molecule\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - pyMMFFMolProperties : PyMMFFMolProperties object as returned\n",
      "                              by MMFFGetMoleculeProperties()\n",
      "                - nonBondedThresh : used to exclude long-range non-bonded\n",
      "                              interactions (defaults to 100.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                              fragments will not be added to the forcefield\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ForceFields::PyForceField* MMFFGetMoleculeForceField(RDKit::ROMol {lvalue},ForceFields::PyMMFFMolProperties* [,double=100.0 [,int=-1 [,bool=True]]])\n",
      "    \n",
      "    MMFFGetMoleculeProperties(...)\n",
      "        MMFFGetMoleculeProperties( (Mol)mol [, (str)mmffVariant='MMFF94' [, (int)mmffVerbosity=0]]) -> MMFFMolProperties :\n",
      "            returns a PyMMFFMolProperties object for a\n",
      "              molecule, which is required by MMFFGetMoleculeForceField()\n",
      "              and can be used to get/set MMFF properties\n",
      "            \n",
      "              \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - mmffVariant : \"MMFF94\" or \"MMFF94s\"\n",
      "                              (defaults to \"MMFF94\")\n",
      "                - mmffVerbosity : 0: none; 1: low; 2: high (defaults to 0).\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ForceFields::PyMMFFMolProperties* MMFFGetMoleculeProperties(RDKit::ROMol {lvalue} [,std::string='MMFF94' [,unsigned int=0]])\n",
      "    \n",
      "    MMFFHasAllMoleculeParams(...)\n",
      "        MMFFHasAllMoleculeParams( (Mol)mol) -> bool :\n",
      "            checks if MMFF parameters are available for all of a molecule's atoms\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                bool MMFFHasAllMoleculeParams(RDKit::ROMol)\n",
      "    \n",
      "    MMFFOptimizeMolecule(...)\n",
      "        MMFFOptimizeMolecule( (Mol)mol [, (str)mmffVariant='MMFF94' [, (int)maxIters=200 [, (float)nonBondedThresh=100.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]]]) -> int :\n",
      "            uses MMFF to optimize a molecule's structure\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - mmffVariant : \"MMFF94\" or \"MMFF94s\"\n",
      "                - maxIters : the maximum number of iterations (defaults to 200)\n",
      "                - nonBondedThresh : used to exclude long-range non-bonded\n",
      "                             interactions (defaults to 100.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                             fragments will not be added to the forcefield\n",
      "            \n",
      "             RETURNS: 0 if the optimization converged, -1 if the forcefield could\n",
      "                      not be set up, 1 if more iterations are required.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int MMFFOptimizeMolecule(RDKit::ROMol {lvalue} [,std::string='MMFF94' [,int=200 [,double=100.0 [,int=-1 [,bool=True]]]]])\n",
      "    \n",
      "    MMFFOptimizeMoleculeConfs(...)\n",
      "        MMFFOptimizeMoleculeConfs( (Mol)self [, (int)numThreads=1 [, (int)maxIters=200 [, (str)mmffVariant='MMFF94' [, (float)nonBondedThresh=10.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]]]]) -> object :\n",
      "            uses MMFF to optimize all of a molecule's conformations\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - numThreads : the number of threads to use, only has an effect if the RDKit\n",
      "                               was built with thread support (defaults to 1)\n",
      "                               If set to zero, the max supported by the system will be used.\n",
      "                - maxIters : the maximum number of iterations (defaults to 200)\n",
      "                - mmffVariant : \"MMFF94\" or \"MMFF94s\"\n",
      "                - nonBondedThresh : used to exclude long-range non-bonded\n",
      "                              interactions (defaults to 100.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                              fragments will not be added to the forcefield.\n",
      "            \n",
      "             RETURNS: 0 if the optimization converged, 1 if more iterations are required.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object MMFFOptimizeMoleculeConfs(RDKit::ROMol {lvalue} [,int=1 [,int=200 [,std::string='MMFF94' [,double=10.0 [,int=-1 [,bool=True]]]]]])\n",
      "    \n",
      "    MMFFSanitizeMolecule(...)\n",
      "        MMFFSanitizeMolecule( (Mol)mol) -> int :\n",
      "            sanitizes a molecule according to MMFF requirements.\n",
      "            \n",
      "                - mol : the molecule of interest.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int MMFFSanitizeMolecule(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    MQNs_(...)\n",
      "        MQNs_( (Mol)mol [, (bool)force=False]) -> list :\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list MQNs_(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    MakePropertyRangeQuery(...)\n",
      "        MakePropertyRangeQuery( (str)name, (float)min, (float)max) -> PropertyRangeQuery :\n",
      "            Generates a Range property for the specified property, between min and max\n",
      "            query = MakePropertyRangeQuery('exactmw', 0, 500)\n",
      "            query.Match( mol )\n",
      "        \n",
      "            C++ signature :\n",
      "                Queries::RangeQuery<double, RDKit::ROMol const&, true>* MakePropertyRangeQuery(std::string,double,double)\n",
      "    \n",
      "    MassEqualsQueryAtom(...)\n",
      "        MassEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Mass is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MassEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MassGreaterQueryAtom(...)\n",
      "        MassGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Mass is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MassGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MassLessQueryAtom(...)\n",
      "        MassLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where Mass is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MassLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MatchOnlyAtRgroupsAdjustParams(...)\n",
      "        MatchOnlyAtRgroupsAdjustParams() -> AdjustQueryParameters :\n",
      "            Only match at the specified rgroup locations in the reactant templates\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolOps::AdjustQueryParameters MatchOnlyAtRgroupsAdjustParams()\n",
      "    \n",
      "    MergeQueryHs(...)\n",
      "        MergeQueryHs( (Mol)mol [, (bool)mergeUnmappedOnly=False]) -> Mol :\n",
      "            merges hydrogens into their neighboring atoms as queries\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MergeQueryHs(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    MinRingSizeEqualsQueryAtom(...)\n",
      "        MinRingSizeEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where MinRingSize is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MinRingSizeEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MinRingSizeGreaterQueryAtom(...)\n",
      "        MinRingSizeGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where MinRingSize is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MinRingSizeGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MinRingSizeLessQueryAtom(...)\n",
      "        MinRingSizeLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where MinRingSize is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MinRingSizeLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    MissingChiralTagQueryAtom(...)\n",
      "        MissingChiralTagQueryAtom([  (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms when MissingChiralTag is True.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* MissingChiralTagQueryAtom([ bool=False])\n",
      "    \n",
      "    MolAddRecursiveQueries(...)\n",
      "        MolAddRecursiveQueries( (Mol)mol, (dict)queries, (str)propName) -> None :\n",
      "            Adds named recursive queries to atoms\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolAddRecursiveQueries(RDKit::ROMol {lvalue},boost::python::dict,std::string)\n",
      "    \n",
      "    MolFragmentToSmiles(...)\n",
      "        MolFragmentToSmiles( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (AtomPairsParameters)atomSymbols=0 [, (AtomPairsParameters)bondSymbols=0 [, (bool)isomericSmiles=False [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False]]]]]]]]]) -> str :\n",
      "            Returns the canonical SMILES string for a fragment of a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse : a list of atoms to include in the fragment\n",
      "                - bondsToUse : (optional) a list of bonds to include in the fragment\n",
      "                               if not provided, all bonds between the atoms provided\n",
      "                               will be included.\n",
      "                - atomSymbols : (optional) a list with the symbols to use for the atoms\n",
      "                                in the SMILES. This should have be mol.GetNumAtoms() long.\n",
      "                - bondSymbols : (optional) a list with the symbols to use for the bonds\n",
      "                                in the SMILES. This should have be mol.GetNumBonds() long.\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolFragmentToSmiles(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=False [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False]]]]]]]]])\n",
      "    \n",
      "    MolFromFASTA(...)\n",
      "        MolFromFASTA( (AtomPairsParameters)text [, (bool)sanitize=True [, (int)flavor=0]]) -> Mol :\n",
      "            Construct a molecule from a FASTA string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the FASTA\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "            - flavor: (optional)\n",
      "              0 Protein, L amino acids (default)\n",
      "              1 Protein, D amino acids\n",
      "              2 RNA, no cap\n",
      "              3 RNA, 5' cap\n",
      "              4 RNA, 3' cap\n",
      "              5 RNA, both caps\n",
      "              6 DNA, no cap\n",
      "              7 DNA, 5' cap\n",
      "              8 DNA, 3' cap\n",
      "              9 DNA, both caps\n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromFASTA(boost::python::api::object [,bool=True [,int=0]])\n",
      "    \n",
      "    MolFromHELM(...)\n",
      "        MolFromHELM( (AtomPairsParameters)text [, (bool)sanitize=True]) -> Mol :\n",
      "            Construct a molecule from a HELM string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the HELM\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromHELM(boost::python::api::object [,bool=True])\n",
      "    \n",
      "    MolFromMol2Block(...)\n",
      "        MolFromMol2Block( (str)molBlock [, (bool)sanitize=True [, (bool)removeHs=True]]) -> Mol :\n",
      "            Construct a molecule from a Tripos Mol2 block.\n",
      "            \n",
      "              NOTE:\n",
      "                 The parser expects the atom-typing scheme used by Corina.\n",
      "                Atom types from Tripos' dbtranslate are less supported.\n",
      "                Other atom typing schemes are unlikely to work.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol2Block: string containing the Mol2 block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMol2Block(std::string [,bool=True [,bool=True]])\n",
      "    \n",
      "    MolFromMol2File(...)\n",
      "        MolFromMol2File( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True]]) -> Mol :\n",
      "            Construct a molecule from a Tripos Mol2 file.\n",
      "            \n",
      "              NOTE:\n",
      "                 The parser expects the atom-typing scheme used by Corina.\n",
      "                Atom types from Tripos' dbtranslate are less supported.\n",
      "                Other atom typing schemes are unlikely to work.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "                                              \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMol2File(char const* [,bool=True [,bool=True]])\n",
      "    \n",
      "    MolFromMolBlock(...)\n",
      "        MolFromMolBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the Mol block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolBlock(boost::python::api::object [,bool=True [,bool=True [,bool=True]]])\n",
      "        \n",
      "        MolFromMolBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the Mol block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolBlock(boost::python::api::object [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromMolFile(...)\n",
      "        MolFromMolFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolFile(char const* [,bool=True [,bool=True [,bool=True]]])\n",
      "        \n",
      "        MolFromMolFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolFile(char const* [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromPDBBlock(...)\n",
      "        MolFromPDBBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (int)flavor=0]]]) -> Mol :\n",
      "            Construct a molecule from a PDB block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the PDB block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - flavor: (optional) \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromPDBBlock(boost::python::api::object [,bool=True [,bool=True [,unsigned int=0]]])\n",
      "    \n",
      "    MolFromPDBFile(...)\n",
      "        MolFromPDBFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (int)flavor=0]]]) -> Mol :\n",
      "            Construct a molecule from a PDB file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - flavor: (optional) \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromPDBFile(char const* [,bool=True [,bool=True [,unsigned int=0]]])\n",
      "    \n",
      "    MolFromQuerySLN(...)\n",
      "        MolFromQuerySLN( (str)SLN [, (bool)mergeHs=True [, (bool)debugParser=False]]) -> Mol :\n",
      "            Construct a query molecule from an SLN string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - SLN: the SLN string\n",
      "            \n",
      "                - mergeHs: (optional) toggles the merging of explicit Hs in the query into the attached\n",
      "                  heavy atoms. Defaults to False.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object suitable for using in substructure queries, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromQuerySLN(std::string [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolFromSLN(...)\n",
      "        MolFromSLN( (str)SLN [, (bool)sanitize=True [, (bool)debugParser=False]]) -> Mol :\n",
      "            Construct a molecule from an SLN string.\n",
      "            \n",
      "                ARGUMENTS:\n",
      "            \n",
      "                - SLN: the SLN string\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "              NOTE: the SLN should not contain query information or properties. To build a\n",
      "                query from SLN, use MolFromQuerySLN.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSLN(std::string [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolFromSequence(...)\n",
      "        MolFromSequence( (AtomPairsParameters)text [, (bool)sanitize=True [, (int)flavor=0]]) -> Mol :\n",
      "            Construct a molecule from a sequence string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the sequence\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - flavor: (optional)\n",
      "                  0 Protein, L amino acids (default)\n",
      "                  1 Protein, D amino acids\n",
      "                  2 RNA, no cap\n",
      "                  3 RNA, 5' cap\n",
      "                  4 RNA, 3' cap\n",
      "                  5 RNA, both caps\n",
      "                  6 DNA, no cap\n",
      "                  7 DNA, 5' cap\n",
      "                  8 DNA, 3' cap\n",
      "                  9 DNA, both caps\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSequence(boost::python::api::object [,bool=True [,int=0]])\n",
      "    \n",
      "    MolFromSmarts(...)\n",
      "        MolFromSmarts( (AtomPairsParameters)SMARTS [, (bool)mergeHs=False [, (dict)replacements={}]]) -> Mol :\n",
      "            Construct a molecule from a SMARTS string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - SMARTS: the smarts string\n",
      "            \n",
      "                - mergeHs: (optional) toggles the merging of explicit Hs in the query into the attached\n",
      "                  atoms.  So, for example, 'C[H]' becomes '[C;!H0]'.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - replacements: (optional) a dictionary of replacement strings (see below)\n",
      "                  Defaults to {}. See the documentation for MolFromSmiles for an explanation.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmarts(boost::python::api::object [,bool=False [,boost::python::dict={}]])\n",
      "    \n",
      "    MolFromSmiles(...)\n",
      "        MolFromSmiles( (AtomPairsParameters)SMILES, (SmilesParserParams)params) -> Mol :\n",
      "            Construct a molecule from a SMILES string.\n",
      "            \n",
      "                 ARGUMENTS:\n",
      "               \n",
      "                   - SMILES: the smiles string\n",
      "               \n",
      "                   - params: used to provide optional parameters for the SMILES parsing\n",
      "               \n",
      "                 RETURNS:\n",
      "               \n",
      "                   a Mol object, None on failure.\n",
      "               \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmiles(boost::python::api::object,RDKit::SmilesParserParams)\n",
      "        \n",
      "        MolFromSmiles( (AtomPairsParameters)SMILES [, (bool)sanitize=True [, (dict)replacements={}]]) -> Mol :\n",
      "            Construct a molecule from a SMILES string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - SMILES: the smiles string\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - replacements: (optional) a dictionary of replacement strings (see below)\n",
      "                  Defaults to {}.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "               The optional replacements dict can be used to do string substitution of abbreviations \n",
      "               in the input SMILES. The set of substitutions is repeatedly looped through until \n",
      "               the string no longer changes. It is the responsiblity of the caller to make sure \n",
      "               that substitutions results in legal and sensible SMILES. \n",
      "             \n",
      "               Examples of replacements: \n",
      "             \n",
      "                 CC{Q}C with {'{Q}':'OCCO'} -> CCOCCOC  \n",
      "                 C{A}C{Q}C with {'{Q}':'OCCO', '{A}':'C1(CC1)'} -> CC1(CC1)COCCOC  \n",
      "                 C{A}C{Q}C with {'{Q}':'{X}CC{X}', '{A}':'C1CC1', '{X}':'N'} -> CC1CC1CCNCCNC  \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmiles(boost::python::api::object [,bool=True [,boost::python::dict={}]])\n",
      "    \n",
      "    MolFromTPLBlock(...)\n",
      "        MolFromTPLBlock( (AtomPairsParameters)tplBlock [, (bool)sanitize=True [, (bool)skipFirstConf=False]]) -> Mol :\n",
      "            Construct a molecule from a TPL block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - skipFirstConf: (optional) skips reading the first conformer.\n",
      "                  Defaults to False.\n",
      "                  This should be set to True when reading TPLs written by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromTPLBlock(boost::python::api::object [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolFromTPLFile(...)\n",
      "        MolFromTPLFile( (str)fileName [, (bool)sanitize=True [, (bool)skipFirstConf=False]]) -> Mol :\n",
      "            Construct a molecule from a TPL file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - skipFirstConf: (optional) skips reading the first conformer.\n",
      "                  Defaults to False.\n",
      "                  This should be set to True when reading TPLs written by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromTPLFile(char const* [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolToFASTA(...)\n",
      "        MolToFASTA( (Mol)mol) -> str :\n",
      "            Returns the FASTA string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToFASTA(RDKit::ROMol)\n",
      "    \n",
      "    MolToHELM(...)\n",
      "        MolToHELM( (Mol)mol) -> str :\n",
      "            Returns the HELM string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToHELM(RDKit::ROMol)\n",
      "    \n",
      "    MolToMolBlock(...)\n",
      "        MolToMolBlock( (Mol)mol [, (bool)includeStereo=False [, (int)confId=-1 [, (bool)kekulize=True [, (bool)forceV3000=False]]]]) -> str :\n",
      "            Returns a Mol block for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - includeStereo: (optional) toggles inclusion of stereochemical\n",
      "                                 information in the output\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - kekulize: (optional) triggers kekulization of the molecule before it's written,\n",
      "                            as suggested by the MDL spec.\n",
      "                - forceV3000 (optional) force generation a V3000 mol block (happens automatically with \n",
      "                             more than 999 atoms or bonds)\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToMolBlock(RDKit::ROMol [,bool=False [,int=-1 [,bool=True [,bool=False]]]])\n",
      "    \n",
      "    MolToMolFile(...)\n",
      "        MolToMolFile( (Mol)mol, (str)filename [, (bool)includeStereo=False [, (int)confId=-1 [, (bool)kekulize=True [, (bool)forceV3000=False]]]]) -> None :\n",
      "            Writes a Mol file for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - filename: the file to write to\n",
      "                - includeStereo: (optional) toggles inclusion of stereochemical\n",
      "                                 information in the output\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - kekulize: (optional) triggers kekulization of the molecule before it's written,\n",
      "                            as suggested by the MDL spec.\n",
      "                - forceV3000 (optional) force generation a V3000 mol block (happens automatically with \n",
      "                             more than 999 atoms or bonds)\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToMolFile(RDKit::ROMol,std::string [,bool=False [,int=-1 [,bool=True [,bool=False]]]])\n",
      "    \n",
      "    MolToPDBBlock(...)\n",
      "        MolToPDBBlock( (Mol)mol [, (int)confId=-1 [, (int)flavor=0]]) -> str :\n",
      "            Returns a PDB block for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - flavor: (optional) \n",
      "                        flavor & 1 : Write MODEL/ENDMDL lines around each record \n",
      "                        flavor & 2 : Don't write any CONECT records \n",
      "                        flavor & 4 : Write CONECT records in both directions \n",
      "                        flavor & 8 : Don't use multiple CONECTs to encode bond order \n",
      "                        flavor & 16 : Write MASTER record \n",
      "                        flavor & 32 : Write TER record \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToPDBBlock(RDKit::ROMol [,int=-1 [,unsigned int=0]])\n",
      "    \n",
      "    MolToPDBFile(...)\n",
      "        MolToPDBFile( (Mol)mol, (str)filename [, (int)confId=-1 [, (int)flavor=0]]) -> None :\n",
      "            Writes a PDB file for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - filename: name of the file to write\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - flavor: (optional) \n",
      "                        flavor & 1 : Write MODEL/ENDMDL lines around each record \n",
      "                        flavor & 2 : Don't write any CONECT records \n",
      "                        flavor & 4 : Write CONECT records in both directions \n",
      "                        flavor & 8 : Don't use multiple CONECTs to encode bond order \n",
      "                        flavor & 16 : Write MASTER record \n",
      "                        flavor & 32 : Write TER record \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToPDBFile(RDKit::ROMol,std::string [,int=-1 [,unsigned int=0]])\n",
      "    \n",
      "    MolToSVG(...)\n",
      "        MolToSVG( (Mol)arg1, (int)mol [, (int)width=300 [, (AtomPairsParameters)height=300 [, (bool)highlightAtoms=None [, (int)kekulize=True [, (int)lineWidthMult=1 [, (bool)fontSize=12 [, (int)includeAtomCircles=True]]]]]]]) -> str :\n",
      "            Returns svg for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToSVG(RDKit::ROMol,unsigned int [,unsigned int=300 [,boost::python::api::object=300 [,bool=None [,unsigned int=True [,unsigned int=1 [,bool=12 [,int=True]]]]]]])\n",
      "    \n",
      "    MolToSequence(...)\n",
      "        MolToSequence( (Mol)mol) -> str :\n",
      "            Returns the sequence string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToSequence(RDKit::ROMol)\n",
      "    \n",
      "    MolToSmarts(...)\n",
      "        MolToSmarts( (Mol)mol [, (bool)isomericSmiles=False]) -> str :\n",
      "            Returns a SMARTS string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMARTS.  Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToSmarts(RDKit::ROMol {lvalue} [,bool=False])\n",
      "    \n",
      "    MolToSmiles(...)\n",
      "        MolToSmiles( (Mol)mol [, (bool)isomericSmiles=False [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False]]]]]]) -> str :\n",
      "            Returns the canonical SMILES string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToSmiles(RDKit::ROMol [,bool=False [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False]]]]]])\n",
      "    \n",
      "    MolToTPLBlock(...)\n",
      "        MolToTPLBlock( (Mol)mol [, (str)partialChargeProp='_GasteigerCharge' [, (bool)writeFirstConfTwice=False]]) -> str :\n",
      "            Returns the Tpl block for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - partialChargeProp: name of the property to use for partial charges\n",
      "                  Defaults to '_GasteigerCharge'.\n",
      "                - writeFirstConfTwice: Defaults to False.\n",
      "                  This should be set to True when writing TPLs to be read by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::string MolToTPLBlock(RDKit::ROMol [,std::string='_GasteigerCharge' [,bool=False]])\n",
      "    \n",
      "    MolToTPLFile(...)\n",
      "        MolToTPLFile( (Mol)mol, (str)fileName [, (str)partialChargeProp='_GasteigerCharge' [, (bool)writeFirstConfTwice=False]]) -> None :\n",
      "            Writes a molecule to a TPL file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - fileName: name of the file to write\n",
      "                - partialChargeProp: name of the property to use for partial charges\n",
      "                  Defaults to '_GasteigerCharge'.\n",
      "                - writeFirstConfTwice: Defaults to False.\n",
      "                  This should be set to True when writing TPLs to be read by \n",
      "                  the CombiCode.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToTPLFile(RDKit::ROMol,std::string [,std::string='_GasteigerCharge' [,bool=False]])\n",
      "    \n",
      "    MurckoDecompose(...)\n",
      "        MurckoDecompose( (Mol)mol) -> Mol :\n",
      "            Do a Murcko decomposition and return the scaffold\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MurckoDecompose(RDKit::ROMol)\n",
      "    \n",
      "    NumRadicalElectronsEqualsQueryAtom(...)\n",
      "        NumRadicalElectronsEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where NumRadicalElectrons is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* NumRadicalElectronsEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    NumRadicalElectronsGreaterQueryAtom(...)\n",
      "        NumRadicalElectronsGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where NumRadicalElectrons is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* NumRadicalElectronsGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    NumRadicalElectronsLessQueryAtom(...)\n",
      "        NumRadicalElectronsLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where NumRadicalElectrons is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* NumRadicalElectronsLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    PEOE_VSA_(...)\n",
      "        PEOE_VSA_( (Mol)mol [, (AtomPairsParameters)bins=[] [, (bool)force=False]]) -> list :\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list PEOE_VSA_(RDKit::ROMol [,boost::python::api::object=[] [,bool=False]])\n",
      "    \n",
      "    ParseMolQueryDefFile(...)\n",
      "        ParseMolQueryDefFile( (AtomPairsParameters)fileobj [, (bool)standardize=True [, (str)delimiter='\\t' [, (str)comment='//' [, (int)nameColumn=0 [, (int)smartsColumn=1]]]]]) -> dict :\n",
      "            reads query definitions from a simply formatted file\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict ParseMolQueryDefFile(boost::python::api::object {lvalue} [,bool=True [,std::string='\\t' [,std::string='//' [,unsigned int=0 [,unsigned int=1]]]]])\n",
      "    \n",
      "    PathToSubmol(...)\n",
      "        PathToSubmol( (Mol)mol, (AtomPairsParameters)path [, (bool)useQuery=False [, (AtomPairsParameters)atomMap=None]]) -> Mol :\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* PathToSubmol(RDKit::ROMol,boost::python::api::object {lvalue} [,bool=False [,boost::python::api::object=None]])\n",
      "    \n",
      "    PatternFingerprint(...)\n",
      "        PatternFingerprint( (Mol)mol [, (int)fpSize=2048 [, (list)atomCounts=[] [, (ExplicitBitVect)setOnlyBits=None]]]) -> ExplicitBitVect :\n",
      "            A fingerprint using SMARTS patterns \n",
      "            \n",
      "              NOTE: This function is experimental. The API or results may change from\n",
      "                release to release.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* PatternFingerprint(RDKit::ROMol [,unsigned int=2048 [,boost::python::list=[] [,ExplicitBitVect*=None]]])\n",
      "    \n",
      "    PreprocessReaction(...)\n",
      "        PreprocessReaction( (ChemicalReaction)reaction [, (dict)queries={} [, (str)propName='molFileValue']]) -> object :\n",
      "            A function for preprocessing reactions with more specific queries.\n",
      "            Queries are indicated by labels on atoms (molFileAlias property by default)\n",
      "            When these labels are found, more specific queries are placed on the atoms.\n",
      "            By default, the available quieries come from \n",
      "              FilterCatalog.GetFlattenedFunctionalGroupHierarchy(True)\n",
      "            \n",
      "            \n",
      "            Sample Usage:\n",
      "              >>> from rdkit import Chem, RDConfig\n",
      "              >>> from rdkit.Chem import MolFromSmiles, AllChem\n",
      "              >>> from rdkit.Chem.rdChemReactions import PreprocessReaction\n",
      "              >>> import os\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','boronic1.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              >>> nWarn\n",
      "              0\n",
      "              >>> nError\n",
      "              0\n",
      "              >>> nReacts\n",
      "              2\n",
      "              >>> nProds\n",
      "              1\n",
      "              >>> reactantLabels\n",
      "              (((0, 'halogen.bromine.aromatic'),), ((1, 'boronicacid'),))\n",
      "            \n",
      "              If there are functional group labels in the input reaction (via atoms with molFileValue properties),\n",
      "              the corresponding atoms will have queries added to them so that they only match such things. We can\n",
      "              see this here:\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> r1 = rxn.GetReactantTemplate(0)\n",
      "              >>> m1 = Chem.MolFromSmiles('CCBr')\n",
      "              >>> m2 = Chem.MolFromSmiles('c1ccccc1Br')\n",
      "              \n",
      "              These both match because the reaction file itself just has R1-Br:\n",
      "              >>> m1.HasSubstructMatch(r1)\n",
      "              True\n",
      "              >>> m2.HasSubstructMatch(r1)\n",
      "              True\n",
      "            \n",
      "              After preprocessing, we only match the aromatic Br:\n",
      "              >>> d = PreprocessReaction(rxn)\n",
      "              >>> m1.HasSubstructMatch(r1)\n",
      "              False\n",
      "              >>> m2.HasSubstructMatch(r1)\n",
      "              True\n",
      "            \n",
      "              We also support or queries in the values field (separated by commas):\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','azide_reaction.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> reactantLabels = PreprocessReaction(rxn)[-1]\n",
      "              >>> reactantLabels\n",
      "              (((1, 'azide'),), ((1, 'carboxylicacid,acidchloride'),))\n",
      "              >>> m1 = Chem.MolFromSmiles('CC(=O)O')\n",
      "              >>> m2 = Chem.MolFromSmiles('CC(=O)Cl')\n",
      "              >>> m3 = Chem.MolFromSmiles('CC(=O)N')\n",
      "              >>> r2 = rxn.GetReactantTemplate(1)\n",
      "              >>> m1.HasSubstructMatch(r2)\n",
      "              True\n",
      "              >>> m2.HasSubstructMatch(r2)\n",
      "              True\n",
      "              >>> m3.HasSubstructMatch(r2)\n",
      "              False\n",
      "            \n",
      "              unrecognized final group types are returned as None:\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value1.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "            \n",
      "              One unrecognized group type in a comma-separated list makes the whole thing fail:\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value2.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value3.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "              >>> rxn = rdChemReactions.ChemicalReaction()\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              >>> reactantLabels\n",
      "              ()\n",
      "              >>> reactantLabels == ()\n",
      "              True\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object PreprocessReaction(RDKit::ChemicalReaction {lvalue} [,boost::python::dict={} [,std::string='molFileValue']])\n",
      "    \n",
      "    RDKFingerprint(...)\n",
      "        RDKFingerprint( (Mol)mol [, (int)minPath=1 [, (int)maxPath=7 [, (int)fpSize=2048 [, (int)nBitsPerHash=2 [, (bool)useHs=True [, (float)tgtDensity=0.0 [, (int)minSize=128 [, (bool)branchedPaths=True [, (bool)useBondOrder=True [, (AtomPairsParameters)atomInvariants=0 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)atomBits=None [, (AtomPairsParameters)bitInfo=None]]]]]]]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns an RDKit topological fingerprint for a molecule\n",
      "            \n",
      "              Explanation of the algorithm below.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                  Defaults to 7.\n",
      "            \n",
      "                - fpSize: (optional) number of bits in the fingerprint\n",
      "                  Defaults to 2048.\n",
      "            \n",
      "                - nBitsPerHash: (optional) number of bits to set per path\n",
      "                  Defaults to 2.\n",
      "            \n",
      "                - useHs: (optional) include paths involving Hs in the fingerprint if the molecule\n",
      "                  has explicit Hs.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - tgtDensity: (optional) fold the fingerprint until this minimum density has\n",
      "                  been reached\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - minSize: (optional) the minimum size the fingerprint will be folded to when\n",
      "                  trying to reach tgtDensity\n",
      "                  Defaults to 128.\n",
      "            \n",
      "                - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                  used in the fingerprint.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - useBondOrder: (optional) if set both bond orders will be used in the path hashes\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - atomInvariants: (optional) a sequence of atom invariants to use in the path hashes\n",
      "                  Defaults to empty.\n",
      "            \n",
      "                - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                  starting from these atoms will be used.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "                - atomBits: (optional) an empty list. If provided, the result will contain a list \n",
      "                  containing the bits each atom sets.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "               - bitInfo: (optional) an empty dict. If provided, the result will contain a dict \n",
      "                 with bits as keys and corresponding bond paths as values.\n",
      "                 Defaults to empty.\n",
      "            \n",
      "              RETURNS: a DataStructs.ExplicitBitVect with _fpSize_ bits\n",
      "            \n",
      "              ALGORITHM:\n",
      "            \n",
      "               This algorithm functions by find all subgraphs between minPath and maxPath in\n",
      "               length.  For each subgraph:\n",
      "            \n",
      "                 1) A hash is calculated.\n",
      "            \n",
      "                 2) The hash is used to seed a random-number generator\n",
      "            \n",
      "                 3) _nBitsPerHash_ random numbers are generated and used to set the corresponding\n",
      "                    bits in the fingerprint\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* RDKFingerprint(RDKit::ROMol [,unsigned int=1 [,unsigned int=7 [,unsigned int=2048 [,unsigned int=2 [,bool=True [,double=0.0 [,unsigned int=128 [,bool=True [,bool=True [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=None [,boost::python::api::object=None]]]]]]]]]]]]])\n",
      "    \n",
      "    RandomTransform(...)\n",
      "        RandomTransform( (Mol)mol [, (int)cid=-1 [, (int)seed=-1]]) -> None :\n",
      "            Perform a random transformation on a molecule\n",
      "                 \n",
      "                 ARGUMENTS\n",
      "                  - mol    molecule that is to be transformed\n",
      "                  - cid    ID of the conformation in the mol to be transformed\n",
      "                           (defaults to first conformation)\n",
      "                  - seed   seed used to initialize the random generator\n",
      "                           (defaults to -1, that is no seeding)\n",
      "                   \n",
      "                \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void RandomTransform(RDKit::ROMol {lvalue} [,int=-1 [,int=-1]])\n",
      "    \n",
      "    ReactionFromMolecule(...)\n",
      "        ReactionFromMolecule( (Mol)arg1) -> ChemicalReaction :\n",
      "            construct a ChemicalReaction from an molecule if the RXN role property of the molecule is set\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ChemicalReaction* ReactionFromMolecule(RDKit::ROMol)\n",
      "    \n",
      "    ReactionFromRxnBlock(...)\n",
      "        ReactionFromRxnBlock( (str)arg1) -> ChemicalReaction :\n",
      "            construct a ChemicalReaction from an string in MDL rxn format\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ChemicalReaction* ReactionFromRxnBlock(std::string)\n",
      "    \n",
      "    ReactionFromRxnFile(...)\n",
      "        ReactionFromRxnFile( (str)arg1) -> ChemicalReaction :\n",
      "            construct a ChemicalReaction from an MDL rxn file\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ChemicalReaction* ReactionFromRxnFile(std::string)\n",
      "    \n",
      "    ReactionFromSmarts(...)\n",
      "        ReactionFromSmarts( (str)SMARTS [, (dict)replacements={} [, (bool)useSmiles=False]]) -> ChemicalReaction :\n",
      "            construct a ChemicalReaction from a reaction SMARTS string. \n",
      "            see the documentation for rdkit.Chem.MolFromSmiles for an explanation\n",
      "            of the replacements argument.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ChemicalReaction* ReactionFromSmarts(char const* [,boost::python::dict={} [,bool=False]])\n",
      "    \n",
      "    ReactionToMolecule(...)\n",
      "        ReactionToMolecule( (ChemicalReaction)reaction) -> Mol :\n",
      "            construct a molecule for a ChemicalReaction with RXN role property set\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReactionToMolecule(RDKit::ChemicalReaction)\n",
      "    \n",
      "    ReactionToRxnBlock(...)\n",
      "        ReactionToRxnBlock( (ChemicalReaction)reaction [, (bool)separateAgents=False]) -> str :\n",
      "            construct a string in MDL rxn format for a ChemicalReaction\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string ReactionToRxnBlock(RDKit::ChemicalReaction [,bool=False])\n",
      "    \n",
      "    ReactionToSmarts(...)\n",
      "        ReactionToSmarts( (ChemicalReaction)reaction) -> str :\n",
      "            construct a reaction SMARTS string for a ChemicalReaction\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string ReactionToSmarts(RDKit::ChemicalReaction)\n",
      "    \n",
      "    ReactionToSmiles(...)\n",
      "        ReactionToSmiles( (ChemicalReaction)reaction [, (bool)canonical=True]) -> str :\n",
      "            construct a reaction SMILES string for a ChemicalReaction\n",
      "        \n",
      "            C++ signature :\n",
      "                std::string ReactionToSmiles(RDKit::ChemicalReaction [,bool=True])\n",
      "    \n",
      "    ReduceProductToSideChains(...)\n",
      "        ReduceProductToSideChains( (Mol)product [, (bool)addDummyAtoms=True]) -> Mol :\n",
      "            reduce the product of a reaction to the side chains added by the reaction.              The output is a molecule with attached wildcards indicating where the product was attached.              The dummy atom has the same reaction-map number as the product atom (if available).\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReduceProductToSideChains(boost::shared_ptr<RDKit::ROMol> [,bool=True])\n",
      "    \n",
      "    RemoveHs(...)\n",
      "        RemoveHs( (Mol)mol [, (bool)implicitOnly=False [, (bool)updateExplicitCount=False [, (bool)sanitize=True]]]) -> Mol :\n",
      "            Removes any hydrogens from the graph of a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - implicitOnly: (optional) if this toggle is set, only implicit Hs will\n",
      "                  be removed from the graph.  Default value is 0 (remove implicit and explicit Hs).\n",
      "            \n",
      "                - updateExplicitCount: (optional) if this toggle is set, the explicit H count on atoms with \n",
      "                  Hs will be updated. Default value is 0 (do not update explicit H count).\n",
      "            \n",
      "                - sanitize: (optional) if this toggle is set, the molecule will be sanitized after the Hs\n",
      "                  are removed. Default value is 1 (do sanitize).\n",
      "            \n",
      "              RETURNS: a new molecule with the Hs removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* RemoveHs(RDKit::ROMol [,bool=False [,bool=False [,bool=True]]])\n",
      "    \n",
      "    RemoveMappingNumbersFromReactions(...)\n",
      "        RemoveMappingNumbersFromReactions( (ChemicalReaction)reaction) -> None :\n",
      "            Removes the mapping numbers from the molecules of a reaction\n",
      "        \n",
      "            C++ signature :\n",
      "                void RemoveMappingNumbersFromReactions(RDKit::ChemicalReaction)\n",
      "    \n",
      "    RemoveStereochemistry(...)\n",
      "        RemoveStereochemistry( (Mol)mol) -> None :\n",
      "            Removes all stereochemistry info from the molecule.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void RemoveStereochemistry(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    RenumberAtoms(...)\n",
      "        RenumberAtoms( (Mol)mol, (AtomPairsParameters)newOrder) -> Mol :\n",
      "            Returns a copy of a molecule with renumbered atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - newOrder: the new ordering the atoms (should be numAtoms long)\n",
      "                     for example: if newOrder is [3,2,0,1], then atom 3 in the original \n",
      "                     molecule will be atom 0 in the new one\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* RenumberAtoms(RDKit::ROMol,boost::python::api::object {lvalue})\n",
      "    \n",
      "    ReplaceCore(...)\n",
      "        ReplaceCore( (Mol)mol, (Mol)core, (AtomPairsParameters)matches [, (bool)replaceDummies=True [, (bool)labelByIndex=False [, (bool)requireDummyMatch=False]]]) -> Mol :\n",
      "            Removes the core of a molecule and labels the sidechains with dummy atoms based on\n",
      "            The matches indices given in the matching vector matches.\n",
      "            Calling:\n",
      "              ReplaceCore(mol,core,mol.GetSubstructMatch(core))\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - matches: a matching vector of the type returned by mol.GetSubstructMatch(...)\n",
      "            \n",
      "                - replaceDummies: toggles replacement of atoms that match dummies in the query\n",
      "            \n",
      "                - labelByIndex: toggles labeling the attachment point dummy atoms with \n",
      "                                the index of the core atom they're attached to.\n",
      "            \n",
      "                - requireDummyMatch: if the molecule has side chains that attach at points not\n",
      "                                     flagged with a dummy, it will be rejected (None is returned)\n",
      "            \n",
      "              RETURNS: a new molecule with the core removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            EXAMPLES:\n",
      "                >>> from rdkit.Chem import MolToSmiles, MolFromSmiles, ReplaceCore\n",
      "                >>> mol = MolFromSmiles('C1ONNCC1')\n",
      "                >>> core = MolFromSmiles('NN')\n",
      "            \n",
      "                Note: Using isomericSmiles is necessary to see the labels.\n",
      "                >>> MolToSmiles(ReplaceCore(mol, core, mol.GetSubstructMatch(core)), isomericSmiles=True)\n",
      "                '[1*]OCCC[2*]'\n",
      "            \n",
      "                Since NN is symmetric, we should actually get two matches here if we don't\n",
      "                uniquify the matches.\n",
      "                >>> [MolToSmiles(ReplaceCore(mol, core, match), isomericSmiles=True)\n",
      "                ...     for match in mol.GetSubstructMatches(core, uniquify=False)]\n",
      "                ['[1*]OCCC[2*]', '[1*]CCCO[2*]']\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceCore(RDKit::ROMol,RDKit::ROMol,boost::python::api::object [,bool=True [,bool=False [,bool=False]]])\n",
      "        \n",
      "        ReplaceCore( (Mol)mol, (Mol)coreQuery [, (bool)replaceDummies=True [, (bool)labelByIndex=False [, (bool)requireDummyMatch=False [, (bool)useChirality=False]]]]) -> Mol :\n",
      "            Removes the core of a molecule and labels the sidechains with dummy atoms.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - replaceDummies: toggles replacement of atoms that match dummies in the query\n",
      "            \n",
      "                - labelByIndex: toggles labeling the attachment point dummy atoms with \n",
      "                                the index of the core atom they're attached to.\n",
      "            \n",
      "                - requireDummyMatch: if the molecule has side chains that attach at points not\n",
      "                                     flagged with a dummy, it will be rejected (None is returned)\n",
      "            \n",
      "                - useChirality: use chirality matching in the coreQuery\n",
      "            \n",
      "              RETURNS: a new molecule with the core removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               >>> from rdkit.Chem import MolToSmiles, MolFromSmiles, MolFromSmarts, ReplaceCore\n",
      "            \n",
      "               Basic usage: remove a core as specified by SMILES (or another molecule).\n",
      "               To get the atom labels which are stored as an isotope of the matched atom, \n",
      "               the output must be written as isomeric smiles.  \n",
      "               A small confusion is that atom isotopes of 0 aren't shown in smiles strings.\n",
      "            \n",
      "               Here we remove a ring and leave the decoration (r-group) behind.\n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('CCCC1CCC1'),MolFromSmiles('C1CCC1')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCC'\n",
      "            \n",
      "               The isotope label by default is matched by the first connection found. In order to\n",
      "               indicate which atom the decoration is attached in the core query, use labelByIndex=True.\n",
      "               Here the attachment is from the third atom in the smiles string, which is indexed by 3\n",
      "               in the core, like all good computer scientists expect, atoms indices start at 0.\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('CCN1CCC1'),MolFromSmiles('C1CCN1'),\n",
      "               ...                         labelByIndex=True),\n",
      "               ...   isomericSmiles=True)\n",
      "               '[3*]CC'\n",
      "            \n",
      "               Non-core matches just return None\n",
      "               >>> ReplaceCore(MolFromSmiles('CCC1CC1'),MolFromSmiles('C1CCC1'))\n",
      "            \n",
      "               The bond between atoms are considered part of the core and are removed as well\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CC2C1CCC2'),MolFromSmiles('C1CCC1')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCC[2*]'\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmiles('N')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCCC[2*]'\n",
      "            \n",
      "               When using dummy atoms, cores should be read in as SMARTS.  When read as SMILES\n",
      "               dummy atoms only match other dummy atoms.\n",
      "               The replaceDummies flag indicates whether matches to the dummy atoms should be considered as part\n",
      "               of the core or as part of the decoration (r-group)\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmarts('[*]N[*]'),\n",
      "               ...                         replaceDummies=True),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CC[2*]'\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmarts('[*]N[*]'),\n",
      "               ...                         replaceDummies=False),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCCC[2*]'\n",
      "            \n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CCC1CN'),MolFromSmarts('C1CCC1[*]'),\n",
      "               ...                         replaceDummies=False),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CN'\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceCore(RDKit::ROMol,RDKit::ROMol [,bool=True [,bool=False [,bool=False [,bool=False]]]])\n",
      "    \n",
      "    ReplaceSidechains(...)\n",
      "        ReplaceSidechains( (Mol)mol, (Mol)coreQuery [, (bool)useChirality=False]) -> Mol :\n",
      "            Replaces sidechains in a molecule with dummy atoms for their attachment points.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a new molecule with the sidechains removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - ReplaceSidechains('CCC1CCC1','C1CCC1') -> '[Xa]C1CCC1'\n",
      "            \n",
      "                - ReplaceSidechains('CCC1CC1','C1CCC1') -> ''\n",
      "            \n",
      "                - ReplaceSidechains('C1CC2C1CCC2','C1CCC1') -> '[Xa]C1CCC1[Xb]'\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceSidechains(RDKit::ROMol,RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    ReplaceSubstructs(...)\n",
      "        ReplaceSubstructs( (Mol)mol, (Mol)query, (Mol)replacement [, (bool)replaceAll=False [, (int)replacementConnectionPoint=0 [, (bool)useChirality=False]]]) -> object :\n",
      "            Replaces atoms matching a substructure query in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as a substructure query\n",
      "            \n",
      "                - replacement: the molecule to be used as the replacement\n",
      "            \n",
      "                - replaceAll: (optional) if this toggle is set, all substructures matching\n",
      "                  the query will be replaced in a single result, otherwise each result will\n",
      "                  contain a separate replacement.\n",
      "                  Default value is False (return multiple replacements)\n",
      "                - replacementConnectionPoint: (optional) index of the atom in the replacement that\n",
      "                  the bond should be made to.\n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a tuple of new molecules with the substructures replaced removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - ReplaceSubstructs('CCOC','OC','NC') -> ('CCNC',)\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','NC') -> ('COCCNC','CNCCOC')\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','NC',True) -> ('CNCCNC',)\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','CN',True,1) -> ('CNCCNC',)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* ReplaceSubstructs(RDKit::ROMol,RDKit::ROMol,RDKit::ROMol [,bool=False [,unsigned int=0 [,bool=False]]])\n",
      "    \n",
      "    RingBondCountEqualsQueryAtom(...)\n",
      "        RingBondCountEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where RingBondCount is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* RingBondCountEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    RingBondCountGreaterQueryAtom(...)\n",
      "        RingBondCountGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where RingBondCount is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* RingBondCountGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    RingBondCountLessQueryAtom(...)\n",
      "        RingBondCountLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where RingBondCount is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* RingBondCountLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    SMR_VSA_(...)\n",
      "        SMR_VSA_( (Mol)mol [, (AtomPairsParameters)bins=[] [, (bool)force=False]]) -> list :\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list SMR_VSA_(RDKit::ROMol [,boost::python::api::object=[] [,bool=False]])\n",
      "    \n",
      "    SanitizeMol(...)\n",
      "        SanitizeMol( (Mol)mol [, (int)sanitizeOps=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL [, (bool)catchErrors=False]]) -> SanitizeFlags :\n",
      "            Kekulize, check valencies, set aromaticity, conjugation and hybridization\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "                - If sanitization fails, an exception will be thrown unless catchErrors is set\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "                - sanitizeOps: (optional) sanitization operations to be carried out\n",
      "                               these should be constructed by or'ing together the\n",
      "                               operations in rdkit.Chem.SanitizeFlags\n",
      "                - catchErrors: (optional) if provided, instead of raising an exception\n",
      "                               when sanitization fails (the default behavior), the \n",
      "                               first operation that failed (as defined in rdkit.Chem.SanitizeFlags)\n",
      "                               is returned. Zero is returned on success.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolOps::SanitizeFlags SanitizeMol(RDKit::ROMol {lvalue} [,unsigned long=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL [,bool=False]])\n",
      "    \n",
      "    SanitizeRxn(...)\n",
      "        SanitizeRxn( (ChemicalReaction)rxn [, (int)sanitizeOps=4294967295 [, (AdjustQueryParameters)params=<rdkit.Chem.rdmolops.AdjustQueryParameters object at 0x7fdd834cc030> [, (bool)catchErrors=False]]]) -> SanitizeFlags :\n",
      "            A function for preprocessing reactions with more specific queries.\n",
      "            Queries are indicated by labels on atoms (molFileAlias property by default)\n",
      "            When these labels are found, more specific queries are placed on the atoms.\n",
      "            By default, the available quieries come from \n",
      "              FilterCatalog.GetFlattenedFunctionalGroupHierarchy(True)\n",
      "            \n",
      "            \n",
      "            Sample Usage:\n",
      "              >>> from rdkit import Chem, RDConfig\n",
      "              >>> from rdkit.Chem import MolFromSmiles, AllChem\n",
      "              >>> from rdkit.Chem.rdChemReactions import PreprocessReaction\n",
      "              >>> import os\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','boronic1.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              >>> nWarn\n",
      "              0\n",
      "              >>> nError\n",
      "              0\n",
      "              >>> nReacts\n",
      "              2\n",
      "              >>> nProds\n",
      "              1\n",
      "              >>> reactantLabels\n",
      "              (((0, 'halogen.bromine.aromatic'),), ((1, 'boronicacid'),))\n",
      "            \n",
      "              If there are functional group labels in the input reaction (via atoms with molFileValue properties),\n",
      "              the corresponding atoms will have queries added to them so that they only match such things. We can\n",
      "              see this here:\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> r1 = rxn.GetReactantTemplate(0)\n",
      "              >>> m1 = Chem.MolFromSmiles('CCBr')\n",
      "              >>> m2 = Chem.MolFromSmiles('c1ccccc1Br')\n",
      "              \n",
      "              These both match because the reaction file itself just has R1-Br:\n",
      "              >>> m1.HasSubstructMatch(r1)\n",
      "              True\n",
      "              >>> m2.HasSubstructMatch(r1)\n",
      "              True\n",
      "            \n",
      "              After preprocessing, we only match the aromatic Br:\n",
      "              >>> d = PreprocessReaction(rxn)\n",
      "              >>> m1.HasSubstructMatch(r1)\n",
      "              False\n",
      "              >>> m2.HasSubstructMatch(r1)\n",
      "              True\n",
      "            \n",
      "              We also support or queries in the values field (separated by commas):\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','azide_reaction.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> reactantLabels = PreprocessReaction(rxn)[-1]\n",
      "              >>> reactantLabels\n",
      "              (((1, 'azide'),), ((1, 'carboxylicacid,acidchloride'),))\n",
      "              >>> m1 = Chem.MolFromSmiles('CC(=O)O')\n",
      "              >>> m2 = Chem.MolFromSmiles('CC(=O)Cl')\n",
      "              >>> m3 = Chem.MolFromSmiles('CC(=O)N')\n",
      "              >>> r2 = rxn.GetReactantTemplate(1)\n",
      "              >>> m1.HasSubstructMatch(r2)\n",
      "              True\n",
      "              >>> m2.HasSubstructMatch(r2)\n",
      "              True\n",
      "              >>> m3.HasSubstructMatch(r2)\n",
      "              False\n",
      "            \n",
      "              unrecognized final group types are returned as None:\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value1.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "            \n",
      "              One unrecognized group type in a comma-separated list makes the whole thing fail:\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value2.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "              >>> testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','bad_value3.rxn')\n",
      "              >>> rxn = AllChem.ReactionFromRxnFile(testFile)\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              Traceback (most recent call last):\n",
      "                ...\n",
      "              RuntimeError: KeyErrorException\n",
      "              >>> rxn = rdChemReactions.ChemicalReaction()\n",
      "              >>> rxn.Initialize()\n",
      "              >>> nWarn,nError,nReacts,nProds,reactantLabels = PreprocessReaction(rxn)\n",
      "              >>> reactantLabels\n",
      "              ()\n",
      "              >>> reactantLabels == ()\n",
      "              True\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::RxnOps::SanitizeRxnFlags SanitizeRxn(RDKit::ChemicalReaction {lvalue} [,unsigned long=4294967295 [,RDKit::MolOps::AdjustQueryParameters=<rdkit.Chem.rdmolops.AdjustQueryParameters object at 0x7fdd834cc030> [,bool=False]]])\n",
      "    \n",
      "    SetAngleDeg(...)\n",
      "        SetAngleDeg( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId, (float)value) -> None :\n",
      "            Sets the angle in degrees between atoms i, j, k; all atoms bonded to atom k are moved\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAngleDeg(RDKit::Conformer {lvalue},unsigned int,unsigned int,unsigned int,double)\n",
      "    \n",
      "    SetAngleRad(...)\n",
      "        SetAngleRad( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId, (float)value) -> None :\n",
      "            Sets the angle in radians between atoms i, j, k; all atoms bonded to atom k are moved\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAngleRad(RDKit::Conformer {lvalue},unsigned int,unsigned int,unsigned int,double)\n",
      "    \n",
      "    SetAromaticity(...)\n",
      "        SetAromaticity( (Mol)mol [, (AromaticityModel)model=rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_DEFAULT]) -> None :\n",
      "            does aromaticity perception\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - model: the model to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAromaticity(RDKit::ROMol {lvalue} [,RDKit::MolOps::AromaticityModel=rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_DEFAULT])\n",
      "    \n",
      "    SetAtomAlias(...)\n",
      "        SetAtomAlias( (Atom)atom, (str)rlabel) -> None :\n",
      "            Sets the atom's MDL alias text.\n",
      "            Setting to an empty string clears the alias.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomAlias(RDKit::Atom*,std::string)\n",
      "    \n",
      "    SetAtomRLabel(...)\n",
      "        SetAtomRLabel( (Atom)atom, (int)rlabel) -> None :\n",
      "            Sets the atom's MDL RLabel (this is an integer from 0 to 99).\n",
      "            Setting to 0 clears the rlabel.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomRLabel(RDKit::Atom*,int)\n",
      "    \n",
      "    SetAtomValue(...)\n",
      "        SetAtomValue( (Atom)atom, (str)rlabel) -> None :\n",
      "            Sets the atom's MDL alias text.\n",
      "            Setting to an empty string clears the alias.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomValue(RDKit::Atom*,std::string)\n",
      "    \n",
      "    SetBondLength(...)\n",
      "        SetBondLength( (Conformer)conf, (int)iAtomId, (int)jAtomId, (float)value) -> None :\n",
      "            Sets the bond length in angstrom between atoms i, j; all atoms bonded to atom j are moved\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetBondLength(RDKit::Conformer {lvalue},unsigned int,unsigned int,double)\n",
      "    \n",
      "    SetConjugation(...)\n",
      "        SetConjugation( (Mol)mol) -> None :\n",
      "            finds conjugated bonds\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetConjugation(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    SetDefaultPickleProperties(...)\n",
      "        SetDefaultPickleProperties( (int)arg1) -> None :\n",
      "            Set the current global mol pickler options.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetDefaultPickleProperties(unsigned int)\n",
      "    \n",
      "    SetDihedralDeg(...)\n",
      "        SetDihedralDeg( (Conformer)arg1, (int)arg2, (int)arg3, (int)arg4, (int)arg5, (float)arg6) -> None :\n",
      "            Sets the dihedral angle in degrees between atoms i, j, k, l; all atoms bonded to atom l are moved\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetDihedralDeg(RDKit::Conformer {lvalue},unsigned int,unsigned int,unsigned int,unsigned int,double)\n",
      "    \n",
      "    SetDihedralRad(...)\n",
      "        SetDihedralRad( (Conformer)conf, (int)iAtomId, (int)jAtomId, (int)kAtomId, (int)lAtomId, (float)value) -> None :\n",
      "            Sets the dihedral angle in radians between atoms i, j, k, l; all atoms bonded to atom l are moved\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetDihedralRad(RDKit::Conformer {lvalue},unsigned int,unsigned int,unsigned int,unsigned int,double)\n",
      "    \n",
      "    SetHybridization(...)\n",
      "        SetHybridization( (Mol)mol) -> None :\n",
      "            Assigns hybridization states to atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetHybridization(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    SetSupplementalSmilesLabel(...)\n",
      "        SetSupplementalSmilesLabel( (Atom)atom, (str)label) -> None :\n",
      "            Sets a supplemental label on an atom that is written to the smiles string.\n",
      "            >>> m = Chem.MolFromSmiles(\"C\")\n",
      "            >>> Chem.SetSupplementalSmilesLabel(m.GetAtomWithIdx(0), '<xxx>')\n",
      "            >>> Chem.MolToSmiles(m)\n",
      "            'C<xxx>'\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetSupplementalSmilesLabel(RDKit::Atom*,std::string)\n",
      "    \n",
      "    ShapeProtrudeDist(...)\n",
      "        ShapeProtrudeDist( (Mol)mol1, (Mol)mol2 [, (int)confId1=-1 [, (int)confId2=-1 [, (float)gridSpacing=0.5 [, (DiscreteValueType)bitsPerPoint=rdkit.DataStructs.cDataStructs.DiscreteValueType.TWOBITVALUE [, (float)vdwScale=0.8 [, (float)stepSize=0.25 [, (int)maxLayers=-1 [, (bool)ignoreHs=True [, (bool)allowReordering=True]]]]]]]]]) -> float :\n",
      "            Compute the shape protrude distance between two molecule based on a predefined alignment\n",
      "              \n",
      "              ARGUMENTS:\n",
      "                - mol1 : The first molecule of interest \n",
      "                - mol2 : The second molecule of interest \n",
      "                - confId1 : Conformer in the first molecule (defaults to first conformer) \n",
      "                - confId2 : Conformer in the second molecule (defaults to first conformer) \n",
      "                - gridSpacing : resolution of the grid used to encode the molecular shapes \n",
      "                - bitsPerPoint : number of bit used to encode the occupancy at each grid point \n",
      "                                      defaults to two bits per grid point \n",
      "                - vdwScale : Scaling factor for the radius of the atoms to determine the base radius \n",
      "                            used in the encoding - grid points inside this sphere carry the maximum occupancy \n",
      "                - stepSize : thickness of the each layer outside the base radius, the occupancy value is decreased \n",
      "                             from layer to layer from the maximum value \n",
      "                - maxLayers : the maximum number of layers - defaults to the number of bits \n",
      "                              used per grid point - e.g. two bits per grid point will allow 3 layers \n",
      "                - ignoreHs : when set, the contribution of Hs to the shape will be ignored\n",
      "                - allowReordering : when set, the order will be automatically updated so that the value calculated\n",
      "                                    is the protrusion of the smaller shape from the larger one.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double ShapeProtrudeDist(RDKit::ROMol,RDKit::ROMol [,int=-1 [,int=-1 [,double=0.5 [,RDKit::DiscreteValueVect::DiscreteValueType=rdkit.DataStructs.cDataStructs.DiscreteValueType.TWOBITVALUE [,double=0.8 [,double=0.25 [,int=-1 [,bool=True [,bool=True]]]]]]]]])\n",
      "    \n",
      "    ShapeTanimotoDist(...)\n",
      "        ShapeTanimotoDist( (Mol)mol1, (Mol)mol2 [, (int)confId1=-1 [, (int)confId2=-1 [, (float)gridSpacing=0.5 [, (DiscreteValueType)bitsPerPoint=rdkit.DataStructs.cDataStructs.DiscreteValueType.TWOBITVALUE [, (float)vdwScale=0.8 [, (float)stepSize=0.25 [, (int)maxLayers=-1 [, (bool)ignoreHs=True]]]]]]]]) -> float :\n",
      "            Compute the shape tanimoto distance between two molecule based on a predefined alignment\n",
      "              \n",
      "              ARGUMENTS:\n",
      "                - mol1 : The first molecule of interest \n",
      "                - mol2 : The second molecule of interest \n",
      "                - confId1 : Conformer in the first molecule (defaults to first conformer) \n",
      "                - confId2 : Conformer in the second molecule (defaults to first conformer) \n",
      "                - gridSpacing : resolution of the grid used to encode the molecular shapes \n",
      "                - bitsPerPoint : number of bits used to encode the occupancy at each grid point \n",
      "                                      defaults to two bits per grid point \n",
      "                - vdwScale : Scaling factor for the radius of the atoms to determine the base radius \n",
      "                            used in the encoding - grid points inside this sphere carry the maximum occupancy \n",
      "                - stepSize : thickness of the each layer outside the base radius, the occupancy value is decreased \n",
      "                             from layer to layer from the maximum value \n",
      "                - maxLayers : the maximum number of layers - defaults to the number of bits \n",
      "                              used per grid point - e.g. two bits per grid point will allow 3 layers \n",
      "                - ignoreHs : when set, the contribution of Hs to the shape will be ignored\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                double ShapeTanimotoDist(RDKit::ROMol,RDKit::ROMol [,int=-1 [,int=-1 [,double=0.5 [,RDKit::DiscreteValueVect::DiscreteValueType=rdkit.DataStructs.cDataStructs.DiscreteValueType.TWOBITVALUE [,double=0.8 [,double=0.25 [,int=-1 [,bool=True]]]]]]]])\n",
      "    \n",
      "    SlogP_VSA_(...)\n",
      "        SlogP_VSA_( (Mol)mol [, (AtomPairsParameters)bins=[] [, (bool)force=False]]) -> list :\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::list SlogP_VSA_(RDKit::ROMol [,boost::python::api::object=[] [,bool=False]])\n",
      "    \n",
      "    SmilesMolSupplierFromText(...)\n",
      "        SmilesMolSupplierFromText( (str)text [, (str)delimiter=' ' [, (int)smilesColumn=0 [, (int)nameColumn=1 [, (bool)titleLine=True [, (bool)sanitize=True]]]]]) -> SmilesMolSupplier :\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SmilesMolSupplier* SmilesMolSupplierFromText(std::string [,std::string=' ' [,int=0 [,int=1 [,bool=True [,bool=True]]]]])\n",
      "    \n",
      "    SplitMolByPDBChainId(...)\n",
      "        SplitMolByPDBChainId( (Mol)mol [, (AtomPairsParameters)whiteList=None [, (bool)negateList=False]]) -> dict :\n",
      "            Splits a molecule into pieces based on PDB chain information.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - whiteList: only residues in this list will be returned\n",
      "                - negateList: if set, negates the white list inclusion logic\n",
      "            \n",
      "              RETURNS: a dictionary keyed by chain id with molecules as the values\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict SplitMolByPDBChainId(RDKit::ROMol [,boost::python::api::object=None [,bool=False]])\n",
      "    \n",
      "    SplitMolByPDBResidues(...)\n",
      "        SplitMolByPDBResidues( (Mol)mol [, (AtomPairsParameters)whiteList=None [, (bool)negateList=False]]) -> dict :\n",
      "            Splits a molecule into pieces based on PDB residue information.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - whiteList: only residues in this list will be returned\n",
      "                - negateList: if set, negates the white list inclusion logic\n",
      "            \n",
      "              RETURNS: a dictionary keyed by residue name with molecules as the values\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict SplitMolByPDBResidues(RDKit::ROMol [,boost::python::api::object=None [,bool=False]])\n",
      "    \n",
      "    TotalDegreeEqualsQueryAtom(...)\n",
      "        TotalDegreeEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalDegree is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalDegreeEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TotalDegreeGreaterQueryAtom(...)\n",
      "        TotalDegreeGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalDegree is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalDegreeGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TotalDegreeLessQueryAtom(...)\n",
      "        TotalDegreeLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalDegree is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalDegreeLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TotalValenceEqualsQueryAtom(...)\n",
      "        TotalValenceEqualsQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalValence is equal to the target value.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalValenceEqualsQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TotalValenceGreaterQueryAtom(...)\n",
      "        TotalValenceGreaterQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalValence is equal to the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalValenceGreaterQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TotalValenceLessQueryAtom(...)\n",
      "        TotalValenceLessQueryAtom( (int)val [, (bool)negate=False]) -> QueryAtom :\n",
      "            Returns a QueryAtom that matches atoms where TotalValence is less than the target value.\n",
      "            NOTE: the direction of comparison is reversed relative to the C++ API\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::QueryAtom* TotalValenceLessQueryAtom(int [,bool=False])\n",
      "    \n",
      "    TransformConformer(...)\n",
      "        TransformConformer( (Conformer)arg1, (AtomPairsParameters)arg2) -> None :\n",
      "            Transform the coordinates of a conformer\n",
      "        \n",
      "            C++ signature :\n",
      "                void TransformConformer(RDKit::Conformer {lvalue},boost::python::api::object)\n",
      "    \n",
      "    TransformMol(mol, tform, confId=-1, keepConfs=False)\n",
      "        Applies the transformation (usually a 4x4 double matrix) to a molecule\n",
      "        if keepConfs is False then all but that conformer are removed\n",
      "    \n",
      "    UFFGetMoleculeForceField(...)\n",
      "        UFFGetMoleculeForceField( (Mol)mol [, (float)vdwThresh=10.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]) -> ForceField :\n",
      "            returns a UFF force field for a molecule\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - vdwThresh : used to exclude long-range van der Waals interactions\n",
      "                              (defaults to 10.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                              fragments will not be added to the forcefield.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ForceFields::PyForceField* UFFGetMoleculeForceField(RDKit::ROMol {lvalue} [,double=10.0 [,int=-1 [,bool=True]]])\n",
      "    \n",
      "    UFFHasAllMoleculeParams(...)\n",
      "        UFFHasAllMoleculeParams( (Mol)mol) -> bool :\n",
      "            checks if UFF parameters are available for all of a molecule's atoms\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                bool UFFHasAllMoleculeParams(RDKit::ROMol)\n",
      "    \n",
      "    UFFOptimizeMolecule(...)\n",
      "        UFFOptimizeMolecule( (Mol)self [, (int)maxIters=200 [, (float)vdwThresh=10.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]]) -> int :\n",
      "            uses UFF to optimize a molecule's structure\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - maxIters : the maximum number of iterations (defaults to 200)\n",
      "                - vdwThresh : used to exclude long-range van der Waals interactions\n",
      "                              (defaults to 10.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                              fragments will not be added to the forcefield.\n",
      "            \n",
      "             RETURNS: 0 if the optimization converged, 1 if more iterations are required.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int UFFOptimizeMolecule(RDKit::ROMol {lvalue} [,int=200 [,double=10.0 [,int=-1 [,bool=True]]]])\n",
      "    \n",
      "    UFFOptimizeMoleculeConfs(...)\n",
      "        UFFOptimizeMoleculeConfs( (Mol)self [, (int)numThreads=1 [, (int)maxIters=200 [, (float)vdwThresh=10.0 [, (int)confId=-1 [, (bool)ignoreInterfragInteractions=True]]]]]) -> object :\n",
      "            uses UFF to optimize all of a molecule's conformations\n",
      "            \n",
      "             \n",
      "             ARGUMENTS:\n",
      "            \n",
      "                - mol : the molecule of interest\n",
      "                - numThreads : the number of threads to use, only has an effect if the RDKit\n",
      "                               was built with thread support (defaults to 1)\n",
      "                               If set to zero, the max supported by the system will be used.\n",
      "                - maxIters : the maximum number of iterations (defaults to 200)\n",
      "                - vdwThresh : used to exclude long-range van der Waals interactions\n",
      "                              (defaults to 10.0)\n",
      "                - confId : indicates which conformer to optimize\n",
      "                - ignoreInterfragInteractions : if true, nonbonded terms between\n",
      "                              fragments will not be added to the forcefield.\n",
      "            \n",
      "             RETURNS: 0 if the optimization converged, 1 if more iterations are required.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object UFFOptimizeMoleculeConfs(RDKit::ROMol {lvalue} [,int=1 [,int=200 [,double=10.0 [,int=-1 [,bool=True]]]]])\n",
      "    \n",
      "    UnfoldedRDKFingerprintCountBased(...)\n",
      "        UnfoldedRDKFingerprintCountBased( (Mol)mol [, (int)minPath=1 [, (int)maxPath=7 [, (bool)useHs=True [, (bool)branchedPaths=True [, (bool)useBondOrder=True [, (AtomPairsParameters)atomInvariants=0 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)atomBits=None [, (AtomPairsParameters)bitInfo=None]]]]]]]]]) -> ULongSparseIntVect :\n",
      "            Returns an unfolded count-based version of the RDKit fingerprint for a molecule\n",
      "            \n",
      "            ARGUMENTS:\n",
      "                \n",
      "                    - mol: the molecule to use\n",
      "                \n",
      "                    - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                      Defaults to 1.\n",
      "                \n",
      "                    - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                      Defaults to 7.\n",
      "                \n",
      "                    - useHs: (optional) include paths involving Hs in the fingerprint if the molecule\n",
      "                      has explicit Hs.\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                      used in the fingerprint.\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - useBondOrder: (optional) if set both bond orders will be used in the path hashes\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - atomInvariants: (optional) a sequence of atom invariants to use in the path hashes\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                    - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                      starting from these atoms will be used.\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                    - atomBits: (optional) an empty list. If provided, the result will contain a list \n",
      "                      containing the bits each atom sets.\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                   - bitInfo: (optional) an empty dict. If provided, the result will contain a dict \n",
      "                     with bits as keys and corresponding bond paths as values.\n",
      "                     Defaults to empty.\n",
      "                 \n",
      "                 \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<unsigned long>* UnfoldedRDKFingerprintCountBased(RDKit::ROMol [,unsigned int=1 [,unsigned int=7 [,bool=True [,bool=True [,bool=True [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=None [,boost::python::api::object=None]]]]]]]]])\n",
      "    \n",
      "    UpdateProductsStereochemistry(...)\n",
      "        UpdateProductsStereochemistry( (ChemicalReaction)reaction) -> None :\n",
      "            Caution: This is an expert-user function which will change a property (molInversionFlag) of your products.          This function is called by default using the RXN or SMARTS parser for reactions and should really only be called if reactions have been constructed some other way.          The function updates the stereochemistry of the product by considering 4 different cases: inversion, retention, removal, and introduction\n",
      "        \n",
      "            C++ signature :\n",
      "                void UpdateProductsStereochemistry(RDKit::ChemicalReaction*)\n",
      "    \n",
      "    WedgeMolBonds(...)\n",
      "        WedgeMolBonds( (Mol)arg1, (Conformer)arg2) -> None :\n",
      "            Set the wedging on single bonds in a molecule.\n",
      "               The wedging scheme used is that from Mol files.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molecule: the molecule to update\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void WedgeMolBonds(RDKit::ROMol {lvalue},RDKit::Conformer const*)\n",
      "    \n",
      "    WrapLogs(...)\n",
      "        WrapLogs() -> None :\n",
      "            Wrap the internal RDKit streams so they go to python's SysStdErr\n",
      "        \n",
      "            C++ signature :\n",
      "                void WrapLogs()\n",
      "    \n",
      "    tossit(...)\n",
      "        tossit() -> None :\n",
      "        \n",
      "            C++ signature :\n",
      "                void tossit()\n",
      "\n",
      "DATA\n",
      "    ADJUST_IGNOREALL = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_IG...\n",
      "    ADJUST_IGNORECHAINS = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST...\n",
      "    ADJUST_IGNOREDUMMIES = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUS...\n",
      "    ADJUST_IGNORENONDUMMIES = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.AD...\n",
      "    ADJUST_IGNORENONE = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_I...\n",
      "    ADJUST_IGNORERINGS = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_...\n",
      "    ALLOW_CHARGE_SEPARATION = rdkit.Chem.rdchem.ResonanceFlags.ALLOW_CHARG...\n",
      "    ALLOW_INCOMPLETE_OCTETS = rdkit.Chem.rdchem.ResonanceFlags.ALLOW_INCOM...\n",
      "    AROMATICITY_CUSTOM = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_...\n",
      "    AROMATICITY_DEFAULT = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY...\n",
      "    AROMATICITY_RDKIT = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_R...\n",
      "    AROMATICITY_SIMPLE = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_...\n",
      "    AllProps = rdkit.Chem.rdchem.PropertyPickleOptions.AllProps\n",
      "    AtomProps = rdkit.Chem.rdchem.PropertyPickleOptions.AtomProps\n",
      "    BondProps = rdkit.Chem.rdchem.PropertyPickleOptions.BondProps\n",
      "    CHI_OTHER = rdkit.Chem.rdchem.ChiralType.CHI_OTHER\n",
      "    CHI_TETRAHEDRAL_CCW = rdkit.Chem.rdchem.ChiralType.CHI_TETRAHEDRAL_CCW\n",
      "    CHI_TETRAHEDRAL_CW = rdkit.Chem.rdchem.ChiralType.CHI_TETRAHEDRAL_CW\n",
      "    CHI_UNSPECIFIED = rdkit.Chem.rdchem.ChiralType.CHI_UNSPECIFIED\n",
      "    COMPOSITE_AND = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_AND\n",
      "    COMPOSITE_OR = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_OR\n",
      "    COMPOSITE_XOR = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_XOR\n",
      "    ComputedProps = rdkit.Chem.rdchem.PropertyPickleOptions.ComputedProps\n",
      "    INCHI_AVAILABLE = True\n",
      "    KEKULE_ALL = rdkit.Chem.rdchem.ResonanceFlags.KEKULE_ALL\n",
      "    LayeredFingerprint_substructLayers = 7\n",
      "    MolProps = rdkit.Chem.rdchem.PropertyPickleOptions.MolProps\n",
      "    NoProps = rdkit.Chem.rdchem.PropertyPickleOptions.NoProps\n",
      "    PrivateProps = rdkit.Chem.rdchem.PropertyPickleOptions.PrivateProps\n",
      "    QueryAtomData = rdkit.Chem.rdchem.PropertyPickleOptions.QueryAtomData\n",
      "    SANITIZE_ADJUSTHS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ADJUSTH...\n",
      "    SANITIZE_ADJUST_REACTANTS = rdkit.Chem.rdChemReactions.SanitizeFlags.S...\n",
      "    SANITIZE_ALL = rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_ALL\n",
      "    SANITIZE_ATOM_MAPS = rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE...\n",
      "    SANITIZE_CLEANUP = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_CLEANUP\n",
      "    SANITIZE_CLEANUPCHIRALITY = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE...\n",
      "    SANITIZE_FINDRADICALS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_FIN...\n",
      "    SANITIZE_KEKULIZE = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_KEKULIZ...\n",
      "    SANITIZE_MERGEHS = rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_M...\n",
      "    SANITIZE_NONE = rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_NONE\n",
      "    SANITIZE_PROPERTIES = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_PROPE...\n",
      "    SANITIZE_RGROUP_NAMES = rdkit.Chem.rdChemReactions.SanitizeFlags.SANIT...\n",
      "    SANITIZE_SETAROMATICITY = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_S...\n",
      "    SANITIZE_SETCONJUGATION = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_S...\n",
      "    SANITIZE_SETHYBRIDIZATION = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE...\n",
      "    SANITIZE_SYMMRINGS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_SYMMRI...\n",
      "    UNCONSTRAINED_ANIONS = rdkit.Chem.rdchem.ResonanceFlags.UNCONSTRAINED_...\n",
      "    UNCONSTRAINED_CATIONS = rdkit.Chem.rdchem.ResonanceFlags.UNCONSTRAINED...\n",
      "    logger = <rdkit.RDLogger.logger object>\n",
      "\n",
      "FILE\n",
      "    /miniconda/lib/python3.6/site-packages/rdkit/Chem/AllChem.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Chem.AllChem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/miniconda/lib/python3.6/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "adrb = pd.read_csv('../data/ChemBL_ADRB1.csv', sep = ';')\n",
    "ht1a = pd.read_csv('../data/ChemBL_5HT1A.csv', sep = ';')\n",
    "\n",
    "clean_ardb = clean_df(adrb)\n",
    "clean_ht1a = clean_df(ht1a, drop_non_binders = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = Pool(6)\n",
    "\n",
    "clean_ardb['Unfolded'] = p.map(get_unfolded_fingerprint, clean_ardb['Smiles'].tolist())\n",
    "clean_ht1a['Unfolded'] = p.map(get_unfolded_fingerprint, clean_ht1a['Smiles'].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = Pool(6)\n",
    "\n",
    "clean_ardb['Morgan'] = p.map(get_morgan_matrix, clean_ardb['Smiles'].tolist())\n",
    "clean_ht1a['Morgan'] = p.map(get_morgan_matrix, clean_ht1a['Smiles'].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 892/892 [03:41<00:00,  4.03it/s]\n"
     ]
    }
   ],
   "source": [
    "clean_ardb['forrest_unfolded'] = clean_ardb['Smiles'].progress_apply(forrest_fingerprint)\n",
    "clean_ardb['forrest_unfolded'] = clean_ht1a['Smiles'].progress_apply(forrest_fingerprint)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f539040dac8>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAD8CAYAAACfF6SlAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAE69JREFUeJzt3X+M5HV9x/Hnu6CIrOUOke15XLpQr0bkIsIGsSbNHhj5ZTxNpDlC9FDMmRaN1mvqIUnVWBKsIq2pRU+hoiILRSkXwFKKbIhJATkE7gApK5xwcOWk4OmitR6++8d8TsZljp2dndmZ28/zkUzm+/18P/P9ft73nXvtd7/zne9GZiJJqsvv9XsAkqT5Z/hLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKrRvvwcAcPDBB+fIyEi/hzErzzzzDAcccEC/h9FT1rgwWOPCMb3OTZs2PZmZr+hkXQMR/iMjI9xxxx39HsasTExMMDY21u9h9JQ1LgzWuHBMrzMiftzpujztI0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFRqIb/hq7zGy/rq+bHfr+af2ZbvSQuWRvyRVyPCXpAoZ/pJUIcNfkio0Y/hHxEsi4vaIuDsi7o2IT5b2wyLitoh4MCKuiIgXl/b9yvxkWT7S2xIkSbPVzpH/r4DjM/N1wFHASRFxHPBp4MLMXA48DZxV+p8FPJ2ZrwIuLP0kSQNkxvDPhqky+6LySOB44KrSfinw9jK9qsxTlp8QEdG1EUuS5iwyc+ZOEfsAm4BXAV8APgPcWo7uiYhlwHcy88iI2AKclJnbyrIfAW/IzCenrXMtsBZgeHj4mPHx8e5VNQ+mpqYYGhrq9zB6qlWNmx/b2ZexrFh6YE/WW+t+XGhqqBGeX+fKlSs3ZeZoJ+tq60temfkscFRELAKuBl7Tqlt5bnWU/7yfMJm5AdgAMDo6mnvbn2Cr4c/GtarxzH59yeuMsRn7dKLW/bjQ1FAjdLfOWV3tk5k/BSaA44BFEbH7h8ehwONlehuwDKAsPxB4qhuDlSR1RztX+7yiHPETEfsDbwbuB24G3lm6rQGuKdMbyzxl+XeznXNLkqR5085pnyXApeW8/+8BV2bmtRFxHzAeEX8L/AC4uPS/GPh6REzSOOJf3YNxS5LmYMbwz8x7gNe3aH8IOLZF+/8Cp3VldJKknvAbvpJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpArNGP4RsSwibo6I+yPi3oj4UGn/REQ8FhF3lccpTa85JyImI+KBiDixlwVIkmZv3zb67ALWZeadEfEyYFNE3FiWXZiZn23uHBFHAKuB1wKvBP4jIv44M5/t5sAlSZ2b8cg/M7dn5p1l+ufA/cDSF3jJKmA8M3+VmQ8Dk8Cx3RisJKk7IjPb7xwxAtwCHAl8BDgT+BlwB43fDp6OiH8Ebs3Mb5TXXAx8JzOvmrautcBagOHh4WPGx8fnWsu8mpqaYmhoqN/D6KlWNW5+bGdfxrJi6YE9WW+t+3GhqaFGeH6dK1eu3JSZo52sq53TPgBExBDwLeDDmfmziLgI+BSQ5fkC4L1AtHj5837CZOYGYAPA6Ohojo2NzXrw/TQxMcHeNubZalXjmeuv68tYtp4xNmOfTtS6HxeaGmqE7tbZ1tU+EfEiGsF/WWZ+GyAzn8jMZzPzN8CXee7UzjZgWdPLDwUe78poJUld0c7VPgFcDNyfmZ9ral/S1O0dwJYyvRFYHRH7RcRhwHLg9u4NWZI0V+2c9nkT8C5gc0TcVdo+BpweEUfROKWzFXg/QGbeGxFXAvfRuFLobK/0kaTBMmP4Z+b3aH0e//oXeM15wHlzGJckqYf8hq8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalCM4Z/RCyLiJsj4v6IuDciPlTaD4qIGyPiwfK8uLRHRHw+IiYj4p6IOLrXRUiSZqedI/9dwLrMfA1wHHB2RBwBrAduyszlwE1lHuBkYHl5rAUu6vqoJUlzMmP4Z+b2zLyzTP8cuB9YCqwCLi3dLgXeXqZXAV/LhluBRRGxpOsjlyR1LDKz/c4RI8AtwJHAI5m5qGnZ05m5OCKuBc7PzO+V9puAj2bmHdPWtZbGbwYMDw8fMz4+PsdS5tfU1BRDQ0P9HkZPtapx82M7+zKWFUsP7Ml6a92PC00NNcLz61y5cuWmzBztZF37ttsxIoaAbwEfzsyfRcQeu7Zoe95PmMzcAGwAGB0dzbGxsXaHMhAmJibY28Y8W61qPHP9dX0Zy9Yzxmbs04la9+NCU0ON0N0627raJyJeRCP4L8vMb5fmJ3afzinPO0r7NmBZ08sPBR7vymglSV0x45F/NA7xLwbuz8zPNS3aCKwBzi/P1zS1fyAixoE3ADszc3tXR63qjPToN451K3bN+NvM1vNP7cm2pX5q57TPm4B3AZsj4q7S9jEaoX9lRJwFPAKcVpZdD5wCTAK/AN7T1RFLkuZsxvAvH9zu6QT/CS36J3D2HMclSeohv+ErSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SapQ23/AXYOjV3/ScLp2/sShpL2TR/6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFVoxvCPiEsiYkdEbGlq+0REPBYRd5XHKU3LzomIyYh4ICJO7NXAJUmda+fI/6vASS3aL8zMo8rjeoCIOAJYDby2vOafImKfbg1WktQdM4Z/Zt4CPNXm+lYB45n5q8x8GJgEjp3D+CRJPTCXc/4fiIh7ymmhxaVtKfBoU59tpU2SNEAiM2fuFDECXJuZR5b5YeBJIIFPAUsy870R8QXgPzPzG6XfxcD1mfmtFutcC6wFGB4ePmZ8fLwrBc2XqakphoaG+rLtzY/tnJftDO8PT/xyXjbVN+3UuGLpgfMzmB7p53t1vtRQIzy/zpUrV27KzNFO1tXRjd0y84nd0xHxZeDaMrsNWNbU9VDg8T2sYwOwAWB0dDTHxsY6GUrfTExM0K8xz9fN1tat2MUFmxf2vf/aqXHrGWPzM5ge6ed7db7UUCN0t86OTvtExJKm2XcAu68E2gisjoj9IuIwYDlw+9yGKEnqthkP6yLicmAMODgitgEfB8Yi4igap322Au8HyMx7I+JK4D5gF3B2Zj7bm6FLkjo1Y/hn5uktmi9+gf7nAefNZVCSpN7yG76SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKzRj+EXFJROyIiC1NbQdFxI0R8WB5XlzaIyI+HxGTEXFPRBzdy8FLkjrTzpH/V4GTprWtB27KzOXATWUe4GRgeXmsBS7qzjAlSd00Y/hn5i3AU9OaVwGXlulLgbc3tX8tG24FFkXEkm4NVpLUHft2+LrhzNwOkJnbI+KQ0r4UeLSp37bStr3zIUr9NbL+ur5sd+v5p/Zlu6pDp+G/J9GiLVt2jFhL49QQw8PDTExMdHkovTU1NdW3Ma9bsWtetjO8//xtq18GucZuvb/6+V6dLzXUCN2ts9PwfyIilpSj/iXAjtK+DVjW1O9Q4PFWK8jMDcAGgNHR0RwbG+twKP0xMTFBv8Z85jwdia5bsYsLNnf7+GCwDHKNW88Y68p6+vlenS811AjdrbPTSz03AmvK9Brgmqb2d5erfo4Ddu4+PSRJGhwzHvJExOXAGHBwRGwDPg6cD1wZEWcBjwCnle7XA6cAk8AvgPf0YMySpDmaMfwz8/Q9LDqhRd8Ezp7roCRJveU3fCWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRXady4vjoitwM+BZ4FdmTkaEQcBVwAjwFbgzzLz6bkNU5LUTd048l+ZmUdl5miZXw/clJnLgZvKvCRpgMzpyH8PVgFjZfpSYAL4aA+203cj66/r9xAkqSNzPfJP4N8jYlNErC1tw5m5HaA8HzLHbUiSuiwys/MXR7wyMx+PiEOAG4EPAhszc1FTn6czc3GL164F1gIMDw8fMz4+3vE4+mFqaoqHdz7b72H01PD+8MQv+z2K3hrkGlcsPbAr65mammJoaKgr6xpUNdQIz69z5cqVm5pOuc/KnE77ZObj5XlHRFwNHAs8ERFLMnN7RCwBduzhtRuADQCjo6M5NjY2l6HMu4mJCS743jP9HkZPrVuxiws29+LM4OAY5Bq3njHWlfVMTEywt/3/mq0aaoTu1tnxaZ+IOCAiXrZ7GngLsAXYCKwp3dYA18x1kJKk7prLIc8wcHVE7F7PNzPz3yLi+8CVEXEW8Ahw2tyHKUnqpo7DPzMfAl7Xov1/gBPmMihJUm/5DV9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVWgwb2Q+C/36U4rrVuxiAfzzaYB16729bsUuzpzFuraef2pXtqvB5pG/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyK+oSqpev+4UAP37RrVH/pJUoZ4d+UfEScA/APsAX8nM83u1LUkLQ6dH4LO9f5F6dOQfEfsAXwBOBo4ATo+II3qxLUnS7PXqtM+xwGRmPpSZ/weMA6t6tC1J0iz1KvyXAo82zW8rbZKkARCZ2f2VRpwGnJiZ7yvz7wKOzcwPNvVZC6wts68GHuj6QHrrYODJfg+ix6xxYbDGhWN6nX+Yma/oZEW9+sB3G7Csaf5Q4PHmDpm5AdjQo+33XETckZmj/R5HL1njwmCNC0c36+zVaZ/vA8sj4rCIeDGwGtjYo21JkmapJ0f+mbkrIj4A3EDjUs9LMvPeXmxLkjR7PbvOPzOvB67v1foHwF57ymoWrHFhsMaFo2t19uQDX0nSYPP2DpJUIcP/BUTEPhHxg4i4tswfFhG3RcSDEXFF+TCbiNivzE+W5SP9HHe7ImJRRFwVET+MiPsj4o0RcVBE3FhqvDEiFpe+ERGfLzXeExFH93v87YiIv4yIeyNiS0RcHhEvWQj7MSIuiYgdEbGlqW3W+y4i1pT+D0bEmn7Usid7qPEz5f16T0RcHRGLmpadU2p8ICJObGo/qbRNRsT6+a7jhbSqsWnZX0VERsTBZb67+zEzfezhAXwE+CZwbZm/Elhdpr8I/HmZ/gvgi2V6NXBFv8feZn2XAu8r0y8GFgF/B6wvbeuBT5fpU4DvAAEcB9zW7/G3Ud9S4GFg/6b9d+ZC2I/AnwJHA1ua2ma174CDgIfK8+Iyvbjftc1Q41uAfcv0p5tqPAK4G9gPOAz4EY2LTfYp04eX9/jdwBH9ru2Faizty2hcMPNj4OBe7Me+Fz+oDxrfTbgJOB64tvyDP9n0xnsjcEOZvgF4Y5net/SLftcwQ32/X4IxprU/ACwp00uAB8r0l4DTW/Ub1AfPfdP8oLJfrgVOXCj7ERiZFoyz2nfA6cCXmtp/p98gPKbXOG3ZO4DLyvQ5wDlNy24o+/a3+7dVv0F4tKoRuAp4HbC1Kfy7uh897bNnfw/8NfCbMv9y4KeZuavMN9+y4re3syjLd5b+g+xw4CfAP5dTW1+JiAOA4czcDlCeDyn997pbdmTmY8BngUeA7TT2yyYW1n5sNtt9t9ft02neS+NIGBZQjRHxNuCxzLx72qKu1mj4txARbwV2ZOam5uYWXbONZYNqXxq/bl6Uma8HnqFxqmBP9roayznvVTROA7wSOIDGnWan25v3Yzv2VNdeW29EnAvsAi7b3dSi215XY0S8FDgX+JtWi1u0dVyj4d/am4C3RcRWGnckPZ7GbwKLImL3dyOab1nx29tZlOUHAk/N54A7sA3Ylpm3lfmraPwweCIilgCU5x1N/V/wlh0D6M3Aw5n5k8z8NfBt4E9YWPux2Wz33d64TykfaL4VOCPLeQ4WTo1/RONg5e6SP4cCd0bEH9DlGg3/FjLznMw8NDNHaHzw993MPAO4GXhn6bYGuKZMbyzzlOXfbXpTDqTM/G/g0Yh4dWk6AbiP361leo3vLlccHAfs3H2KYYA9AhwXES+NiOC5GhfMfpxmtvvuBuAtEbG4/Jb0ltI2sKLxR6I+CrwtM3/RtGgjsLpcsXUYsBy4nb3sVjOZuTkzD8nMkZI/24Cjy//X7u7Hfn/YMegPYIznrvY5nMYbahL4F2C/0v6SMj9Zlh/e73G3WdtRwB3APcC/0rhS4OU0Puh+sDwfVPoGjT/Q8yNgMzDa7/G3WeMngR8CW4Cv07gaZK/fj8DlND7H+HUJiLM62Xc0zptPlsd7+l1XGzVO0ji/fVd5fLGp/7mlxgeAk5vaTwH+qyw7t991zVTjtOVbee4D367uR7/hK0kV8rSPJFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUL/D8hvYy7Q4ivBAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clean_ardb['forrest_unfolded'].apply(len).hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdd7c489470>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAEJCAYAAAB8Pye7AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAFeJJREFUeJzt3X2QXfV93/H314AxZikPVtgqgrGcRo2DLZtYW8ZOppldu1PzkFRQhwaGYpTgUR6wx50oTeTMxE6bMlacYKeuncRKocjjhzUJTkxBdopVKcQt2EYUI0AhyEaDhVRRYqxYhiaz7rd/nN/C5Wa19+o+7N3L7/2aubPn/M7Td3977ueePffccyMzkSTV5SWjLkCStPQMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFThx1AQArVqzI1atXj7qM53z3u9/l1FNPHXUZixqHGmE86hyHGmE86hyHGmE86uymxt27dz+Vmd/X0wYyc+SPdevW5XKyc+fOUZfQ0TjUmDkedY5DjZnjUec41Jg5HnV2UyNwb/aYu572kaQKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekCi2L2zv0Y/XmOwa+zk1r59jQxXr3b7lk4NuWpKXgkb8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekCnUM/4g4NyJ2RsTeiHgoIt5d2n8jIp6IiPvL4+KWZd4TEfsi4pGIeOswfwFJ0vHr5t4+c8CmzLwvIk4DdkfEnWXahzLzd1pnjojzgCuA1wDfD3wxIv5xZn5vkIVLknrX8cg/Mw9l5n1l+DvAXmDVIousB2Yz828z8zFgH3DBIIqVJA3GcZ3zj4jVwI8AXy5N74yIByLipog4s7StAr7ZstgBFn+xkCQtscjM7maMmAD+HLg+Mz8bEZPAU0ACvwmszMyfjYiPAndn5ifKcjcC2zPz1rb1bQQ2AkxOTq6bnZ3t6RfY88SRnpZbzOQpcPjZzvOtXXX6wLfdraNHjzIxMTGy7XdrHOochxphPOochxphPOrspsaZmZndmTnVy/q7up9/RJwE3Ap8MjM/C5CZh1um/yFwexk9AJzbsvg5wMH2dWbmVmArwNTUVE5PT/dQPl3dd/94bVo7xw17OnfN/qumB77tbu3atYte+2wpjUOd41AjjEed41AjjEedw66xm6t9ArgR2JuZH2xpX9ky22XAg2X4NuCKiDg5Il4FrAG+MriSJUn96ubI/8eAq4E9EXF/afs14MqIOJ/mtM9+4OcAMvOhiLgFeJjmSqHrvNJHkpaXjuGfmV8CYoFJ2xdZ5nrg+j7qkiQNkZ/wlaQKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRXqGP4RcW5E7IyIvRHxUES8u7SfFRF3RsSj5eeZpT0i4sMRsS8iHoiINwz7l5AkHZ9ujvzngE2Z+cPAG4HrIuI8YDOwIzPXADvKOMBFwJry2Aj8/sCrliT1pWP4Z+ahzLyvDH8H2AusAtYD28ps24BLy/B64OPZuAc4IyJWDrxySVLPIjO7nzliNXAX8Frg8cw8o2Xa05l5ZkTcDmzJzC+V9h3Ar2bmvW3r2kjznwGTk5PrZmdne/oF9jxxpKflFjN5Chx+tvN8a1edPvBtd+vo0aNMTEyMbPvdGoc6x6FGGI86x6FGGI86u6lxZmZmd2ZO9bL+E7udMSImgFuBf5OZfxMRx5x1gba/9wqTmVuBrQBTU1M5PT3dbSkvsGHzHT0tt5hNa+e4YU/nrtl/1fTAt92tXbt20WufLaVxqHMcaoTxqHMcaoTxqHPYNXZ1tU9EnEQT/J/MzM+W5sPzp3PKzydL+wHg3JbFzwEODqZcSdIgdHO1TwA3Ansz84Mtk24DrinD1wCfa2l/e7nq543Akcw8NMCaJUl96ua0z48BVwN7IuL+0vZrwBbgloi4FngcuLxM2w5cDOwDngF+ZqAVS5L61jH8yxu3xzrB/5YF5k/guj7rkiQNkZ/wlaQKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqUMfwj4ibIuLJiHiwpe03IuKJiLi/PC5umfaeiNgXEY9ExFuHVbgkqXfdHPnfDFy4QPuHMvP88tgOEBHnAVcArynL/F5EnDCoYiVJg9Ex/DPzLuBbXa5vPTCbmX+bmY8B+4AL+qhPkjQE/Zzzf2dEPFBOC51Z2lYB32yZ50BpkyQtI5GZnWeKWA3cnpmvLeOTwFNAAr8JrMzMn42IjwJ3Z+Ynynw3Atsz89YF1rkR2AgwOTm5bnZ2tqdfYM8TR3pabjGTp8DhZzvPt3bV6QPfdreOHj3KxMTEyLbfrXGocxxqhPGocxxqhPGos5saZ2ZmdmfmVC/rP7GXhTLz8PxwRPwhcHsZPQCc2zLrOcDBY6xjK7AVYGpqKqenp3sphQ2b7+hpucVsWjvHDXs6d83+q6YHvu1u7dq1i177bCmNQ53jUCOMR53jUCOMR53DrrGn0z4RsbJl9DJg/kqg24ArIuLkiHgVsAb4Sn8lSpIGrePhbUR8GpgGVkTEAeB9wHREnE9z2mc/8HMAmflQRNwCPAzMAddl5veGU7okqVcdwz8zr1yg+cZF5r8euL6foiRJw+UnfCWpQj294Stp+Fa3Xcywae3cUC5waLd/yyVD34ZGzyN/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekCnlXT0nVa7+D6lIa1V1UPfKXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoUMf0mqUMfwj4ibIuLJiHiwpe2siLgzIh4tP88s7RERH46IfRHxQES8YZjFS5J6082R/83AhW1tm4EdmbkG2FHGAS4C1pTHRuD3B1OmJGmQOoZ/Zt4FfKuteT2wrQxvAy5taf94Nu4BzoiIlYMqVpI0GL2e85/MzEMA5efZpX0V8M2W+Q6UNknSMhKZ2XmmiNXA7Zn52jL+7cw8o2X605l5ZkTcAbw/M79U2ncAv5KZuxdY50aaU0NMTk6um52d7ekX2PPEkZ6WW8zkKXD42c7zrV11+sC33a2jR48yMTExsu13axzqXK41tu/b3e6X/epnv16ufdmuvc5h5Ei3jtXf3fTlzMzM7syc6mW7vX6N4+GIWJmZh8ppnSdL+wHg3Jb5zgEOLrSCzNwKbAWYmprK6enpngrZMISvX9u0do4b9nTumv1XTQ98293atWsXvfbZUhqHOpdrje37drf7Zb/62a+Xa1+2a69zGDnSrWP197D7stfTPrcB15Tha4DPtbS/vVz180bgyPzpIUnS8tHxMCIiPg1MAysi4gDwPmALcEtEXAs8DlxeZt8OXAzsA54BfmYINUuS+tQx/DPzymNMessC8yZwXb9FSZKGy0/4SlKFDH9JqpDhL0kVMvwlqULDv2hYLyqrj/N66E1r5wZyDfX+LZf0vQ5Jz/PIX5IqZPhLUoUMf0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKuQnfMfQ6s13DOyTs5Lq5JG/JFXI8JekChn+klQhz/n34XjvcClJy4VH/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqZPhLUoX6urdPROwHvgN8D5jLzKmIOAv4DLAa2A/8q8x8ur8yVbth3kep03cj7N9yydC2LY3KII78ZzLz/MycKuObgR2ZuQbYUcYlScvIME77rAe2leFtwKVD2IYkqQ+Rmb0vHPEY8DSQwMcyc2tEfDszz2iZ5+nMPHOBZTcCGwEmJyfXzc7O9lTDnieO9LTcYiZPgcPPDny1AzUONcJ41NmpxrWrTl+6Ylq079tL1Zf9/L5Hjx5lYmJigNUMR3udw8iRbh2rv7vpy5mZmd0tZ12OS7/h//2ZeTAizgbuBN4F3NZN+LeamprKe++9t6cahnEueNPaOW7Ys7y/6mAcaoTxqLNTjaM659++by9VX/bz++7atYvp6enBFTMk7XWO8rs5jtXf3fRlRPQc/n3tSZl5sPx8MiL+BLgAOBwRKzPzUESsBJ7sZxuS6rFUIdzpTf4a9HzOPyJOjYjT5oeBfw48CNwGXFNmuwb4XL9FSpIGq58j/0ngTyJifj2fyswvRMRXgVsi4lrgceDy/suUJA1Sz+Gfmd8AXr9A+18Db+mnKEnScPkJX0mqkOEvSRUy/CWpQoa/JFXI8JekChn+klQhw1+SKmT4S1KFDH9JqpDhL0kVMvwlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFTL8JalChr8kVcjwl6QKGf6SVCHDX5IqdOKoC5C0vKzefEfPy25aO8eGPpbX0vHIX5IqZPhLUoUMf0mqkOf8pQ76OQcuLVce+UtShYYW/hFxYUQ8EhH7ImLzsLYjSTp+Qwn/iDgB+ChwEXAecGVEnDeMbUmSjt+wjvwvAPZl5jcy8++AWWD9kLYlSTpOwwr/VcA3W8YPlDZJ0jIQmTn4lUZcDrw1M99Rxq8GLsjMd7XMsxHYWEZ/CHhk4IX0bgXw1KiL6GAcaoTxqHMcaoTxqHMcaoTxqLObGl+Zmd/Xy8qHdannAeDclvFzgIOtM2TmVmDrkLbfl4i4NzOnRl3HYsahRhiPOsehRhiPOsehRhiPOodd47BO+3wVWBMRr4qIlwJXALcNaVuSpOM0lCP/zJyLiHcCfwacANyUmQ8NY1uSpOM3tE/4ZuZ2YPuw1j9ky/J0VJtxqBHGo85xqBHGo85xqBHGo86h1jiUN3wlScubt3eQpAq9aMM/Ik6IiP8VEbcvMO2XIuLhiHggInZExCtL+0xE3N/y+L8RcWmZ9hct7Qcj4k9L+3REHGmZ9t5h11mmfSAiHoqIvRHx4YiI0r4uIvaU22q0tp8VEXdGxKPl55mjqDEiXh4Rd0TEX5ZpW1rm3xAR/6elL98xwn7cVW5PMl/L2aX95Ij4TOnfL0fE6m5rHHSdEXFa2/76VET87gj78rci4sHy+OmW9leVvnq09N1LS/uo+vJYdX6y/M0fjIibIuKk0t7zc3wINd4cEY+11HJ+aY+yX+wr63tDx+Iy80X5AH4J+BRw+wLTZoCXl+FfAD6zwDxnAd+an69t2q3A28vw9ELbGGadwI8C/4PmzfQTgLuB6TLtK8CbgAA+D1xU2j8AbC7Dm4HfGkWNwMuBmTLPS4G/aKlxA/CRZdKPu4CpBdb1i8AflOErFtp3lrLOtuV3Az8+or68BLiT5n3EU4F7gX9Qpt0CXFGG/wD4hRH25WJ1XkzzvAng0y11Ti+0jRHVeDPwUwus62Ka53sAbwS+3Km2F+WRf0ScQ9OB/3mh6Zm5MzOfKaP30HwOod1PAZ9vmW9+3acBbwb+dIR1JvAymvA8GTgJOBwRK2l2kruz2SM+DlxallkPbCvD21ral7TGzHwmM3eWZf8OuI+F+79rg66xw+Za+/GPgbfM/7cwyjojYg1wNs2Lac/6qPE84M8zcy4zvwt8Dbiw9M2bafoKXrjvjaIvF6yzLLM9C5qDqFHtl8escRHrgY+X8u8Bzih5cEwvyvAHfhf4FeD/dTHvtTSvmO2uoHn1b3cZsCMz/6al7U0R8bWI+HxEvGbYdWbm3cBO4FB5/Flm7qW5hcaBlmVab6sxmZmHyvKHaIJiFDU+JyLOAH4S2NHS/Lbyb+sfR0TrBwVHUeN/Kf9a/3pLKD1365LMnAOOAK8YcZ0AV9IcObZewbFkfUkTUBdFc1pvBc1R7bk0ffPt0lfwwn1yyftykTqfU073XA18oaW5l+f4sGq8vvxdPxQRJ5e2476lzosu/CPiJ4AnM3N3F/P+a2AK+O229pXAWprPKbS7khe+KNxH8xHr1wP/iS7/I+inzoj4QeCHaY4UVgFvjogfp/mXr13Pl3MNqcb5+U+k6ccPZ+Y3SvN/BVZn5uuAL/L8UeEoarwqM9cC/7Q8rp5fzQKr7tjHw+zLov1gZUn7MjP/G82l3f+z1HE3MMfi/bXkfblIna1+D7grM+f/izru5/gQa3wP8Grgn9Ccmv7V+dUssOrF+7KX81jL+QG8n+ZVbz/wv4FngE8sMN8/A/YCZy8w7d3A1gXaXwH8NfCyRba/H1gxzDqBfwv8esv4e2mOMFYCf9nSfiXwsTL8CLCyDK8EHhlFjS3jN9EE/7G2fQJwZJQ1trRvoJw/pzkgeFMZPpHm3isx4r58PfBXo+zLBeb5FM+fQ38KOLG0v4nmP5eR9OWx6mwZfx9NuL9kkWX20+E5PswaW9qnKe8lAB8DrmyZ9tzz/Zjr7dTR4/zgGG/UAD8CfB1Yc4zl7qG8KdnW/vPAtra2fzi/w9LcyvrxbnbgfuoEfprmaO5EmvO/O4CfLNO+SvOGz/wbvheX9t/mhW/4fmCENf4HmjfNX9K2zMqW4cuAe0ZRYxlfUeY5ieZ89M+X8et44ZuUtwx7v1ysL8v0LcC/G3FfngC8ogy/DniQ5wP/j3jhG76/OMK+XKzOd9AcbZ/Stkxfz/EB1zh/ABc0p5W2lPFLeOEbvl/pWNfxdvY4PXjhK+O/B/5FGf4izRtm95fHbS3LrAaeYIFXfporQC5sa3sn8BDNebp7gB8ddp1l5/gYzRHDw8AHW9Y1VXaWrwMfadlpX0ETGo+Wn2eNokaaUxdZ2ueXeUeZ9v6WvtwJvHpENZ5Kc+XMA6We/wicUKa9jCbM9tG8KfgDo/x7l+nfaO+rEfTly0ptD9M8D85vWdcPlL7aV/ru5BH25WJ1ztE8b+aXee8gnuMDrvG/A3tonuOfACZKe9B8gdbXy/S/d6Va+8NP+EpShV50b/hKkjoz/CWpQoa/JFXI8JekChn+krTEyo3jnoyIB7uY95Xlxm8PRHPDwb5uOzHP8JekpXczne/XM+93aO7b8zqaS0XfP4gCDH9JWmKZeRfNXYOfExH/KCK+EBG7o7mF/KvLpPN4/v5XO2lu4tY3w1+SloetwLsycx3wyzT3GILmw2VvK8OXAadFRLc3wDumoX2HrySpOxExQfO9DX/Uclfr+Tt2/jLwkYjYANxFcweC9pvRHTfDX5JG7yU0t74+v31CZh4E/iU89yLxtsw8MogNSpJGKJvvB3ksIi6H576W8fVleEVEzGf1e2juiNs3w1+SllhEzN+n/4ci4kBEXAtcBVwbEV+juZHc/Bu708AjEfFXwCRw/UBq8MZuklQfj/wlqUKGvyRVyPCXpAoZ/pJUIcNfkipk+EtShQx/SaqQ4S9JFfr/a4QCaZOI67AAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clean_ardb['Unfolded'].apply(lambda x: x.GetLength()).hist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdd6454e470>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAD/JJREFUeJzt3X+s3XV9x/HnW67Ir0mBjhvSdrts1k0iWcbuCNNFL3YxUo1lCRgMm51r1mRzTAeLdDOZy/4SN0AlRtdQZ1060SHaxsGcAU7MjJDxSzqojgZruVBBww+9OuPq3vvjfMju2lvuKff86H2f5yM5Od8fn+/383n33Pu63/s533MbmYkkqa6XjHoAkqTBMuglqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKmxj1AABWrlyZU1NTox7GUfvhD3/IySefPOphDNW41Txu9YI1Lyf33nvv9zLzZxdrd0wE/dTUFPfcc8+oh3HUOp0OMzMzox7GUI1bzeNWL1jzchIR3+6lnVM3klScQS9JxRn0klScQS9JxRn0klScQS9JxRn0klScQS9JxRn0klTcMfHJWOlYtfvx5/i9Lf88kr73feDNI+lX9XhFL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVJxBL0nFGfSSVFxPQR8RfxoRD0XEf0TEpyPihIg4OyLujohHIuIzEXF8a/uytr637Z8aZAGSpBe2aNBHxCrgT4DpzHw1cBxwGXANcH1mrgWeATa1QzYBz2TmK4DrWztJ0oj0OnUzAZwYERPAScAB4A3AzW3/duDitryhrdP2r4uI6M9wJUlHa9Ggz8zHgb8F9tMN+OeAe4FnM/NgazYLrGrLq4DH2rEHW/sz+jtsSVKvFv3PwSPiNLpX6WcDzwL/BFy0QNN8/pAX2Df/vJuBzQCTk5N0Op3eRnwMmZubW5bjXopxq3nyRLjq3IOLNxyAUf07j9trDPVrXjTogd8CvpWZ3wWIiFuA1wArImKiXbWvBp5o7WeBNcBsm+o5FXj60JNm5lZgK8D09HTOzMwssZTh63Q6LMdxL8W41XzDjp1cu7uXb5P+23f5zEj6HbfXGOrX3Msc/X7ggog4qc21rwMeBu4ELmltNgI72/Kutk7bf0dmHnZFL0kajl7m6O+m+6bqfcDudsxW4GrgyojYS3cOfls7ZBtwRtt+JbBlAOOWJPWop99JM/P9wPsP2fwocP4CbX8MXLr0oUmS+sFPxkpScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBVn0EtScQa9JBXXU9BHxIqIuDkivhEReyLiNyLi9Ij4ckQ80p5Pa20jIj4SEXsj4sGIOG+wJUiSXkivV/QfBv4lM38Z+BVgD7AFuD0z1wK3t3WAi4C17bEZ+FhfRyxJOiqLBn1EvBx4HbANIDN/kpnPAhuA7a3ZduDitrwB+FR23QWsiIiz+j5ySVJPermi/wXgu8DfR8T9EXFjRJwMTGbmAYD2fGZrvwp4bN7xs22bJGkEJnpscx5wRWbeHREf5v+maRYSC2zLwxpFbKY7tcPk5CSdTqeHoRxb5ubmluW4l2Lcap48Ea469+BI+h7Vv/O4vcZQv+Zegn4WmM3Mu9v6zXSD/smIOCszD7SpmafmtV8z7/jVwBOHnjQztwJbAaanp3NmZubFVTBCnU6H5TjupRi3mm/YsZNrd/fybdJ/+y6fGUm/4/YaQ/2aF526yczvAI9FxC+1TeuAh4FdwMa2bSOwsy3vAt7R7r65AHju+SkeSdLw9XqpcgWwIyKOBx4F3kn3h8RnI2ITsB+4tLW9FVgP7AV+1NpKkkakp6DPzAeA6QV2rVugbQLvWuK4JEl94idjJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJak4g16SijPoJam4noM+Io6LiPsj4ott/eyIuDsiHomIz0TE8W37y9r63rZ/ajBDlyT14miu6N8N7Jm3fg1wfWauBZ4BNrXtm4BnMvMVwPWtnSRpRHoK+ohYDbwZuLGtB/AG4ObWZDtwcVve0NZp+9e19pKkEej1iv5DwHuB/2nrZwDPZubBtj4LrGrLq4DHANr+51p7SdIITCzWICLeAjyVmfdGxMzzmxdomj3sm3/ezcBmgMnJSTqdTi/jPabMzc0ty3EvxbjVPHkiXHXuwcUbDsCo/p3H7TWG+jUvGvTAa4G3RsR64ATg5XSv8FdExES7al8NPNHazwJrgNmImABOBZ4+9KSZuRXYCjA9PZ0zMzNLLGX4Op0Oy3HcSzFuNd+wYyfX7u7l26T/9l0+M5J+x+01hvo1Lzp1k5l/npmrM3MKuAy4IzMvB+4ELmnNNgI72/Kutk7bf0dmHnZFL0kajqXcR381cGVE7KU7B7+tbd8GnNG2XwlsWdoQJUlLcVS/k2ZmB+i05UeB8xdo82Pg0j6MTZLUB34yVpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqTiDXpKKM+glqbhFgz4i1kTEnRGxJyIeioh3t+2nR8SXI+KR9nxa2x4R8ZGI2BsRD0bEeYMuQpJ0ZL1c0R8ErsrMVwEXAO+KiHOALcDtmbkWuL2tA1wErG2PzcDH+j5qSVLPFg36zDyQmfe15R8Ae4BVwAZge2u2Hbi4LW8APpVddwErIuKsvo9cktSTyMzeG0dMAV8BXg3sz8wV8/Y9k5mnRcQXgQ9k5r+17bcDV2fmPYecazPdK34mJyd/7aabblpiKcM3NzfHKaecMuphDNW41fzU08/x5H+Npu9zV506kn7H7TWG5VvzhRdeeG9mTi/WbqLXE0bEKcDngPdk5vcj4ohNF9h22E+TzNwKbAWYnp7OmZmZXodyzOh0OizHcS/FuNV8w46dXLu752+Tvtp3+cxI+h231xjq19zTXTcR8VK6Ib8jM29pm598fkqmPT/Vts8Ca+Ydvhp4oj/DlSQdrV7uuglgG7AnM6+bt2sXsLEtbwR2ztv+jnb3zQXAc5l5oI9jliQdhV5+J30t8LvA7oh4oG37C+ADwGcjYhOwH7i07bsVWA/sBX4EvLOvI5YkHZVFg769qXqkCfl1C7RP4F1LHJckqU/8ZKwkFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFWfQS1JxBr0kFTeQoI+IN0XENyNib0RsGUQfkqTe9D3oI+I44KPARcA5wNsj4px+9yNJ6s0grujPB/Zm5qOZ+RPgJmDDAPqRJPVgEEG/Cnhs3vps2yZJGoGJAZwzFtiWhzWK2AxsbqtzEfHNAYxl0FYC3xv1IIZs3GoeWb1xzSh6BcbvNYblW/PP99JoEEE/C6yZt74aeOLQRpm5Fdg6gP6HJiLuyczpUY9jmMat5nGrF6y5okFM3fw7sDYizo6I44HLgF0D6EeS1IO+X9Fn5sGI+GPgS8BxwCcy86F+9yNJ6s0gpm7IzFuBWwdx7mPMsp56epHGreZxqxesuZzIPOx9UklSIf4JBEkqbmyDPiLWRMSdEbEnIh6KiHe37adHxJcj4pH2fNohx/16RPw0Ii5p6xdGxAPzHj+OiIuP0OfbIuLh1t8/Dr7Kw/ofas0R8XOtv/sj4sGIWD+cSv/fGPpSc9v2wXaOPRHxkYg47Fbixc47DCOo+W8i4hvtNf58RKwYfJWHjWGoNc9r+2cRkRGxcnDV9UFmjuUDOAs4ry3/DPCfdP9kwweBLW37FuCaecccB9xB9/2HSxY45+nA08BJC+xbC9wPnNbWzxyDmrcCf9iWzwH2LdeagdcAX237jgO+Bsws0N8Rz1u45jcCE235mnGoubVdQ/emk28DK4dd89E8xvaKPjMPZOZ9bfkHwB66n+DdAGxvzbYD869UrwA+Bzx1hNNeAtyWmT9aYN8fAB/NzGdan0c6x8CMoOYEXt6WT2WBz1MMWh9rTuAE4HjgZcBLgScX6PKFzjsUw645M/81Mw+21bvofnZmqEbwOgNcD7yXBT4QeqwZ26CfLyKmgF8F7gYmM/MAdL94gDNbm1XAbwMff4FTXQZ8+gj7Xgm8MiK+GhF3RcSb+jP6F2dINf8V8DsRMUv3qumKPgz9RVtKzZn5NeBO4EB7fCkz9yzQzYLnHZUh1Tzf7wO39a+CozeMmiPircDjmfn1gRXSR2Mf9BFxCt2f6u/JzO+/QNMPAVdn5k+PcJ6zgHPp/iq3kAm60zczwNuBG0cxlwlDrfntwCczczWwHviHiBjJ19xSa46IVwCvonu1ugp4Q0S8blDj7Ydh1xwR7wMOAjuWOvYXaxg1R8RJwPuAv+zn2AdpIPfRLxcR8VK6XxQ7MvOWtvnJiDgrMw+0IHv+17pp4Kb2vsxKYH1EHMzML7T9bwM+n5n/fYTuZoG72v5vRfdv+6yl+0nioRlyzZuAN0H3SikiTmjnGeq0VT9qpvta3ZWZc+2ctwEXAF85pLsjnXeohlwzEbEReAuwLtsE9rANseZfBM4Gvt6OXw3cFxHnZ+Z3BlrkizS2V/TtnfRtwJ7MvG7erl3Axra8EdgJkJlnZ+ZUZk4BNwN/NC/woHv1eqQpDIAvABe2vlfSncp5tA+l9GwENe8H1rW+X0V37vO7fSilZ32seT/w+oiYaIHyerrzwIda8LzDNOya2zTk1cBbj/BezcANs+bM3J2ZZ847fpbuG8HHZMgDY33XzW/SfRPlQeCB9lgPnAHcDjzSnk9f4NhPMu8OFGAKeBx4ySHt/pruFz90/6rndcDDwG7gsjGo+Ry6dzB8vfX1xuVaM907MP6O7jf9w8B189rdCEy35UXPW7DmvXT/NPnzfX28es2HHL+PY/yuGz8ZK0nFje3UjSSNC4Nekooz6CWpOINekooz6CWpOINekooz6CWpOINekor7XzME9bvEw1PYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clean_ardb['Morgan'].apply(lambda x: len(x)).hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_ardb['Y'] = clean_ardb['Standard Value'].apply(lambda x: 0 if x < 1000 else 1)\n",
    "clean_ht1a['Y'] = clean_ht1a['Standard Value'].apply(lambda x: 0 if x < 1000 else 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_morgan = clean_ardb['Morgan'].append(clean_ht1a['Morgan'])\n",
    "X_morgan = np.array([np.array(i) for i in X_train_morgan.values])\n",
    "\n",
    "y_train_morgan = clean_ardb['Y'].append(clean_ht1a['Y'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_morgan, X_test_morgan, \\\n",
    "y_train_morgan, y_test_morgan = train_test_split(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_ardb['Unfolded'] = clean_ardb['Unfolded'].apply(conv_rdkit_to_csr)\n",
    "clean_ht1a['Unfolded'] = clean_ht1a['Unfolded'].apply(conv_rdkit_to_csr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/miniconda/lib/python3.6/site-packages/sklearn/utils/validation.py:475: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.\n",
      "  warnings.warn(msg, DataConversionWarning)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'X_test' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-100-0cd6a0288913>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mclf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRMTClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train_morgan\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train_morgan\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mtest_preds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepsilon_multiple\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'confusion_matrix \\n'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_preds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X_test' is not defined"
     ]
    }
   ],
   "source": [
    "clf = RMTClassifier()\n",
    "clf.fit(X_train_morgan, y_train_morgan)\n",
    "test_preds = clf.predict(X_train_morgan, epsilon_multiple = 1)\n",
    "\n",
    "print('confusion_matrix \\n',confusion_matrix(y_test, test_preds))\n",
    "print(classification_report(y_test,test_preds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X_train' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-87-af365212e028>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mclf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRMTClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mtest_preds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepsilon_multiple\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'confusion_matrix \\n'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mconfusion_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_preds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X_train' is not defined"
     ]
    }
   ],
   "source": [
    "clf = RMTClassifier()\n",
    "clf.fit(X_train, y_train)\n",
    "test_preds = clf.predict(X_test, epsilon_multiple = 1)\n",
    "\n",
    "print('confusion_matrix \\n',confusion_matrix(y_test, test_preds))\n",
    "print(classification_report(y_test,test_preds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
